// aws-cdk-lib/aws-lambda/lib/runtime-management.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeManagementMode = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Specify the runtime update mode.
 */
class RuntimeManagementMode {
    /**
     * You specify a runtime version in your function configuration.
     * The function uses this runtime version indefinitely.
     * In the rare case in which a new runtime version is incompatible with an existing function,
     * you can use this mode to roll back your function to an earlier runtime version.
     */
    static manual(arn) {
        return new RuntimeManagementMode('Manual', arn);
    }
    constructor(mode, arn) {
        this.mode = mode;
        this.arn = arn;
        if (arn) {
            this.runtimeManagementConfig = {
                runtimeVersionArn: arn,
                updateRuntimeOn: mode,
            };
        }
        else {
            this.runtimeManagementConfig = {
                updateRuntimeOn: mode,
            };
        }
    }
}
_a = JSII_RTTI_SYMBOL_1;
RuntimeManagementMode[_a] = { fqn: "aws-cdk-lib.aws_lambda.RuntimeManagementMode", version: "2.75.1" };
/**
 * Automatically update to the most recent and secure runtime version using Two-phase runtime version rollout.
 * We recommend this mode for most customers so that you always benefit from runtime updates.
 */
RuntimeManagementMode.AUTO = new RuntimeManagementMode('Auto');
/**
 * When you update your function, Lambda updates the runtime of your function to the most recent and secure runtime version.
 * This approach synchronizes runtime updates with function deployments,
 * giving you control over when Lambda applies runtime updates.
 * With this mode, you can detect and mitigate rare runtime update incompatibilities early.
 * When using this mode, you must regularly update your functions to keep their runtime up to date.
 */
RuntimeManagementMode.FUNCTION_UPDATE = new RuntimeManagementMode('FunctionUpdate');
exports.RuntimeManagementMode = RuntimeManagementMode;
