// aws-cdk-lib/pipelines/lib/blueprint/file-set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSet = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * A set of files traveling through the deployment pipeline
 *
 * Individual steps in the pipeline produce or consume
 * `FileSet`s.
 */
class FileSet {
    constructor(
    /** Human-readable descriptor for this file set (does not need to be unique) */
    id, producer) {
        this.id = id;
        /**
         * The primary output of a file set producer
         *
         * The primary output of a FileSet is itself.
         */
        this.primaryOutput = this;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(producer);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, FileSet);
            }
            throw error;
        }
        this._producer = producer;
    }
    /**
     * The Step that produces this FileSet
     */
    get producer() {
        if (!this._producer) {
            throw new Error(`FileSet '${this.id}' doesn\'t have a producer; call 'fileSet.producedBy()'`);
        }
        return this._producer;
    }
    /**
     * Mark the given Step as the producer for this FileSet
     *
     * This method can only be called once.
     */
    producedBy(producer) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(producer);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.producedBy);
            }
            throw error;
        }
        if (this._producer) {
            throw new Error(`FileSet '${this.id}' already has a producer (${this._producer}) while setting producer: ${producer}`);
        }
        this._producer = producer;
    }
    /**
     * Return a string representation of this FileSet
     */
    toString() {
        return `FileSet(${this.id})`;
    }
}
_a = JSII_RTTI_SYMBOL_1;
FileSet[_a] = { fqn: "aws-cdk-lib.pipelines.FileSet", version: "2.75.1" };
exports.FileSet = FileSet;
