// aws-cdk-lib/pipelines/lib/blueprint/step.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Step = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../../core");
const step_output_1 = require("../helpers-internal/step-output");
/**
 * A generic Step which can be added to a Pipeline
 *
 * Steps can be used to add Sources, Build Actions and Validations
 * to your pipeline.
 *
 * This class is abstract. See specific subclasses of Step for
 * useful steps to add to your Pipeline
 */
class Step {
    /**
     * Define a sequence of steps to be executed in order.
     *
     * If you need more fine-grained step ordering, use the `addStepDependency()`
     * API. For example, if you want `secondStep` to occur after `firstStep`, call
     * `secondStep.addStepDependency(firstStep)`.
     */
    static sequence(steps) {
        for (let i = 1; i < steps.length; i++) {
            steps[i].addStepDependency(steps[i - 1]);
        }
        return steps;
    }
    constructor(
    /** Identifier for this step */
    id) {
        this.id = id;
        /**
         * The list of FileSets consumed by this Step
         */
        this.dependencyFileSets = [];
        /**
         * Whether or not this is a Source step
         *
         * What it means to be a Source step depends on the engine.
         */
        this.isSource = false;
        this._dependencies = new Set();
        if (core_1.Token.isUnresolved(id)) {
            throw new Error(`Step id cannot be unresolved, got '${id}'`);
        }
    }
    /**
     * Return the steps this step depends on, based on the FileSets it requires
     */
    get dependencies() {
        return Array.from(new Set([
            ...this.dependencyFileSets.map(f => f.producer),
            ...this._dependencies,
        ]));
    }
    /**
     * Return a string representation of this Step
     */
    toString() {
        return `${this.constructor.name}(${this.id})`;
    }
    /**
     * The primary FileSet produced by this Step
     *
     * Not all steps produce an output FileSet--if they do
     * you can substitute the `Step` object for the `FileSet` object.
     */
    get primaryOutput() {
        // Accessor so it can be mutable in children
        return this._primaryOutput;
    }
    /**
     * Add a dependency on another step.
     */
    addStepDependency(step) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(step);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.addStepDependency);
            }
            throw error;
        }
        this._dependencies.add(step);
    }
    /**
     * Add an additional FileSet to the set of file sets required by this step
     *
     * This will lead to a dependency on the producer of that file set.
     */
    addDependencyFileSet(fs) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_pipelines_FileSet(fs);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.addDependencyFileSet);
            }
            throw error;
        }
        this.dependencyFileSets.push(fs);
    }
    /**
     * Configure the given FileSet as the primary output of this step
     */
    configurePrimaryOutput(fs) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_pipelines_FileSet(fs);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.configurePrimaryOutput);
            }
            throw error;
        }
        this._primaryOutput = fs;
    }
    /**
     * Crawl the given structure for references to StepOutputs and add dependencies on all steps found
     *
     * Should be called in the constructor of subclasses based on what the user
     * passes in as construction properties. The format of the structure passed in
     * here does not have to correspond exactly to what gets rendered into the
     * engine, it just needs to contain the same data.
     */
    discoverReferencedOutputs(structure) {
        for (const output of step_output_1.StepOutput.findAll(structure)) {
            this._dependencies.add(output.step);
            step_output_1.StepOutput.recordProducer(output);
        }
    }
    /**
     * StackOutputReferences this step consumes.
     */
    get consumedStackOutputs() {
        return [];
    }
}
_a = JSII_RTTI_SYMBOL_1;
Step[_a] = { fqn: "aws-cdk-lib.pipelines.Step", version: "2.75.1" };
exports.Step = Step;
