// aws-cdk-lib/aws-stepfunctions-tasks/lib/dynamodb/private/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJsonPath = exports.transformAttributeValueMap = exports.getDynamoResourceArn = exports.DynamoMethod = void 0;
const sfn = require("../../../../aws-stepfunctions");
const task_utils_1 = require("../../private/task-utils");
var DynamoMethod;
(function (DynamoMethod) {
    DynamoMethod["GET"] = "Get";
    DynamoMethod["PUT"] = "Put";
    DynamoMethod["DELETE"] = "Delete";
    DynamoMethod["UPDATE"] = "Update";
})(DynamoMethod = exports.DynamoMethod || (exports.DynamoMethod = {}));
function getDynamoResourceArn(method) {
    return (0, task_utils_1.integrationResourceArn)('dynamodb', `${method.toLowerCase()}Item`, sfn.IntegrationPattern.REQUEST_RESPONSE);
}
exports.getDynamoResourceArn = getDynamoResourceArn;
function transformAttributeValueMap(attrMap) {
    const transformedValue = {};
    for (const key in attrMap) {
        if (key) {
            transformedValue[key] = attrMap[key].toObject();
        }
    }
    return attrMap ? transformedValue : undefined;
}
exports.transformAttributeValueMap = transformAttributeValueMap;
function validateJsonPath(value) {
    if (!value.startsWith('$')) {
        throw new Error("Data JSON path values must either be exactly equal to '$' or start with '$.'");
    }
}
exports.validateJsonPath = validateJsonPath;
