// aws-cdk-lib/aws-codepipeline-actions/lib/elastic-beanstalk/deploy-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticBeanstalkDeployAction = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const codepipeline = require("../../../aws-codepipeline");
const action_1 = require("../action");
const common_1 = require("../common");
/**
 * CodePipeline action to deploy an AWS ElasticBeanstalk Application.
 */
class ElasticBeanstalkDeployAction extends action_1.Action {
    constructor(props) {
        super({
            ...props,
            category: codepipeline.ActionCategory.DEPLOY,
            provider: 'ElasticBeanstalk',
            artifactBounds: (0, common_1.deployArtifactBounds)(),
            inputs: [props.input],
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_codepipeline_actions_ElasticBeanstalkDeployActionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, ElasticBeanstalkDeployAction);
            }
            throw error;
        }
        this.applicationName = props.applicationName;
        this.environmentName = props.environmentName;
    }
    bound(_scope, _stage, options) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_codepipeline_IStage(_stage);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_codepipeline_ActionBindOptions(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bound);
            }
            throw error;
        }
        // Per https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/AWSHowTo.iam.managed-policies.html
        // it doesn't seem we can scope this down further for the codepipeline action.
        options.role.addManagedPolicy({ managedPolicyArn: 'arn:aws:iam::aws:policy/AdministratorAccess-AWSElasticBeanstalk' });
        // the Action's Role needs to read from the Bucket to get artifacts
        options.bucket.grantRead(options.role);
        return {
            configuration: {
                ApplicationName: this.applicationName,
                EnvironmentName: this.environmentName,
            },
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkDeployAction[_a] = { fqn: "aws-cdk-lib.aws_codepipeline_actions.ElasticBeanstalkDeployAction", version: "2.75.1" };
exports.ElasticBeanstalkDeployAction = ElasticBeanstalkDeployAction;
