// aws-cdk-lib/aws-cognito/lib/user-pool-idps/saml.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPoolIdentityProviderSaml = exports.UserPoolIdentityProviderSamlMetadata = exports.UserPoolIdentityProviderSamlMetadataType = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../../core");
const user_pool_idp_base_1 = require("./private/user-pool-idp-base");
const cognito_generated_1 = require("../cognito.generated");
/**
 * Metadata types that can be used for a SAML user pool identity provider.
 */
var UserPoolIdentityProviderSamlMetadataType;
(function (UserPoolIdentityProviderSamlMetadataType) {
    /** Metadata provided via a URL. */
    UserPoolIdentityProviderSamlMetadataType["URL"] = "url";
    /** Metadata provided via the contents of a file. */
    UserPoolIdentityProviderSamlMetadataType["FILE"] = "file";
})(UserPoolIdentityProviderSamlMetadataType = exports.UserPoolIdentityProviderSamlMetadataType || (exports.UserPoolIdentityProviderSamlMetadataType = {}));
/**
 * Metadata for a SAML user pool identity provider.
 */
class UserPoolIdentityProviderSamlMetadata {
    /**
     * Specify SAML metadata via a URL.
     */
    static url(url) {
        return new UserPoolIdentityProviderSamlMetadata(url, UserPoolIdentityProviderSamlMetadataType.URL);
    }
    /**
     * Specify SAML metadata via the contents of a file.
     */
    static file(fileContent) {
        return new UserPoolIdentityProviderSamlMetadata(fileContent, UserPoolIdentityProviderSamlMetadataType.FILE);
    }
    /**
     * Construct the metadata for a SAML identity provider.
     *
     * @param metadataContent A URL hosting SAML metadata, or the content of a file containing SAML metadata.
     * @param metadataType The type of metadata, either a URL or file content.
     */
    constructor(metadataContent, metadataType) {
        this.metadataContent = metadataContent;
        this.metadataType = metadataType;
    }
}
_a = JSII_RTTI_SYMBOL_1;
UserPoolIdentityProviderSamlMetadata[_a] = { fqn: "aws-cdk-lib.aws_cognito.UserPoolIdentityProviderSamlMetadata", version: "2.75.1" };
exports.UserPoolIdentityProviderSamlMetadata = UserPoolIdentityProviderSamlMetadata;
/**
 * Represents a identity provider that integrates with SAML.
 * @resource AWS::Cognito::UserPoolIdentityProvider
 */
class UserPoolIdentityProviderSaml extends user_pool_idp_base_1.UserPoolIdentityProviderBase {
    constructor(scope, id, props) {
        super(scope, id, props);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cognito_UserPoolIdentityProviderSamlProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, UserPoolIdentityProviderSaml);
            }
            throw error;
        }
        this.validateName(props.name);
        const { metadataType, metadataContent } = props.metadata;
        const resource = new cognito_generated_1.CfnUserPoolIdentityProvider(this, 'Resource', {
            userPoolId: props.userPool.userPoolId,
            providerName: this.getProviderName(props.name),
            providerType: 'SAML',
            providerDetails: {
                IDPSignout: props.idpSignout ?? false,
                MetadataURL: metadataType === UserPoolIdentityProviderSamlMetadataType.URL ? metadataContent : undefined,
                MetadataFile: metadataType === UserPoolIdentityProviderSamlMetadataType.FILE ? metadataContent : undefined,
            },
            idpIdentifiers: props.identifiers,
            attributeMapping: super.configureAttributeMapping(),
        });
        this.providerName = super.getResourceNameAttribute(resource.ref);
    }
    getProviderName(name) {
        if (name) {
            this.validateName(name);
            return name;
        }
        const uniqueName = core_1.Names.uniqueResourceName(this, {
            maxLength: 32,
        });
        if (uniqueName.length < 3) {
            return `${uniqueName}saml`;
        }
        return uniqueName;
    }
    validateName(name) {
        if (name && !core_1.Token.isUnresolved(name) && (name.length < 3 || name.length > 32)) {
            throw new Error(`Expected provider name to be between 3 and 32 characters, received ${name} (${name.length} characters)`);
        }
    }
}
_b = JSII_RTTI_SYMBOL_1;
UserPoolIdentityProviderSaml[_b] = { fqn: "aws-cdk-lib.aws_cognito.UserPoolIdentityProviderSaml", version: "2.75.1" };
exports.UserPoolIdentityProviderSaml = UserPoolIdentityProviderSaml;
