// aws-cdk-lib/aws-codebuild/lib/report-group.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportGroup = exports.ReportGroupType = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("../../aws-iam");
const cdk = require("../../core");
const codebuild_generated_1 = require("./codebuild.generated");
const report_group_utils_1 = require("./report-group-utils");
class ReportGroupBase extends cdk.Resource {
    grantWrite(identity) {
        const typeAction = this.type === ReportGroupType.CODE_COVERAGE ? 'codebuild:BatchPutCodeCoverages' : 'codebuild:BatchPutTestCases';
        const ret = iam.Grant.addToPrincipal({
            grantee: identity,
            actions: [
                'codebuild:CreateReport',
                'codebuild:UpdateReport',
                typeAction,
            ],
            resourceArns: [this.reportGroupArn],
        });
        if (this.exportBucket) {
            this.exportBucket.grantWrite(identity);
        }
        return ret;
    }
}
/**
 * The type of reports in the report group.
 */
var ReportGroupType;
(function (ReportGroupType) {
    /**
     * The report group contains test reports.
     */
    ReportGroupType["TEST"] = "TEST";
    /**
     * The report group contains code coverage reports.
     */
    ReportGroupType["CODE_COVERAGE"] = "CODE_COVERAGE";
})(ReportGroupType = exports.ReportGroupType || (exports.ReportGroupType = {}));
/**
 * The ReportGroup resource class.
 */
class ReportGroup extends ReportGroupBase {
    /**
     * Reference an existing ReportGroup,
     * defined outside of the CDK code,
     * by name.
     */
    static fromReportGroupName(scope, id, reportGroupName) {
        class Import extends ReportGroupBase {
            constructor() {
                super(...arguments);
                this.reportGroupName = reportGroupName;
                this.reportGroupArn = (0, report_group_utils_1.renderReportGroupArn)(scope, reportGroupName);
                this.exportBucket = undefined;
                this.type = undefined;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props = {}) {
        super(scope, id, {
            physicalName: props.reportGroupName,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_codebuild_ReportGroupProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, ReportGroup);
            }
            throw error;
        }
        this.type = props.type ? props.type : ReportGroupType.TEST;
        const resource = new codebuild_generated_1.CfnReportGroup(this, 'Resource', {
            type: this.type,
            exportConfig: {
                exportConfigType: props.exportBucket ? 'S3' : 'NO_EXPORT',
                s3Destination: props.exportBucket
                    ? {
                        bucket: props.exportBucket.bucketName,
                        encryptionDisabled: props.exportBucket.encryptionKey ? false : undefined,
                        encryptionKey: props.exportBucket.encryptionKey?.keyArn,
                        packaging: props.zipExport ? 'ZIP' : undefined,
                    }
                    : undefined,
            },
            name: props.reportGroupName,
        });
        resource.applyRemovalPolicy(props.removalPolicy, {
            default: cdk.RemovalPolicy.RETAIN,
        });
        this.reportGroupArn = this.getResourceArnAttribute(resource.attrArn, (0, report_group_utils_1.reportGroupArnComponents)(this.physicalName));
        this.reportGroupName = this.getResourceNameAttribute(
        // there is no separate name attribute,
        // so use Fn::Select + Fn::Split to make one
        cdk.Fn.select(1, cdk.Fn.split('/', resource.ref)));
        this.exportBucket = props.exportBucket;
    }
}
_a = JSII_RTTI_SYMBOL_1;
ReportGroup[_a] = { fqn: "aws-cdk-lib.aws_codebuild.ReportGroup", version: "2.75.1" };
exports.ReportGroup = ReportGroup;
