// aws-cdk-lib/aws-elasticloadbalancingv2-targets/lib/alb-target.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlbTarget = exports.AlbArnTarget = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const elbv2 = require("../../aws-elasticloadbalancingv2");
/**
 * A single Application Load Balancer as the target for load balancing.
 */
class AlbArnTarget {
    /**
     * Create a new alb target
     *
     * @param albArn The ARN of the application load balancer to load balance to
     * @param port The port on which the target is listening
     */
    constructor(albArn, port) {
        this.albArn = albArn;
        this.port = port;
    }
    /**
     * Register this alb target with a load balancer
     *
     * Don't call this, it is called automatically when you add the target to a
     * load balancer.
     */
    attachToNetworkTargetGroup(targetGroup) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_INetworkTargetGroup(targetGroup);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.attachToNetworkTargetGroup);
            }
            throw error;
        }
        return this.attach(targetGroup);
    }
    attach(_targetGroup) {
        return {
            targetType: elbv2.TargetType.ALB,
            targetJson: { id: this.albArn, port: this.port },
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
AlbArnTarget[_a] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_targets.AlbArnTarget", version: "2.75.1" };
exports.AlbArnTarget = AlbArnTarget;
/**
 * A single Application Load Balancer as the target for load balancing.
 */
class AlbTarget extends AlbArnTarget {
    /**
     * @param alb The application load balancer to load balance to
     * @param port The port on which the target is listening
     */
    constructor(alb, port) {
        super(alb.loadBalancerArn, port);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_ApplicationLoadBalancer(alb);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, AlbTarget);
            }
            throw error;
        }
    }
}
_b = JSII_RTTI_SYMBOL_1;
AlbTarget[_b] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_targets.AlbTarget", version: "2.75.1" };
exports.AlbTarget = AlbTarget;
