// aws-cdk-lib/aws-fsx/lib/fsx-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FSxMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class FSxMetrics {
    static dataReadBytesSum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'DataReadBytes',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static dataWriteBytesSum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'DataWriteBytes',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static freeStorageCapacityAverage(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'FreeStorageCapacity',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static freeDataStorageCapacitySum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'FreeDataStorageCapacity',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static dataReadOperationsSum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'DataReadOperations',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static dataWriteOperationsSum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'DataWriteOperations',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static metadataOperationsSum(dimensions) {
        return {
            namespace: 'AWS/FSx',
            metricName: 'MetadataOperations',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.FSxMetrics = FSxMetrics;
