// aws-cdk-lib/aws-globalaccelerator/lib/accelerator.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Accelerator = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("../../core");
const ga = require("./globalaccelerator.generated");
const listener_1 = require("./listener");
/**
 * The Accelerator construct
 */
class Accelerator extends cdk.Resource {
    /**
     * import from attributes
     */
    static fromAcceleratorAttributes(scope, id, attrs) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_AcceleratorAttributes(attrs);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromAcceleratorAttributes);
            }
            throw error;
        }
        class Import extends cdk.Resource {
            constructor() {
                super(...arguments);
                this.acceleratorArn = attrs.acceleratorArn;
                this.dnsName = attrs.dnsName;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props = {}) {
        super(scope, id);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_AcceleratorProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, Accelerator);
            }
            throw error;
        }
        this.validateAcceleratorName(props.acceleratorName);
        const name = props.acceleratorName ?? cdk.Names.uniqueResourceName(this, {
            maxLength: 64,
        });
        const resource = new ga.CfnAccelerator(this, 'Resource', {
            enabled: props.enabled ?? true,
            name,
        });
        this.acceleratorArn = resource.attrAcceleratorArn;
        this.dnsName = resource.attrDnsName;
    }
    /**
     * Add a listener to the accelerator
     */
    addListener(id, options) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_ListenerOptions(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.addListener);
            }
            throw error;
        }
        return new listener_1.Listener(this, id, {
            accelerator: this,
            ...options,
        });
    }
    validateAcceleratorName(name) {
        if (!cdk.Token.isUnresolved(name) && name !== undefined && (name.length < 1 || name.length > 64)) {
            throw new Error(`Invalid acceleratorName value ${name}, must have length between 1 and 64, got: ${name.length}`);
        }
    }
}
_a = JSII_RTTI_SYMBOL_1;
Accelerator[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator.Accelerator", version: "2.75.1" };
exports.Accelerator = Accelerator;
