// aws-cdk-lib/aws-cognito/lib/user-pool-email.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPoolEmail = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const punycode_1 = require("punycode/");
/**
 * Configure how Cognito sends emails
 */
class UserPoolEmail {
    /**
     * Send email using Cognito
     */
    static withCognito(replyTo) {
        return new CognitoEmail(replyTo);
    }
    /**
     * Send email using SES
     */
    static withSES(options) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cognito_UserPoolSESOptions(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.withSES);
            }
            throw error;
        }
        return new SESEmail(options);
    }
}
_a = JSII_RTTI_SYMBOL_1;
UserPoolEmail[_a] = { fqn: "aws-cdk-lib.aws_cognito.UserPoolEmail", version: "2.75.1" };
exports.UserPoolEmail = UserPoolEmail;
class CognitoEmail extends UserPoolEmail {
    constructor(replyTo) {
        super();
        this.replyTo = replyTo;
    }
    _bind(_scope) {
        return {
            replyToEmailAddress: encodeAndTest(this.replyTo),
            emailSendingAccount: 'COGNITO_DEFAULT',
        };
    }
}
class SESEmail extends UserPoolEmail {
    constructor(options) {
        super();
        this.options = options;
    }
    _bind(scope) {
        const region = core_1.Stack.of(scope).region;
        if (core_1.Token.isUnresolved(region) && !this.options.sesRegion) {
            throw new Error('Your stack region cannot be determined so "sesRegion" is required in SESOptions');
        }
        let from = encodeAndTest(this.options.fromEmail);
        if (this.options.fromName) {
            const fromName = formatFromName(this.options.fromName);
            from = `${fromName} <${from}>`;
        }
        if (this.options.sesVerifiedDomain) {
            const domainFromEmail = this.options.fromEmail.split('@').pop();
            if (domainFromEmail !== this.options.sesVerifiedDomain) {
                throw new Error('"fromEmail" contains a different domain than the "sesVerifiedDomain"');
            }
        }
        return {
            from,
            replyToEmailAddress: encodeAndTest(this.options.replyTo),
            configurationSet: this.options.configurationSetName,
            emailSendingAccount: 'DEVELOPER',
            sourceArn: core_1.Stack.of(scope).formatArn({
                service: 'ses',
                resource: 'identity',
                resourceName: encodeAndTest(this.options.sesVerifiedDomain ?? this.options.fromEmail),
                region: this.options.sesRegion ?? region,
            }),
        };
    }
}
function encodeAndTest(input) {
    if (input) {
        const local = input.split('@')[0];
        if (!/[\p{ASCII}]+/u.test(local)) {
            throw new Error('the local part of the email address must use ASCII characters only');
        }
        return (0, punycode_1.toASCII)(input);
    }
    else {
        return undefined;
    }
}
/**
 * Formats `fromName` to comply RFC 5322
 *
 * @see https://www.rfc-editor.org/rfc/rfc5322#section-3.4
 */
function formatFromName(fromName) {
    // mime encode for non US-ASCII characters
    // see RFC 2047 for details https://www.rfc-editor.org/rfc/rfc2047
    if (!isAscii(fromName)) {
        const base64Name = Buffer.from(fromName, 'utf-8').toString('base64');
        return `=?UTF-8?B?${base64Name}?=`;
    }
    // makes a quoted-string unless fromName is a phrase (only atext and space)
    // or a quoted-string already
    if (!(isSimplePhrase(fromName) || isQuotedString(fromName))) {
        // in quoted-string, `\` and `"` should be escaped by `\`
        // e.g. `"foo \"bar\" \\baz"`
        const quotedName = fromName.replace(/[\\"]/g, (ch) => `\\${ch}`);
        return `"${quotedName}"`;
    }
    // otherwise, returns as is
    return fromName;
}
/**
 * Returns whether the input is a printable US-ASCII string
 */
function isAscii(input) {
    // U+0020 (space) - U+007E (`~`)
    return /^[\u0020-\u007E]+$/u.test(input);
}
/**
 * Returns whether the input is a phrase excluding quoted-string
 *
 * @see https://www.rfc-editor.org/rfc/rfc5322#section-3.2
 */
function isSimplePhrase(input) {
    return /^[\w !#$%&'*+-\/=?^_`{|}~]+$/.test(input);
}
/**
 * Returns whether the input is already a quoted-string
 *
 * @see https://www.rfc-editor.org/rfc/rfc5322#section-3.2.4
 */
function isQuotedString(input) {
    // in quoted-string, `\` and `"` should be esacaped by `\`
    //
    // match: `"foo.bar"` / `"foo \"bar\""` / `"foo \\ bar"`
    // not match: `"bare " dquote"` / `"unclosed escape \"` / `"unclosed dquote`
    return /^"(?:[^\\"]|\\.)*"$/.test(input);
}
