// aws-cdk-lib/aws-cognito/lib/user-pool-resource-server.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPoolResourceServer = exports.ResourceServerScope = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const cognito_generated_1 = require("./cognito.generated");
/**
 * A scope for ResourceServer
 */
class ResourceServerScope {
    constructor(props) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cognito_ResourceServerScopeProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, ResourceServerScope);
            }
            throw error;
        }
        this.scopeName = props.scopeName;
        this.scopeDescription = props.scopeDescription;
    }
}
_a = JSII_RTTI_SYMBOL_1;
ResourceServerScope[_a] = { fqn: "aws-cdk-lib.aws_cognito.ResourceServerScope", version: "2.75.1" };
exports.ResourceServerScope = ResourceServerScope;
/**
 * Defines a User Pool OAuth2.0 Resource Server
 */
class UserPoolResourceServer extends core_1.Resource {
    /**
     * Import a user pool resource client given its id.
     */
    static fromUserPoolResourceServerId(scope, id, userPoolResourceServerId) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.userPoolResourceServerId = userPoolResourceServerId;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props) {
        super(scope, id, {
            physicalName: props.identifier,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cognito_UserPoolResourceServerProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, UserPoolResourceServer);
            }
            throw error;
        }
        const resource = new cognito_generated_1.CfnUserPoolResourceServer(this, 'Resource', {
            identifier: this.physicalName,
            name: props.userPoolResourceServerName ?? this.physicalName,
            scopes: props.scopes,
            userPoolId: props.userPool.userPoolId,
        });
        this.userPoolResourceServerId = resource.ref;
    }
}
_b = JSII_RTTI_SYMBOL_1;
UserPoolResourceServer[_b] = { fqn: "aws-cdk-lib.aws_cognito.UserPoolResourceServer", version: "2.75.1" };
exports.UserPoolResourceServer = UserPoolResourceServer;
