// aws-cdk-lib/aws-stepfunctions-tasks/lib/aws-sdk/call-aws-service.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallAwsService = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("../../../aws-iam");
const sfn = require("../../../aws-stepfunctions");
const core_1 = require("../../../core");
const task_utils_1 = require("../private/task-utils");
/**
 * A StepFunctions task to call an AWS service API
 */
class CallAwsService extends sfn.TaskStateBase {
    constructor(scope, id, props) {
        super(scope, id, props);
        this.props = props;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_stepfunctions_tasks_CallAwsServiceProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CallAwsService);
            }
            throw error;
        }
        if (this.props.integrationPattern === sfn.IntegrationPattern.RUN_JOB) {
            throw new Error('The RUN_JOB integration pattern is not supported for CallAwsService');
        }
        const iamServiceMap = {
            sfn: 'states',
        };
        const iamService = iamServiceMap[props.service] ?? props.service;
        this.taskPolicies = [
            new iam.PolicyStatement({
                resources: props.iamResources,
                // The prefix and the action name are case insensitive
                // https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_action.html
                actions: [props.iamAction ?? `${iamService}:${props.action}`],
            }),
            ...props.additionalIamStatements ?? [],
        ];
    }
    /**
     * @internal
     */
    _renderTask() {
        let service = this.props.service;
        if (!core_1.Token.isUnresolved(service)) {
            service = service.toLowerCase();
        }
        return {
            Resource: (0, task_utils_1.integrationResourceArn)('aws-sdk', `${service}:${this.props.action}`, this.props.integrationPattern),
            Parameters: sfn.FieldUtils.renderObject(this.props.parameters) ?? {}, // Parameters is required for aws-sdk
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
CallAwsService[_a] = { fqn: "aws-cdk-lib.aws_stepfunctions_tasks.CallAwsService", version: "2.75.1" };
exports.CallAwsService = CallAwsService;
