// aws-cdk-lib/aws-gamelift/lib/gamelift-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameLiftMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class GameLiftMetrics {
    static activeInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'ActiveInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static percentIdleInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'PercentIdleInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static desiredInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'DesiredInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static idleInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'IdleInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static maxInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'MaxInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static minInstancesAverage(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'MinInstances',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static instanceInterruptionsSum(dimensions) {
        return {
            namespace: 'AWS/GameLift',
            metricName: 'InstanceInterruptions',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.GameLiftMetrics = GameLiftMetrics;
