// aws-cdk-lib/aws-docdb/lib/docdb-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocDBMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class DocDBMetrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/DocDB',
            metricName: 'CPUUtilization',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static databaseConnectionsAverage(dimensions) {
        return {
            namespace: 'AWS/DocDB',
            metricName: 'DatabaseConnections',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static engineUptimeAverage(dimensions) {
        return {
            namespace: 'AWS/DocDB',
            metricName: 'EngineUptime',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static readThroughputSum(dimensions) {
        return {
            namespace: 'AWS/DocDB',
            metricName: 'ReadThroughput',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static writeThroughputSum(dimensions) {
        return {
            namespace: 'AWS/DocDB',
            metricName: 'WriteThroughput',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.DocDBMetrics = DocDBMetrics;
