// aws-cdk-lib/pipelines/lib/legacy/validation/_files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Files = void 0;
const s3assets = require("../../../../aws-s3-assets");
/**
 * Additional files to use in a shell script
 */
class Files {
    /**
     * Use the files from a CodePipeline artifact
     */
    static fromArtifact(artifact) {
        if (!artifact) {
            // Typechecking may mess up
            throw new Error('Files.fromArtifact(): input artifact is required, got undefined');
        }
        return {
            bind: () => ({ artifact }),
            grantRead: () => { },
        };
    }
    /**
     * Create a new asset to bundle up the files in a directory on disk
     */
    static fromDirectory(directoryPath) {
        let realFiles;
        return {
            bind(scope) {
                realFiles = Files.fromAsset(new s3assets.Asset(scope, directoryPath, {
                    path: directoryPath,
                }));
                return realFiles.bind(scope);
            },
            grantRead(grantee) {
                if (!realFiles) {
                    throw new Error('bind() must be called first');
                }
                realFiles.grantRead(grantee);
            },
        };
    }
    /**
     * Use an existing asset as a file source
     */
    static fromAsset(asset) {
        return {
            bind: () => ({
                commands: [
                    `echo "Downloading additional files from ${asset.s3ObjectUrl}"`,
                    `aws s3 cp ${asset.s3ObjectUrl} /tmp/dl.zip`,
                    'unzip /tmp/dl.zip -d .',
                ],
            }),
            grantRead: (grantee) => asset.grantRead(grantee),
        };
    }
    constructor() {
    }
}
exports.Files = Files;
