// aws-cdk-lib/aws-pipes/lib/pipes.generated.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnPipe = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("../../core");
const cfn_parse = require("../../core/lib/helpers-internal");
/**
 * Determine whether the given properties match those of a `CfnPipeProps`
 *
 * @param properties - the TypeScript properties of a `CfnPipeProps`
 *
 * @returns the result of the validation.
 */
function CfnPipePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('desiredState', cdk.validateString)(properties.desiredState));
    errors.collect(cdk.propertyValidator('enrichment', cdk.validateString)(properties.enrichment));
    errors.collect(cdk.propertyValidator('enrichmentParameters', CfnPipe_PipeEnrichmentParametersPropertyValidator)(properties.enrichmentParameters));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('roleArn', cdk.requiredValidator)(properties.roleArn));
    errors.collect(cdk.propertyValidator('roleArn', cdk.validateString)(properties.roleArn));
    errors.collect(cdk.propertyValidator('source', cdk.requiredValidator)(properties.source));
    errors.collect(cdk.propertyValidator('source', cdk.validateString)(properties.source));
    errors.collect(cdk.propertyValidator('sourceParameters', CfnPipe_PipeSourceParametersPropertyValidator)(properties.sourceParameters));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    errors.collect(cdk.propertyValidator('target', cdk.requiredValidator)(properties.target));
    errors.collect(cdk.propertyValidator('target', cdk.validateString)(properties.target));
    errors.collect(cdk.propertyValidator('targetParameters', CfnPipe_PipeTargetParametersPropertyValidator)(properties.targetParameters));
    return errors.wrap('supplied properties not correct for "CfnPipeProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe` resource
 *
 * @param properties - the TypeScript properties of a `CfnPipeProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe` resource.
 */
// @ts-ignore TS6133
function cfnPipePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipePropsValidator(properties).assertSuccess();
    return {
        RoleArn: cdk.stringToCloudFormation(properties.roleArn),
        Source: cdk.stringToCloudFormation(properties.source),
        Target: cdk.stringToCloudFormation(properties.target),
        Description: cdk.stringToCloudFormation(properties.description),
        DesiredState: cdk.stringToCloudFormation(properties.desiredState),
        Enrichment: cdk.stringToCloudFormation(properties.enrichment),
        EnrichmentParameters: cfnPipePipeEnrichmentParametersPropertyToCloudFormation(properties.enrichmentParameters),
        Name: cdk.stringToCloudFormation(properties.name),
        SourceParameters: cfnPipePipeSourceParametersPropertyToCloudFormation(properties.sourceParameters),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
        TargetParameters: cfnPipePipeTargetParametersPropertyToCloudFormation(properties.targetParameters),
    };
}
// @ts-ignore TS6133
function CfnPipePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('roleArn', 'RoleArn', cfn_parse.FromCloudFormation.getString(properties.RoleArn));
    ret.addPropertyResult('source', 'Source', cfn_parse.FromCloudFormation.getString(properties.Source));
    ret.addPropertyResult('target', 'Target', cfn_parse.FromCloudFormation.getString(properties.Target));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('desiredState', 'DesiredState', properties.DesiredState != null ? cfn_parse.FromCloudFormation.getString(properties.DesiredState) : undefined);
    ret.addPropertyResult('enrichment', 'Enrichment', properties.Enrichment != null ? cfn_parse.FromCloudFormation.getString(properties.Enrichment) : undefined);
    ret.addPropertyResult('enrichmentParameters', 'EnrichmentParameters', properties.EnrichmentParameters != null ? CfnPipePipeEnrichmentParametersPropertyFromCloudFormation(properties.EnrichmentParameters) : undefined);
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('sourceParameters', 'SourceParameters', properties.SourceParameters != null ? CfnPipePipeSourceParametersPropertyFromCloudFormation(properties.SourceParameters) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addPropertyResult('targetParameters', 'TargetParameters', properties.TargetParameters != null ? CfnPipePipeTargetParametersPropertyFromCloudFormation(properties.TargetParameters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Pipes::Pipe`
 *
 * Create a pipe. Amazon EventBridge Pipes connect event sources to targets and reduces the need for specialized knowledge and integration code.
 *
 * @cloudformationResource AWS::Pipes::Pipe
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-pipes-pipe.html
 */
class CfnPipe extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnPipePropsFromCloudFormation(resourceProperties);
        const ret = new CfnPipe(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Pipes::Pipe`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnPipe.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_pipes_CfnPipeProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnPipe);
            }
            throw error;
        }
        cdk.requireProperty(props, 'roleArn', this);
        cdk.requireProperty(props, 'source', this);
        cdk.requireProperty(props, 'target', this);
        this.attrArn = cdk.Token.asString(this.getAtt('Arn', cdk.ResolutionTypeHint.STRING));
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrCurrentState = cdk.Token.asString(this.getAtt('CurrentState', cdk.ResolutionTypeHint.STRING));
        this.attrLastModifiedTime = cdk.Token.asString(this.getAtt('LastModifiedTime', cdk.ResolutionTypeHint.STRING));
        this.attrStateReason = cdk.Token.asString(this.getAtt('StateReason', cdk.ResolutionTypeHint.STRING));
        this.roleArn = props.roleArn;
        this.source = props.source;
        this.target = props.target;
        this.description = props.description;
        this.desiredState = props.desiredState;
        this.enrichment = props.enrichment;
        this.enrichmentParameters = props.enrichmentParameters;
        this.name = props.name;
        this.sourceParameters = props.sourceParameters;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Pipes::Pipe", props.tags, { tagPropertyName: 'tags' });
        this.targetParameters = props.targetParameters;
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnPipe.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            roleArn: this.roleArn,
            source: this.source,
            target: this.target,
            description: this.description,
            desiredState: this.desiredState,
            enrichment: this.enrichment,
            enrichmentParameters: this.enrichmentParameters,
            name: this.name,
            sourceParameters: this.sourceParameters,
            tags: this.tags.renderTags(),
            targetParameters: this.targetParameters,
        };
    }
    renderProperties(props) {
        return cfnPipePropsToCloudFormation(props);
    }
}
_a = JSII_RTTI_SYMBOL_1;
CfnPipe[_a] = { fqn: "aws-cdk-lib.aws_pipes.CfnPipe", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnPipe.CFN_RESOURCE_TYPE_NAME = "AWS::Pipes::Pipe";
exports.CfnPipe = CfnPipe;
/**
 * Determine whether the given properties match those of a `AwsVpcConfigurationProperty`
 *
 * @param properties - the TypeScript properties of a `AwsVpcConfigurationProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_AwsVpcConfigurationPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('assignPublicIp', cdk.validateString)(properties.assignPublicIp));
    errors.collect(cdk.propertyValidator('securityGroups', cdk.listValidator(cdk.validateString))(properties.securityGroups));
    errors.collect(cdk.propertyValidator('subnets', cdk.requiredValidator)(properties.subnets));
    errors.collect(cdk.propertyValidator('subnets', cdk.listValidator(cdk.validateString))(properties.subnets));
    return errors.wrap('supplied properties not correct for "AwsVpcConfigurationProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.AwsVpcConfiguration` resource
 *
 * @param properties - the TypeScript properties of a `AwsVpcConfigurationProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.AwsVpcConfiguration` resource.
 */
// @ts-ignore TS6133
function cfnPipeAwsVpcConfigurationPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_AwsVpcConfigurationPropertyValidator(properties).assertSuccess();
    return {
        AssignPublicIp: cdk.stringToCloudFormation(properties.assignPublicIp),
        SecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroups),
        Subnets: cdk.listMapper(cdk.stringToCloudFormation)(properties.subnets),
    };
}
// @ts-ignore TS6133
function CfnPipeAwsVpcConfigurationPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('assignPublicIp', 'AssignPublicIp', properties.AssignPublicIp != null ? cfn_parse.FromCloudFormation.getString(properties.AssignPublicIp) : undefined);
    ret.addPropertyResult('securityGroups', 'SecurityGroups', properties.SecurityGroups != null ? cfn_parse.FromCloudFormation.getStringArray(properties.SecurityGroups) : undefined);
    ret.addPropertyResult('subnets', 'Subnets', cfn_parse.FromCloudFormation.getStringArray(properties.Subnets));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchArrayPropertiesProperty`
 *
 * @param properties - the TypeScript properties of a `BatchArrayPropertiesProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchArrayPropertiesPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('size', cdk.validateNumber)(properties.size));
    return errors.wrap('supplied properties not correct for "BatchArrayPropertiesProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchArrayProperties` resource
 *
 * @param properties - the TypeScript properties of a `BatchArrayPropertiesProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchArrayProperties` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchArrayPropertiesPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchArrayPropertiesPropertyValidator(properties).assertSuccess();
    return {
        Size: cdk.numberToCloudFormation(properties.size),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchArrayPropertiesPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('size', 'Size', properties.Size != null ? cfn_parse.FromCloudFormation.getNumber(properties.Size) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchContainerOverridesProperty`
 *
 * @param properties - the TypeScript properties of a `BatchContainerOverridesProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchContainerOverridesPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('command', cdk.listValidator(cdk.validateString))(properties.command));
    errors.collect(cdk.propertyValidator('environment', cdk.listValidator(CfnPipe_BatchEnvironmentVariablePropertyValidator))(properties.environment));
    errors.collect(cdk.propertyValidator('instanceType', cdk.validateString)(properties.instanceType));
    errors.collect(cdk.propertyValidator('resourceRequirements', cdk.listValidator(CfnPipe_BatchResourceRequirementPropertyValidator))(properties.resourceRequirements));
    return errors.wrap('supplied properties not correct for "BatchContainerOverridesProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchContainerOverrides` resource
 *
 * @param properties - the TypeScript properties of a `BatchContainerOverridesProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchContainerOverrides` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchContainerOverridesPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchContainerOverridesPropertyValidator(properties).assertSuccess();
    return {
        Command: cdk.listMapper(cdk.stringToCloudFormation)(properties.command),
        Environment: cdk.listMapper(cfnPipeBatchEnvironmentVariablePropertyToCloudFormation)(properties.environment),
        InstanceType: cdk.stringToCloudFormation(properties.instanceType),
        ResourceRequirements: cdk.listMapper(cfnPipeBatchResourceRequirementPropertyToCloudFormation)(properties.resourceRequirements),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchContainerOverridesPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('command', 'Command', properties.Command != null ? cfn_parse.FromCloudFormation.getStringArray(properties.Command) : undefined);
    ret.addPropertyResult('environment', 'Environment', properties.Environment != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeBatchEnvironmentVariablePropertyFromCloudFormation)(properties.Environment) : undefined);
    ret.addPropertyResult('instanceType', 'InstanceType', properties.InstanceType != null ? cfn_parse.FromCloudFormation.getString(properties.InstanceType) : undefined);
    ret.addPropertyResult('resourceRequirements', 'ResourceRequirements', properties.ResourceRequirements != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeBatchResourceRequirementPropertyFromCloudFormation)(properties.ResourceRequirements) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchEnvironmentVariableProperty`
 *
 * @param properties - the TypeScript properties of a `BatchEnvironmentVariableProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchEnvironmentVariablePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "BatchEnvironmentVariableProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchEnvironmentVariable` resource
 *
 * @param properties - the TypeScript properties of a `BatchEnvironmentVariableProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchEnvironmentVariable` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchEnvironmentVariablePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchEnvironmentVariablePropertyValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchEnvironmentVariablePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('value', 'Value', properties.Value != null ? cfn_parse.FromCloudFormation.getString(properties.Value) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchJobDependencyProperty`
 *
 * @param properties - the TypeScript properties of a `BatchJobDependencyProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchJobDependencyPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('jobId', cdk.validateString)(properties.jobId));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "BatchJobDependencyProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchJobDependency` resource
 *
 * @param properties - the TypeScript properties of a `BatchJobDependencyProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchJobDependency` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchJobDependencyPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchJobDependencyPropertyValidator(properties).assertSuccess();
    return {
        JobId: cdk.stringToCloudFormation(properties.jobId),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchJobDependencyPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('jobId', 'JobId', properties.JobId != null ? cfn_parse.FromCloudFormation.getString(properties.JobId) : undefined);
    ret.addPropertyResult('type', 'Type', properties.Type != null ? cfn_parse.FromCloudFormation.getString(properties.Type) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchResourceRequirementProperty`
 *
 * @param properties - the TypeScript properties of a `BatchResourceRequirementProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchResourceRequirementPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "BatchResourceRequirementProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchResourceRequirement` resource
 *
 * @param properties - the TypeScript properties of a `BatchResourceRequirementProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchResourceRequirement` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchResourceRequirementPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchResourceRequirementPropertyValidator(properties).assertSuccess();
    return {
        Type: cdk.stringToCloudFormation(properties.type),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchResourceRequirementPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addPropertyResult('value', 'Value', cfn_parse.FromCloudFormation.getString(properties.Value));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `BatchRetryStrategyProperty`
 *
 * @param properties - the TypeScript properties of a `BatchRetryStrategyProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_BatchRetryStrategyPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('attempts', cdk.validateNumber)(properties.attempts));
    return errors.wrap('supplied properties not correct for "BatchRetryStrategyProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchRetryStrategy` resource
 *
 * @param properties - the TypeScript properties of a `BatchRetryStrategyProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.BatchRetryStrategy` resource.
 */
// @ts-ignore TS6133
function cfnPipeBatchRetryStrategyPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_BatchRetryStrategyPropertyValidator(properties).assertSuccess();
    return {
        Attempts: cdk.numberToCloudFormation(properties.attempts),
    };
}
// @ts-ignore TS6133
function CfnPipeBatchRetryStrategyPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('attempts', 'Attempts', properties.Attempts != null ? cfn_parse.FromCloudFormation.getNumber(properties.Attempts) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `CapacityProviderStrategyItemProperty`
 *
 * @param properties - the TypeScript properties of a `CapacityProviderStrategyItemProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_CapacityProviderStrategyItemPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('base', cdk.validateNumber)(properties.base));
    errors.collect(cdk.propertyValidator('capacityProvider', cdk.requiredValidator)(properties.capacityProvider));
    errors.collect(cdk.propertyValidator('capacityProvider', cdk.validateString)(properties.capacityProvider));
    errors.collect(cdk.propertyValidator('weight', cdk.validateNumber)(properties.weight));
    return errors.wrap('supplied properties not correct for "CapacityProviderStrategyItemProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.CapacityProviderStrategyItem` resource
 *
 * @param properties - the TypeScript properties of a `CapacityProviderStrategyItemProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.CapacityProviderStrategyItem` resource.
 */
// @ts-ignore TS6133
function cfnPipeCapacityProviderStrategyItemPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_CapacityProviderStrategyItemPropertyValidator(properties).assertSuccess();
    return {
        Base: cdk.numberToCloudFormation(properties.base),
        CapacityProvider: cdk.stringToCloudFormation(properties.capacityProvider),
        Weight: cdk.numberToCloudFormation(properties.weight),
    };
}
// @ts-ignore TS6133
function CfnPipeCapacityProviderStrategyItemPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('base', 'Base', properties.Base != null ? cfn_parse.FromCloudFormation.getNumber(properties.Base) : undefined);
    ret.addPropertyResult('capacityProvider', 'CapacityProvider', cfn_parse.FromCloudFormation.getString(properties.CapacityProvider));
    ret.addPropertyResult('weight', 'Weight', properties.Weight != null ? cfn_parse.FromCloudFormation.getNumber(properties.Weight) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `DeadLetterConfigProperty`
 *
 * @param properties - the TypeScript properties of a `DeadLetterConfigProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_DeadLetterConfigPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('arn', cdk.validateString)(properties.arn));
    return errors.wrap('supplied properties not correct for "DeadLetterConfigProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.DeadLetterConfig` resource
 *
 * @param properties - the TypeScript properties of a `DeadLetterConfigProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.DeadLetterConfig` resource.
 */
// @ts-ignore TS6133
function cfnPipeDeadLetterConfigPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_DeadLetterConfigPropertyValidator(properties).assertSuccess();
    return {
        Arn: cdk.stringToCloudFormation(properties.arn),
    };
}
// @ts-ignore TS6133
function CfnPipeDeadLetterConfigPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('arn', 'Arn', properties.Arn != null ? cfn_parse.FromCloudFormation.getString(properties.Arn) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsContainerOverrideProperty`
 *
 * @param properties - the TypeScript properties of a `EcsContainerOverrideProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsContainerOverridePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('command', cdk.listValidator(cdk.validateString))(properties.command));
    errors.collect(cdk.propertyValidator('cpu', cdk.validateNumber)(properties.cpu));
    errors.collect(cdk.propertyValidator('environment', cdk.listValidator(CfnPipe_EcsEnvironmentVariablePropertyValidator))(properties.environment));
    errors.collect(cdk.propertyValidator('environmentFiles', cdk.listValidator(CfnPipe_EcsEnvironmentFilePropertyValidator))(properties.environmentFiles));
    errors.collect(cdk.propertyValidator('memory', cdk.validateNumber)(properties.memory));
    errors.collect(cdk.propertyValidator('memoryReservation', cdk.validateNumber)(properties.memoryReservation));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('resourceRequirements', cdk.listValidator(CfnPipe_EcsResourceRequirementPropertyValidator))(properties.resourceRequirements));
    return errors.wrap('supplied properties not correct for "EcsContainerOverrideProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsContainerOverride` resource
 *
 * @param properties - the TypeScript properties of a `EcsContainerOverrideProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsContainerOverride` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsContainerOverridePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsContainerOverridePropertyValidator(properties).assertSuccess();
    return {
        Command: cdk.listMapper(cdk.stringToCloudFormation)(properties.command),
        Cpu: cdk.numberToCloudFormation(properties.cpu),
        Environment: cdk.listMapper(cfnPipeEcsEnvironmentVariablePropertyToCloudFormation)(properties.environment),
        EnvironmentFiles: cdk.listMapper(cfnPipeEcsEnvironmentFilePropertyToCloudFormation)(properties.environmentFiles),
        Memory: cdk.numberToCloudFormation(properties.memory),
        MemoryReservation: cdk.numberToCloudFormation(properties.memoryReservation),
        Name: cdk.stringToCloudFormation(properties.name),
        ResourceRequirements: cdk.listMapper(cfnPipeEcsResourceRequirementPropertyToCloudFormation)(properties.resourceRequirements),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsContainerOverridePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('command', 'Command', properties.Command != null ? cfn_parse.FromCloudFormation.getStringArray(properties.Command) : undefined);
    ret.addPropertyResult('cpu', 'Cpu', properties.Cpu != null ? cfn_parse.FromCloudFormation.getNumber(properties.Cpu) : undefined);
    ret.addPropertyResult('environment', 'Environment', properties.Environment != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeEcsEnvironmentVariablePropertyFromCloudFormation)(properties.Environment) : undefined);
    ret.addPropertyResult('environmentFiles', 'EnvironmentFiles', properties.EnvironmentFiles != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeEcsEnvironmentFilePropertyFromCloudFormation)(properties.EnvironmentFiles) : undefined);
    ret.addPropertyResult('memory', 'Memory', properties.Memory != null ? cfn_parse.FromCloudFormation.getNumber(properties.Memory) : undefined);
    ret.addPropertyResult('memoryReservation', 'MemoryReservation', properties.MemoryReservation != null ? cfn_parse.FromCloudFormation.getNumber(properties.MemoryReservation) : undefined);
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('resourceRequirements', 'ResourceRequirements', properties.ResourceRequirements != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeEcsResourceRequirementPropertyFromCloudFormation)(properties.ResourceRequirements) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsEnvironmentFileProperty`
 *
 * @param properties - the TypeScript properties of a `EcsEnvironmentFileProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsEnvironmentFilePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "EcsEnvironmentFileProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEnvironmentFile` resource
 *
 * @param properties - the TypeScript properties of a `EcsEnvironmentFileProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEnvironmentFile` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsEnvironmentFilePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsEnvironmentFilePropertyValidator(properties).assertSuccess();
    return {
        Type: cdk.stringToCloudFormation(properties.type),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsEnvironmentFilePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addPropertyResult('value', 'Value', cfn_parse.FromCloudFormation.getString(properties.Value));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsEnvironmentVariableProperty`
 *
 * @param properties - the TypeScript properties of a `EcsEnvironmentVariableProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsEnvironmentVariablePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "EcsEnvironmentVariableProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEnvironmentVariable` resource
 *
 * @param properties - the TypeScript properties of a `EcsEnvironmentVariableProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEnvironmentVariable` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsEnvironmentVariablePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsEnvironmentVariablePropertyValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsEnvironmentVariablePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('value', 'Value', properties.Value != null ? cfn_parse.FromCloudFormation.getString(properties.Value) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsEphemeralStorageProperty`
 *
 * @param properties - the TypeScript properties of a `EcsEphemeralStorageProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsEphemeralStoragePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('sizeInGiB', cdk.requiredValidator)(properties.sizeInGiB));
    errors.collect(cdk.propertyValidator('sizeInGiB', cdk.validateNumber)(properties.sizeInGiB));
    return errors.wrap('supplied properties not correct for "EcsEphemeralStorageProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEphemeralStorage` resource
 *
 * @param properties - the TypeScript properties of a `EcsEphemeralStorageProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsEphemeralStorage` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsEphemeralStoragePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsEphemeralStoragePropertyValidator(properties).assertSuccess();
    return {
        SizeInGiB: cdk.numberToCloudFormation(properties.sizeInGiB),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsEphemeralStoragePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('sizeInGiB', 'SizeInGiB', cfn_parse.FromCloudFormation.getNumber(properties.SizeInGiB));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsInferenceAcceleratorOverrideProperty`
 *
 * @param properties - the TypeScript properties of a `EcsInferenceAcceleratorOverrideProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsInferenceAcceleratorOverridePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('deviceName', cdk.validateString)(properties.deviceName));
    errors.collect(cdk.propertyValidator('deviceType', cdk.validateString)(properties.deviceType));
    return errors.wrap('supplied properties not correct for "EcsInferenceAcceleratorOverrideProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsInferenceAcceleratorOverride` resource
 *
 * @param properties - the TypeScript properties of a `EcsInferenceAcceleratorOverrideProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsInferenceAcceleratorOverride` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsInferenceAcceleratorOverridePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsInferenceAcceleratorOverridePropertyValidator(properties).assertSuccess();
    return {
        DeviceName: cdk.stringToCloudFormation(properties.deviceName),
        DeviceType: cdk.stringToCloudFormation(properties.deviceType),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsInferenceAcceleratorOverridePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('deviceName', 'DeviceName', properties.DeviceName != null ? cfn_parse.FromCloudFormation.getString(properties.DeviceName) : undefined);
    ret.addPropertyResult('deviceType', 'DeviceType', properties.DeviceType != null ? cfn_parse.FromCloudFormation.getString(properties.DeviceType) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsResourceRequirementProperty`
 *
 * @param properties - the TypeScript properties of a `EcsResourceRequirementProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsResourceRequirementPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "EcsResourceRequirementProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsResourceRequirement` resource
 *
 * @param properties - the TypeScript properties of a `EcsResourceRequirementProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsResourceRequirement` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsResourceRequirementPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsResourceRequirementPropertyValidator(properties).assertSuccess();
    return {
        Type: cdk.stringToCloudFormation(properties.type),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsResourceRequirementPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addPropertyResult('value', 'Value', cfn_parse.FromCloudFormation.getString(properties.Value));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `EcsTaskOverrideProperty`
 *
 * @param properties - the TypeScript properties of a `EcsTaskOverrideProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_EcsTaskOverridePropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('containerOverrides', cdk.listValidator(CfnPipe_EcsContainerOverridePropertyValidator))(properties.containerOverrides));
    errors.collect(cdk.propertyValidator('cpu', cdk.validateString)(properties.cpu));
    errors.collect(cdk.propertyValidator('ephemeralStorage', CfnPipe_EcsEphemeralStoragePropertyValidator)(properties.ephemeralStorage));
    errors.collect(cdk.propertyValidator('executionRoleArn', cdk.validateString)(properties.executionRoleArn));
    errors.collect(cdk.propertyValidator('inferenceAcceleratorOverrides', cdk.listValidator(CfnPipe_EcsInferenceAcceleratorOverridePropertyValidator))(properties.inferenceAcceleratorOverrides));
    errors.collect(cdk.propertyValidator('memory', cdk.validateString)(properties.memory));
    errors.collect(cdk.propertyValidator('taskRoleArn', cdk.validateString)(properties.taskRoleArn));
    return errors.wrap('supplied properties not correct for "EcsTaskOverrideProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsTaskOverride` resource
 *
 * @param properties - the TypeScript properties of a `EcsTaskOverrideProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.EcsTaskOverride` resource.
 */
// @ts-ignore TS6133
function cfnPipeEcsTaskOverridePropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_EcsTaskOverridePropertyValidator(properties).assertSuccess();
    return {
        ContainerOverrides: cdk.listMapper(cfnPipeEcsContainerOverridePropertyToCloudFormation)(properties.containerOverrides),
        Cpu: cdk.stringToCloudFormation(properties.cpu),
        EphemeralStorage: cfnPipeEcsEphemeralStoragePropertyToCloudFormation(properties.ephemeralStorage),
        ExecutionRoleArn: cdk.stringToCloudFormation(properties.executionRoleArn),
        InferenceAcceleratorOverrides: cdk.listMapper(cfnPipeEcsInferenceAcceleratorOverridePropertyToCloudFormation)(properties.inferenceAcceleratorOverrides),
        Memory: cdk.stringToCloudFormation(properties.memory),
        TaskRoleArn: cdk.stringToCloudFormation(properties.taskRoleArn),
    };
}
// @ts-ignore TS6133
function CfnPipeEcsTaskOverridePropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('containerOverrides', 'ContainerOverrides', properties.ContainerOverrides != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeEcsContainerOverridePropertyFromCloudFormation)(properties.ContainerOverrides) : undefined);
    ret.addPropertyResult('cpu', 'Cpu', properties.Cpu != null ? cfn_parse.FromCloudFormation.getString(properties.Cpu) : undefined);
    ret.addPropertyResult('ephemeralStorage', 'EphemeralStorage', properties.EphemeralStorage != null ? CfnPipeEcsEphemeralStoragePropertyFromCloudFormation(properties.EphemeralStorage) : undefined);
    ret.addPropertyResult('executionRoleArn', 'ExecutionRoleArn', properties.ExecutionRoleArn != null ? cfn_parse.FromCloudFormation.getString(properties.ExecutionRoleArn) : undefined);
    ret.addPropertyResult('inferenceAcceleratorOverrides', 'InferenceAcceleratorOverrides', properties.InferenceAcceleratorOverrides != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeEcsInferenceAcceleratorOverridePropertyFromCloudFormation)(properties.InferenceAcceleratorOverrides) : undefined);
    ret.addPropertyResult('memory', 'Memory', properties.Memory != null ? cfn_parse.FromCloudFormation.getString(properties.Memory) : undefined);
    ret.addPropertyResult('taskRoleArn', 'TaskRoleArn', properties.TaskRoleArn != null ? cfn_parse.FromCloudFormation.getString(properties.TaskRoleArn) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `FilterProperty`
 *
 * @param properties - the TypeScript properties of a `FilterProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_FilterPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('pattern', cdk.validateString)(properties.pattern));
    return errors.wrap('supplied properties not correct for "FilterProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.Filter` resource
 *
 * @param properties - the TypeScript properties of a `FilterProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.Filter` resource.
 */
// @ts-ignore TS6133
function cfnPipeFilterPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_FilterPropertyValidator(properties).assertSuccess();
    return {
        Pattern: cdk.stringToCloudFormation(properties.pattern),
    };
}
// @ts-ignore TS6133
function CfnPipeFilterPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('pattern', 'Pattern', properties.Pattern != null ? cfn_parse.FromCloudFormation.getString(properties.Pattern) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `FilterCriteriaProperty`
 *
 * @param properties - the TypeScript properties of a `FilterCriteriaProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_FilterCriteriaPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('filters', cdk.listValidator(CfnPipe_FilterPropertyValidator))(properties.filters));
    return errors.wrap('supplied properties not correct for "FilterCriteriaProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.FilterCriteria` resource
 *
 * @param properties - the TypeScript properties of a `FilterCriteriaProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.FilterCriteria` resource.
 */
// @ts-ignore TS6133
function cfnPipeFilterCriteriaPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_FilterCriteriaPropertyValidator(properties).assertSuccess();
    return {
        Filters: cdk.listMapper(cfnPipeFilterPropertyToCloudFormation)(properties.filters),
    };
}
// @ts-ignore TS6133
function CfnPipeFilterCriteriaPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('filters', 'Filters', properties.Filters != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeFilterPropertyFromCloudFormation)(properties.Filters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `MQBrokerAccessCredentialsProperty`
 *
 * @param properties - the TypeScript properties of a `MQBrokerAccessCredentialsProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_MQBrokerAccessCredentialsPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('basicAuth', cdk.requiredValidator)(properties.basicAuth));
    errors.collect(cdk.propertyValidator('basicAuth', cdk.validateString)(properties.basicAuth));
    return errors.wrap('supplied properties not correct for "MQBrokerAccessCredentialsProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.MQBrokerAccessCredentials` resource
 *
 * @param properties - the TypeScript properties of a `MQBrokerAccessCredentialsProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.MQBrokerAccessCredentials` resource.
 */
// @ts-ignore TS6133
function cfnPipeMQBrokerAccessCredentialsPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_MQBrokerAccessCredentialsPropertyValidator(properties).assertSuccess();
    return {
        BasicAuth: cdk.stringToCloudFormation(properties.basicAuth),
    };
}
// @ts-ignore TS6133
function CfnPipeMQBrokerAccessCredentialsPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('basicAuth', 'BasicAuth', cfn_parse.FromCloudFormation.getString(properties.BasicAuth));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `MSKAccessCredentialsProperty`
 *
 * @param properties - the TypeScript properties of a `MSKAccessCredentialsProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_MSKAccessCredentialsPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('clientCertificateTlsAuth', cdk.validateString)(properties.clientCertificateTlsAuth));
    errors.collect(cdk.propertyValidator('saslScram512Auth', cdk.validateString)(properties.saslScram512Auth));
    return errors.wrap('supplied properties not correct for "MSKAccessCredentialsProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.MSKAccessCredentials` resource
 *
 * @param properties - the TypeScript properties of a `MSKAccessCredentialsProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.MSKAccessCredentials` resource.
 */
// @ts-ignore TS6133
function cfnPipeMSKAccessCredentialsPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_MSKAccessCredentialsPropertyValidator(properties).assertSuccess();
    return {
        ClientCertificateTlsAuth: cdk.stringToCloudFormation(properties.clientCertificateTlsAuth),
        SaslScram512Auth: cdk.stringToCloudFormation(properties.saslScram512Auth),
    };
}
// @ts-ignore TS6133
function CfnPipeMSKAccessCredentialsPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('clientCertificateTlsAuth', 'ClientCertificateTlsAuth', properties.ClientCertificateTlsAuth != null ? cfn_parse.FromCloudFormation.getString(properties.ClientCertificateTlsAuth) : undefined);
    ret.addPropertyResult('saslScram512Auth', 'SaslScram512Auth', properties.SaslScram512Auth != null ? cfn_parse.FromCloudFormation.getString(properties.SaslScram512Auth) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `NetworkConfigurationProperty`
 *
 * @param properties - the TypeScript properties of a `NetworkConfigurationProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_NetworkConfigurationPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('awsvpcConfiguration', CfnPipe_AwsVpcConfigurationPropertyValidator)(properties.awsvpcConfiguration));
    return errors.wrap('supplied properties not correct for "NetworkConfigurationProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.NetworkConfiguration` resource
 *
 * @param properties - the TypeScript properties of a `NetworkConfigurationProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.NetworkConfiguration` resource.
 */
// @ts-ignore TS6133
function cfnPipeNetworkConfigurationPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_NetworkConfigurationPropertyValidator(properties).assertSuccess();
    return {
        AwsvpcConfiguration: cfnPipeAwsVpcConfigurationPropertyToCloudFormation(properties.awsvpcConfiguration),
    };
}
// @ts-ignore TS6133
function CfnPipeNetworkConfigurationPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('awsvpcConfiguration', 'AwsvpcConfiguration', properties.AwsvpcConfiguration != null ? CfnPipeAwsVpcConfigurationPropertyFromCloudFormation(properties.AwsvpcConfiguration) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeEnrichmentHttpParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeEnrichmentHttpParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeEnrichmentHttpParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('headerParameters', cdk.hashValidator(cdk.validateString))(properties.headerParameters));
    errors.collect(cdk.propertyValidator('pathParameterValues', cdk.listValidator(cdk.validateString))(properties.pathParameterValues));
    errors.collect(cdk.propertyValidator('queryStringParameters', cdk.hashValidator(cdk.validateString))(properties.queryStringParameters));
    return errors.wrap('supplied properties not correct for "PipeEnrichmentHttpParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeEnrichmentHttpParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeEnrichmentHttpParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeEnrichmentHttpParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeEnrichmentHttpParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeEnrichmentHttpParametersPropertyValidator(properties).assertSuccess();
    return {
        HeaderParameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.headerParameters),
        PathParameterValues: cdk.listMapper(cdk.stringToCloudFormation)(properties.pathParameterValues),
        QueryStringParameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.queryStringParameters),
    };
}
// @ts-ignore TS6133
function CfnPipePipeEnrichmentHttpParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('headerParameters', 'HeaderParameters', properties.HeaderParameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.HeaderParameters) : undefined);
    ret.addPropertyResult('pathParameterValues', 'PathParameterValues', properties.PathParameterValues != null ? cfn_parse.FromCloudFormation.getStringArray(properties.PathParameterValues) : undefined);
    ret.addPropertyResult('queryStringParameters', 'QueryStringParameters', properties.QueryStringParameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.QueryStringParameters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeEnrichmentParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeEnrichmentParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeEnrichmentParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('httpParameters', CfnPipe_PipeEnrichmentHttpParametersPropertyValidator)(properties.httpParameters));
    errors.collect(cdk.propertyValidator('inputTemplate', cdk.validateString)(properties.inputTemplate));
    return errors.wrap('supplied properties not correct for "PipeEnrichmentParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeEnrichmentParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeEnrichmentParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeEnrichmentParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeEnrichmentParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeEnrichmentParametersPropertyValidator(properties).assertSuccess();
    return {
        HttpParameters: cfnPipePipeEnrichmentHttpParametersPropertyToCloudFormation(properties.httpParameters),
        InputTemplate: cdk.stringToCloudFormation(properties.inputTemplate),
    };
}
// @ts-ignore TS6133
function CfnPipePipeEnrichmentParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('httpParameters', 'HttpParameters', properties.HttpParameters != null ? CfnPipePipeEnrichmentHttpParametersPropertyFromCloudFormation(properties.HttpParameters) : undefined);
    ret.addPropertyResult('inputTemplate', 'InputTemplate', properties.InputTemplate != null ? cfn_parse.FromCloudFormation.getString(properties.InputTemplate) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceActiveMQBrokerParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceActiveMQBrokerParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceActiveMQBrokerParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('credentials', cdk.requiredValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('credentials', CfnPipe_MQBrokerAccessCredentialsPropertyValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('queueName', cdk.requiredValidator)(properties.queueName));
    errors.collect(cdk.propertyValidator('queueName', cdk.validateString)(properties.queueName));
    return errors.wrap('supplied properties not correct for "PipeSourceActiveMQBrokerParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceActiveMQBrokerParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceActiveMQBrokerParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceActiveMQBrokerParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceActiveMQBrokerParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceActiveMQBrokerParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        Credentials: cfnPipeMQBrokerAccessCredentialsPropertyToCloudFormation(properties.credentials),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        QueueName: cdk.stringToCloudFormation(properties.queueName),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceActiveMQBrokerParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('credentials', 'Credentials', CfnPipeMQBrokerAccessCredentialsPropertyFromCloudFormation(properties.Credentials));
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('queueName', 'QueueName', cfn_parse.FromCloudFormation.getString(properties.QueueName));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceDynamoDBStreamParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceDynamoDBStreamParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceDynamoDBStreamParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('deadLetterConfig', CfnPipe_DeadLetterConfigPropertyValidator)(properties.deadLetterConfig));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('maximumRecordAgeInSeconds', cdk.validateNumber)(properties.maximumRecordAgeInSeconds));
    errors.collect(cdk.propertyValidator('maximumRetryAttempts', cdk.validateNumber)(properties.maximumRetryAttempts));
    errors.collect(cdk.propertyValidator('onPartialBatchItemFailure', cdk.validateString)(properties.onPartialBatchItemFailure));
    errors.collect(cdk.propertyValidator('parallelizationFactor', cdk.validateNumber)(properties.parallelizationFactor));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.requiredValidator)(properties.startingPosition));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.validateString)(properties.startingPosition));
    return errors.wrap('supplied properties not correct for "PipeSourceDynamoDBStreamParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceDynamoDBStreamParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceDynamoDBStreamParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceDynamoDBStreamParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceDynamoDBStreamParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceDynamoDBStreamParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        DeadLetterConfig: cfnPipeDeadLetterConfigPropertyToCloudFormation(properties.deadLetterConfig),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        MaximumRecordAgeInSeconds: cdk.numberToCloudFormation(properties.maximumRecordAgeInSeconds),
        MaximumRetryAttempts: cdk.numberToCloudFormation(properties.maximumRetryAttempts),
        OnPartialBatchItemFailure: cdk.stringToCloudFormation(properties.onPartialBatchItemFailure),
        ParallelizationFactor: cdk.numberToCloudFormation(properties.parallelizationFactor),
        StartingPosition: cdk.stringToCloudFormation(properties.startingPosition),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceDynamoDBStreamParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('deadLetterConfig', 'DeadLetterConfig', properties.DeadLetterConfig != null ? CfnPipeDeadLetterConfigPropertyFromCloudFormation(properties.DeadLetterConfig) : undefined);
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('maximumRecordAgeInSeconds', 'MaximumRecordAgeInSeconds', properties.MaximumRecordAgeInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumRecordAgeInSeconds) : undefined);
    ret.addPropertyResult('maximumRetryAttempts', 'MaximumRetryAttempts', properties.MaximumRetryAttempts != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumRetryAttempts) : undefined);
    ret.addPropertyResult('onPartialBatchItemFailure', 'OnPartialBatchItemFailure', properties.OnPartialBatchItemFailure != null ? cfn_parse.FromCloudFormation.getString(properties.OnPartialBatchItemFailure) : undefined);
    ret.addPropertyResult('parallelizationFactor', 'ParallelizationFactor', properties.ParallelizationFactor != null ? cfn_parse.FromCloudFormation.getNumber(properties.ParallelizationFactor) : undefined);
    ret.addPropertyResult('startingPosition', 'StartingPosition', cfn_parse.FromCloudFormation.getString(properties.StartingPosition));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceKinesisStreamParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceKinesisStreamParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceKinesisStreamParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('deadLetterConfig', CfnPipe_DeadLetterConfigPropertyValidator)(properties.deadLetterConfig));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('maximumRecordAgeInSeconds', cdk.validateNumber)(properties.maximumRecordAgeInSeconds));
    errors.collect(cdk.propertyValidator('maximumRetryAttempts', cdk.validateNumber)(properties.maximumRetryAttempts));
    errors.collect(cdk.propertyValidator('onPartialBatchItemFailure', cdk.validateString)(properties.onPartialBatchItemFailure));
    errors.collect(cdk.propertyValidator('parallelizationFactor', cdk.validateNumber)(properties.parallelizationFactor));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.requiredValidator)(properties.startingPosition));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.validateString)(properties.startingPosition));
    errors.collect(cdk.propertyValidator('startingPositionTimestamp', cdk.validateString)(properties.startingPositionTimestamp));
    return errors.wrap('supplied properties not correct for "PipeSourceKinesisStreamParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceKinesisStreamParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceKinesisStreamParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceKinesisStreamParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceKinesisStreamParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceKinesisStreamParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        DeadLetterConfig: cfnPipeDeadLetterConfigPropertyToCloudFormation(properties.deadLetterConfig),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        MaximumRecordAgeInSeconds: cdk.numberToCloudFormation(properties.maximumRecordAgeInSeconds),
        MaximumRetryAttempts: cdk.numberToCloudFormation(properties.maximumRetryAttempts),
        OnPartialBatchItemFailure: cdk.stringToCloudFormation(properties.onPartialBatchItemFailure),
        ParallelizationFactor: cdk.numberToCloudFormation(properties.parallelizationFactor),
        StartingPosition: cdk.stringToCloudFormation(properties.startingPosition),
        StartingPositionTimestamp: cdk.stringToCloudFormation(properties.startingPositionTimestamp),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceKinesisStreamParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('deadLetterConfig', 'DeadLetterConfig', properties.DeadLetterConfig != null ? CfnPipeDeadLetterConfigPropertyFromCloudFormation(properties.DeadLetterConfig) : undefined);
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('maximumRecordAgeInSeconds', 'MaximumRecordAgeInSeconds', properties.MaximumRecordAgeInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumRecordAgeInSeconds) : undefined);
    ret.addPropertyResult('maximumRetryAttempts', 'MaximumRetryAttempts', properties.MaximumRetryAttempts != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumRetryAttempts) : undefined);
    ret.addPropertyResult('onPartialBatchItemFailure', 'OnPartialBatchItemFailure', properties.OnPartialBatchItemFailure != null ? cfn_parse.FromCloudFormation.getString(properties.OnPartialBatchItemFailure) : undefined);
    ret.addPropertyResult('parallelizationFactor', 'ParallelizationFactor', properties.ParallelizationFactor != null ? cfn_parse.FromCloudFormation.getNumber(properties.ParallelizationFactor) : undefined);
    ret.addPropertyResult('startingPosition', 'StartingPosition', cfn_parse.FromCloudFormation.getString(properties.StartingPosition));
    ret.addPropertyResult('startingPositionTimestamp', 'StartingPositionTimestamp', properties.StartingPositionTimestamp != null ? cfn_parse.FromCloudFormation.getString(properties.StartingPositionTimestamp) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceManagedStreamingKafkaParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceManagedStreamingKafkaParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceManagedStreamingKafkaParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('consumerGroupId', cdk.validateString)(properties.consumerGroupId));
    errors.collect(cdk.propertyValidator('credentials', CfnPipe_MSKAccessCredentialsPropertyValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.validateString)(properties.startingPosition));
    errors.collect(cdk.propertyValidator('topicName', cdk.requiredValidator)(properties.topicName));
    errors.collect(cdk.propertyValidator('topicName', cdk.validateString)(properties.topicName));
    return errors.wrap('supplied properties not correct for "PipeSourceManagedStreamingKafkaParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceManagedStreamingKafkaParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceManagedStreamingKafkaParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceManagedStreamingKafkaParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceManagedStreamingKafkaParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceManagedStreamingKafkaParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        ConsumerGroupID: cdk.stringToCloudFormation(properties.consumerGroupId),
        Credentials: cfnPipeMSKAccessCredentialsPropertyToCloudFormation(properties.credentials),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        StartingPosition: cdk.stringToCloudFormation(properties.startingPosition),
        TopicName: cdk.stringToCloudFormation(properties.topicName),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceManagedStreamingKafkaParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('consumerGroupId', 'ConsumerGroupID', properties.ConsumerGroupID != null ? cfn_parse.FromCloudFormation.getString(properties.ConsumerGroupID) : undefined);
    ret.addPropertyResult('credentials', 'Credentials', properties.Credentials != null ? CfnPipeMSKAccessCredentialsPropertyFromCloudFormation(properties.Credentials) : undefined);
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('startingPosition', 'StartingPosition', properties.StartingPosition != null ? cfn_parse.FromCloudFormation.getString(properties.StartingPosition) : undefined);
    ret.addPropertyResult('topicName', 'TopicName', cfn_parse.FromCloudFormation.getString(properties.TopicName));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('activeMqBrokerParameters', CfnPipe_PipeSourceActiveMQBrokerParametersPropertyValidator)(properties.activeMqBrokerParameters));
    errors.collect(cdk.propertyValidator('dynamoDbStreamParameters', CfnPipe_PipeSourceDynamoDBStreamParametersPropertyValidator)(properties.dynamoDbStreamParameters));
    errors.collect(cdk.propertyValidator('filterCriteria', CfnPipe_FilterCriteriaPropertyValidator)(properties.filterCriteria));
    errors.collect(cdk.propertyValidator('kinesisStreamParameters', CfnPipe_PipeSourceKinesisStreamParametersPropertyValidator)(properties.kinesisStreamParameters));
    errors.collect(cdk.propertyValidator('managedStreamingKafkaParameters', CfnPipe_PipeSourceManagedStreamingKafkaParametersPropertyValidator)(properties.managedStreamingKafkaParameters));
    errors.collect(cdk.propertyValidator('rabbitMqBrokerParameters', CfnPipe_PipeSourceRabbitMQBrokerParametersPropertyValidator)(properties.rabbitMqBrokerParameters));
    errors.collect(cdk.propertyValidator('selfManagedKafkaParameters', CfnPipe_PipeSourceSelfManagedKafkaParametersPropertyValidator)(properties.selfManagedKafkaParameters));
    errors.collect(cdk.propertyValidator('sqsQueueParameters', CfnPipe_PipeSourceSqsQueueParametersPropertyValidator)(properties.sqsQueueParameters));
    return errors.wrap('supplied properties not correct for "PipeSourceParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceParametersPropertyValidator(properties).assertSuccess();
    return {
        ActiveMQBrokerParameters: cfnPipePipeSourceActiveMQBrokerParametersPropertyToCloudFormation(properties.activeMqBrokerParameters),
        DynamoDBStreamParameters: cfnPipePipeSourceDynamoDBStreamParametersPropertyToCloudFormation(properties.dynamoDbStreamParameters),
        FilterCriteria: cfnPipeFilterCriteriaPropertyToCloudFormation(properties.filterCriteria),
        KinesisStreamParameters: cfnPipePipeSourceKinesisStreamParametersPropertyToCloudFormation(properties.kinesisStreamParameters),
        ManagedStreamingKafkaParameters: cfnPipePipeSourceManagedStreamingKafkaParametersPropertyToCloudFormation(properties.managedStreamingKafkaParameters),
        RabbitMQBrokerParameters: cfnPipePipeSourceRabbitMQBrokerParametersPropertyToCloudFormation(properties.rabbitMqBrokerParameters),
        SelfManagedKafkaParameters: cfnPipePipeSourceSelfManagedKafkaParametersPropertyToCloudFormation(properties.selfManagedKafkaParameters),
        SqsQueueParameters: cfnPipePipeSourceSqsQueueParametersPropertyToCloudFormation(properties.sqsQueueParameters),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('activeMqBrokerParameters', 'ActiveMQBrokerParameters', properties.ActiveMQBrokerParameters != null ? CfnPipePipeSourceActiveMQBrokerParametersPropertyFromCloudFormation(properties.ActiveMQBrokerParameters) : undefined);
    ret.addPropertyResult('dynamoDbStreamParameters', 'DynamoDBStreamParameters', properties.DynamoDBStreamParameters != null ? CfnPipePipeSourceDynamoDBStreamParametersPropertyFromCloudFormation(properties.DynamoDBStreamParameters) : undefined);
    ret.addPropertyResult('filterCriteria', 'FilterCriteria', properties.FilterCriteria != null ? CfnPipeFilterCriteriaPropertyFromCloudFormation(properties.FilterCriteria) : undefined);
    ret.addPropertyResult('kinesisStreamParameters', 'KinesisStreamParameters', properties.KinesisStreamParameters != null ? CfnPipePipeSourceKinesisStreamParametersPropertyFromCloudFormation(properties.KinesisStreamParameters) : undefined);
    ret.addPropertyResult('managedStreamingKafkaParameters', 'ManagedStreamingKafkaParameters', properties.ManagedStreamingKafkaParameters != null ? CfnPipePipeSourceManagedStreamingKafkaParametersPropertyFromCloudFormation(properties.ManagedStreamingKafkaParameters) : undefined);
    ret.addPropertyResult('rabbitMqBrokerParameters', 'RabbitMQBrokerParameters', properties.RabbitMQBrokerParameters != null ? CfnPipePipeSourceRabbitMQBrokerParametersPropertyFromCloudFormation(properties.RabbitMQBrokerParameters) : undefined);
    ret.addPropertyResult('selfManagedKafkaParameters', 'SelfManagedKafkaParameters', properties.SelfManagedKafkaParameters != null ? CfnPipePipeSourceSelfManagedKafkaParametersPropertyFromCloudFormation(properties.SelfManagedKafkaParameters) : undefined);
    ret.addPropertyResult('sqsQueueParameters', 'SqsQueueParameters', properties.SqsQueueParameters != null ? CfnPipePipeSourceSqsQueueParametersPropertyFromCloudFormation(properties.SqsQueueParameters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceRabbitMQBrokerParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceRabbitMQBrokerParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceRabbitMQBrokerParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('credentials', cdk.requiredValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('credentials', CfnPipe_MQBrokerAccessCredentialsPropertyValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('queueName', cdk.requiredValidator)(properties.queueName));
    errors.collect(cdk.propertyValidator('queueName', cdk.validateString)(properties.queueName));
    errors.collect(cdk.propertyValidator('virtualHost', cdk.validateString)(properties.virtualHost));
    return errors.wrap('supplied properties not correct for "PipeSourceRabbitMQBrokerParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceRabbitMQBrokerParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceRabbitMQBrokerParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceRabbitMQBrokerParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceRabbitMQBrokerParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceRabbitMQBrokerParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        Credentials: cfnPipeMQBrokerAccessCredentialsPropertyToCloudFormation(properties.credentials),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        QueueName: cdk.stringToCloudFormation(properties.queueName),
        VirtualHost: cdk.stringToCloudFormation(properties.virtualHost),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceRabbitMQBrokerParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('credentials', 'Credentials', CfnPipeMQBrokerAccessCredentialsPropertyFromCloudFormation(properties.Credentials));
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('queueName', 'QueueName', cfn_parse.FromCloudFormation.getString(properties.QueueName));
    ret.addPropertyResult('virtualHost', 'VirtualHost', properties.VirtualHost != null ? cfn_parse.FromCloudFormation.getString(properties.VirtualHost) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceSelfManagedKafkaParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceSelfManagedKafkaParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceSelfManagedKafkaParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('additionalBootstrapServers', cdk.listValidator(cdk.validateString))(properties.additionalBootstrapServers));
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('consumerGroupId', cdk.validateString)(properties.consumerGroupId));
    errors.collect(cdk.propertyValidator('credentials', CfnPipe_SelfManagedKafkaAccessConfigurationCredentialsPropertyValidator)(properties.credentials));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    errors.collect(cdk.propertyValidator('serverRootCaCertificate', cdk.validateString)(properties.serverRootCaCertificate));
    errors.collect(cdk.propertyValidator('startingPosition', cdk.validateString)(properties.startingPosition));
    errors.collect(cdk.propertyValidator('topicName', cdk.requiredValidator)(properties.topicName));
    errors.collect(cdk.propertyValidator('topicName', cdk.validateString)(properties.topicName));
    errors.collect(cdk.propertyValidator('vpc', CfnPipe_SelfManagedKafkaAccessConfigurationVpcPropertyValidator)(properties.vpc));
    return errors.wrap('supplied properties not correct for "PipeSourceSelfManagedKafkaParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceSelfManagedKafkaParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceSelfManagedKafkaParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceSelfManagedKafkaParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceSelfManagedKafkaParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceSelfManagedKafkaParametersPropertyValidator(properties).assertSuccess();
    return {
        AdditionalBootstrapServers: cdk.listMapper(cdk.stringToCloudFormation)(properties.additionalBootstrapServers),
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        ConsumerGroupID: cdk.stringToCloudFormation(properties.consumerGroupId),
        Credentials: cfnPipeSelfManagedKafkaAccessConfigurationCredentialsPropertyToCloudFormation(properties.credentials),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
        ServerRootCaCertificate: cdk.stringToCloudFormation(properties.serverRootCaCertificate),
        StartingPosition: cdk.stringToCloudFormation(properties.startingPosition),
        TopicName: cdk.stringToCloudFormation(properties.topicName),
        Vpc: cfnPipeSelfManagedKafkaAccessConfigurationVpcPropertyToCloudFormation(properties.vpc),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceSelfManagedKafkaParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('additionalBootstrapServers', 'AdditionalBootstrapServers', properties.AdditionalBootstrapServers != null ? cfn_parse.FromCloudFormation.getStringArray(properties.AdditionalBootstrapServers) : undefined);
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('consumerGroupId', 'ConsumerGroupID', properties.ConsumerGroupID != null ? cfn_parse.FromCloudFormation.getString(properties.ConsumerGroupID) : undefined);
    ret.addPropertyResult('credentials', 'Credentials', properties.Credentials != null ? CfnPipeSelfManagedKafkaAccessConfigurationCredentialsPropertyFromCloudFormation(properties.Credentials) : undefined);
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addPropertyResult('serverRootCaCertificate', 'ServerRootCaCertificate', properties.ServerRootCaCertificate != null ? cfn_parse.FromCloudFormation.getString(properties.ServerRootCaCertificate) : undefined);
    ret.addPropertyResult('startingPosition', 'StartingPosition', properties.StartingPosition != null ? cfn_parse.FromCloudFormation.getString(properties.StartingPosition) : undefined);
    ret.addPropertyResult('topicName', 'TopicName', cfn_parse.FromCloudFormation.getString(properties.TopicName));
    ret.addPropertyResult('vpc', 'Vpc', properties.Vpc != null ? CfnPipeSelfManagedKafkaAccessConfigurationVpcPropertyFromCloudFormation(properties.Vpc) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeSourceSqsQueueParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeSourceSqsQueueParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeSourceSqsQueueParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchSize', cdk.validateNumber)(properties.batchSize));
    errors.collect(cdk.propertyValidator('maximumBatchingWindowInSeconds', cdk.validateNumber)(properties.maximumBatchingWindowInSeconds));
    return errors.wrap('supplied properties not correct for "PipeSourceSqsQueueParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceSqsQueueParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeSourceSqsQueueParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeSourceSqsQueueParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeSourceSqsQueueParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeSourceSqsQueueParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchSize: cdk.numberToCloudFormation(properties.batchSize),
        MaximumBatchingWindowInSeconds: cdk.numberToCloudFormation(properties.maximumBatchingWindowInSeconds),
    };
}
// @ts-ignore TS6133
function CfnPipePipeSourceSqsQueueParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchSize', 'BatchSize', properties.BatchSize != null ? cfn_parse.FromCloudFormation.getNumber(properties.BatchSize) : undefined);
    ret.addPropertyResult('maximumBatchingWindowInSeconds', 'MaximumBatchingWindowInSeconds', properties.MaximumBatchingWindowInSeconds != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaximumBatchingWindowInSeconds) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetBatchJobParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetBatchJobParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetBatchJobParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('arrayProperties', CfnPipe_BatchArrayPropertiesPropertyValidator)(properties.arrayProperties));
    errors.collect(cdk.propertyValidator('containerOverrides', CfnPipe_BatchContainerOverridesPropertyValidator)(properties.containerOverrides));
    errors.collect(cdk.propertyValidator('dependsOn', cdk.listValidator(CfnPipe_BatchJobDependencyPropertyValidator))(properties.dependsOn));
    errors.collect(cdk.propertyValidator('jobDefinition', cdk.requiredValidator)(properties.jobDefinition));
    errors.collect(cdk.propertyValidator('jobDefinition', cdk.validateString)(properties.jobDefinition));
    errors.collect(cdk.propertyValidator('jobName', cdk.requiredValidator)(properties.jobName));
    errors.collect(cdk.propertyValidator('jobName', cdk.validateString)(properties.jobName));
    errors.collect(cdk.propertyValidator('parameters', cdk.hashValidator(cdk.validateString))(properties.parameters));
    errors.collect(cdk.propertyValidator('retryStrategy', CfnPipe_BatchRetryStrategyPropertyValidator)(properties.retryStrategy));
    return errors.wrap('supplied properties not correct for "PipeTargetBatchJobParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetBatchJobParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetBatchJobParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetBatchJobParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetBatchJobParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetBatchJobParametersPropertyValidator(properties).assertSuccess();
    return {
        ArrayProperties: cfnPipeBatchArrayPropertiesPropertyToCloudFormation(properties.arrayProperties),
        ContainerOverrides: cfnPipeBatchContainerOverridesPropertyToCloudFormation(properties.containerOverrides),
        DependsOn: cdk.listMapper(cfnPipeBatchJobDependencyPropertyToCloudFormation)(properties.dependsOn),
        JobDefinition: cdk.stringToCloudFormation(properties.jobDefinition),
        JobName: cdk.stringToCloudFormation(properties.jobName),
        Parameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.parameters),
        RetryStrategy: cfnPipeBatchRetryStrategyPropertyToCloudFormation(properties.retryStrategy),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetBatchJobParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('arrayProperties', 'ArrayProperties', properties.ArrayProperties != null ? CfnPipeBatchArrayPropertiesPropertyFromCloudFormation(properties.ArrayProperties) : undefined);
    ret.addPropertyResult('containerOverrides', 'ContainerOverrides', properties.ContainerOverrides != null ? CfnPipeBatchContainerOverridesPropertyFromCloudFormation(properties.ContainerOverrides) : undefined);
    ret.addPropertyResult('dependsOn', 'DependsOn', properties.DependsOn != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeBatchJobDependencyPropertyFromCloudFormation)(properties.DependsOn) : undefined);
    ret.addPropertyResult('jobDefinition', 'JobDefinition', cfn_parse.FromCloudFormation.getString(properties.JobDefinition));
    ret.addPropertyResult('jobName', 'JobName', cfn_parse.FromCloudFormation.getString(properties.JobName));
    ret.addPropertyResult('parameters', 'Parameters', properties.Parameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Parameters) : undefined);
    ret.addPropertyResult('retryStrategy', 'RetryStrategy', properties.RetryStrategy != null ? CfnPipeBatchRetryStrategyPropertyFromCloudFormation(properties.RetryStrategy) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetCloudWatchLogsParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetCloudWatchLogsParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetCloudWatchLogsParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('logStreamName', cdk.validateString)(properties.logStreamName));
    errors.collect(cdk.propertyValidator('timestamp', cdk.validateString)(properties.timestamp));
    return errors.wrap('supplied properties not correct for "PipeTargetCloudWatchLogsParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetCloudWatchLogsParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetCloudWatchLogsParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetCloudWatchLogsParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetCloudWatchLogsParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetCloudWatchLogsParametersPropertyValidator(properties).assertSuccess();
    return {
        LogStreamName: cdk.stringToCloudFormation(properties.logStreamName),
        Timestamp: cdk.stringToCloudFormation(properties.timestamp),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetCloudWatchLogsParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('logStreamName', 'LogStreamName', properties.LogStreamName != null ? cfn_parse.FromCloudFormation.getString(properties.LogStreamName) : undefined);
    ret.addPropertyResult('timestamp', 'Timestamp', properties.Timestamp != null ? cfn_parse.FromCloudFormation.getString(properties.Timestamp) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetEcsTaskParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetEcsTaskParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetEcsTaskParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('capacityProviderStrategy', cdk.listValidator(CfnPipe_CapacityProviderStrategyItemPropertyValidator))(properties.capacityProviderStrategy));
    errors.collect(cdk.propertyValidator('enableEcsManagedTags', cdk.validateBoolean)(properties.enableEcsManagedTags));
    errors.collect(cdk.propertyValidator('enableExecuteCommand', cdk.validateBoolean)(properties.enableExecuteCommand));
    errors.collect(cdk.propertyValidator('group', cdk.validateString)(properties.group));
    errors.collect(cdk.propertyValidator('launchType', cdk.validateString)(properties.launchType));
    errors.collect(cdk.propertyValidator('networkConfiguration', CfnPipe_NetworkConfigurationPropertyValidator)(properties.networkConfiguration));
    errors.collect(cdk.propertyValidator('overrides', CfnPipe_EcsTaskOverridePropertyValidator)(properties.overrides));
    errors.collect(cdk.propertyValidator('placementConstraints', cdk.listValidator(CfnPipe_PlacementConstraintPropertyValidator))(properties.placementConstraints));
    errors.collect(cdk.propertyValidator('placementStrategy', cdk.listValidator(CfnPipe_PlacementStrategyPropertyValidator))(properties.placementStrategy));
    errors.collect(cdk.propertyValidator('platformVersion', cdk.validateString)(properties.platformVersion));
    errors.collect(cdk.propertyValidator('propagateTags', cdk.validateString)(properties.propagateTags));
    errors.collect(cdk.propertyValidator('referenceId', cdk.validateString)(properties.referenceId));
    errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateCfnTag))(properties.tags));
    errors.collect(cdk.propertyValidator('taskCount', cdk.validateNumber)(properties.taskCount));
    errors.collect(cdk.propertyValidator('taskDefinitionArn', cdk.requiredValidator)(properties.taskDefinitionArn));
    errors.collect(cdk.propertyValidator('taskDefinitionArn', cdk.validateString)(properties.taskDefinitionArn));
    return errors.wrap('supplied properties not correct for "PipeTargetEcsTaskParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetEcsTaskParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetEcsTaskParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetEcsTaskParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetEcsTaskParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetEcsTaskParametersPropertyValidator(properties).assertSuccess();
    return {
        CapacityProviderStrategy: cdk.listMapper(cfnPipeCapacityProviderStrategyItemPropertyToCloudFormation)(properties.capacityProviderStrategy),
        EnableECSManagedTags: cdk.booleanToCloudFormation(properties.enableEcsManagedTags),
        EnableExecuteCommand: cdk.booleanToCloudFormation(properties.enableExecuteCommand),
        Group: cdk.stringToCloudFormation(properties.group),
        LaunchType: cdk.stringToCloudFormation(properties.launchType),
        NetworkConfiguration: cfnPipeNetworkConfigurationPropertyToCloudFormation(properties.networkConfiguration),
        Overrides: cfnPipeEcsTaskOverridePropertyToCloudFormation(properties.overrides),
        PlacementConstraints: cdk.listMapper(cfnPipePlacementConstraintPropertyToCloudFormation)(properties.placementConstraints),
        PlacementStrategy: cdk.listMapper(cfnPipePlacementStrategyPropertyToCloudFormation)(properties.placementStrategy),
        PlatformVersion: cdk.stringToCloudFormation(properties.platformVersion),
        PropagateTags: cdk.stringToCloudFormation(properties.propagateTags),
        ReferenceId: cdk.stringToCloudFormation(properties.referenceId),
        Tags: cdk.listMapper(cdk.cfnTagToCloudFormation)(properties.tags),
        TaskCount: cdk.numberToCloudFormation(properties.taskCount),
        TaskDefinitionArn: cdk.stringToCloudFormation(properties.taskDefinitionArn),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetEcsTaskParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('capacityProviderStrategy', 'CapacityProviderStrategy', properties.CapacityProviderStrategy != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeCapacityProviderStrategyItemPropertyFromCloudFormation)(properties.CapacityProviderStrategy) : undefined);
    ret.addPropertyResult('enableEcsManagedTags', 'EnableECSManagedTags', properties.EnableECSManagedTags != null ? cfn_parse.FromCloudFormation.getBoolean(properties.EnableECSManagedTags) : undefined);
    ret.addPropertyResult('enableExecuteCommand', 'EnableExecuteCommand', properties.EnableExecuteCommand != null ? cfn_parse.FromCloudFormation.getBoolean(properties.EnableExecuteCommand) : undefined);
    ret.addPropertyResult('group', 'Group', properties.Group != null ? cfn_parse.FromCloudFormation.getString(properties.Group) : undefined);
    ret.addPropertyResult('launchType', 'LaunchType', properties.LaunchType != null ? cfn_parse.FromCloudFormation.getString(properties.LaunchType) : undefined);
    ret.addPropertyResult('networkConfiguration', 'NetworkConfiguration', properties.NetworkConfiguration != null ? CfnPipeNetworkConfigurationPropertyFromCloudFormation(properties.NetworkConfiguration) : undefined);
    ret.addPropertyResult('overrides', 'Overrides', properties.Overrides != null ? CfnPipeEcsTaskOverridePropertyFromCloudFormation(properties.Overrides) : undefined);
    ret.addPropertyResult('placementConstraints', 'PlacementConstraints', properties.PlacementConstraints != null ? cfn_parse.FromCloudFormation.getArray(CfnPipePlacementConstraintPropertyFromCloudFormation)(properties.PlacementConstraints) : undefined);
    ret.addPropertyResult('placementStrategy', 'PlacementStrategy', properties.PlacementStrategy != null ? cfn_parse.FromCloudFormation.getArray(CfnPipePlacementStrategyPropertyFromCloudFormation)(properties.PlacementStrategy) : undefined);
    ret.addPropertyResult('platformVersion', 'PlatformVersion', properties.PlatformVersion != null ? cfn_parse.FromCloudFormation.getString(properties.PlatformVersion) : undefined);
    ret.addPropertyResult('propagateTags', 'PropagateTags', properties.PropagateTags != null ? cfn_parse.FromCloudFormation.getString(properties.PropagateTags) : undefined);
    ret.addPropertyResult('referenceId', 'ReferenceId', properties.ReferenceId != null ? cfn_parse.FromCloudFormation.getString(properties.ReferenceId) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getArray(cfn_parse.FromCloudFormation.getCfnTag)(properties.Tags) : undefined);
    ret.addPropertyResult('taskCount', 'TaskCount', properties.TaskCount != null ? cfn_parse.FromCloudFormation.getNumber(properties.TaskCount) : undefined);
    ret.addPropertyResult('taskDefinitionArn', 'TaskDefinitionArn', cfn_parse.FromCloudFormation.getString(properties.TaskDefinitionArn));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetEventBridgeEventBusParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetEventBridgeEventBusParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetEventBridgeEventBusParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('detailType', cdk.validateString)(properties.detailType));
    errors.collect(cdk.propertyValidator('endpointId', cdk.validateString)(properties.endpointId));
    errors.collect(cdk.propertyValidator('resources', cdk.listValidator(cdk.validateString))(properties.resources));
    errors.collect(cdk.propertyValidator('source', cdk.validateString)(properties.source));
    errors.collect(cdk.propertyValidator('time', cdk.validateString)(properties.time));
    return errors.wrap('supplied properties not correct for "PipeTargetEventBridgeEventBusParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetEventBridgeEventBusParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetEventBridgeEventBusParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetEventBridgeEventBusParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetEventBridgeEventBusParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetEventBridgeEventBusParametersPropertyValidator(properties).assertSuccess();
    return {
        DetailType: cdk.stringToCloudFormation(properties.detailType),
        EndpointId: cdk.stringToCloudFormation(properties.endpointId),
        Resources: cdk.listMapper(cdk.stringToCloudFormation)(properties.resources),
        Source: cdk.stringToCloudFormation(properties.source),
        Time: cdk.stringToCloudFormation(properties.time),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetEventBridgeEventBusParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('detailType', 'DetailType', properties.DetailType != null ? cfn_parse.FromCloudFormation.getString(properties.DetailType) : undefined);
    ret.addPropertyResult('endpointId', 'EndpointId', properties.EndpointId != null ? cfn_parse.FromCloudFormation.getString(properties.EndpointId) : undefined);
    ret.addPropertyResult('resources', 'Resources', properties.Resources != null ? cfn_parse.FromCloudFormation.getStringArray(properties.Resources) : undefined);
    ret.addPropertyResult('source', 'Source', properties.Source != null ? cfn_parse.FromCloudFormation.getString(properties.Source) : undefined);
    ret.addPropertyResult('time', 'Time', properties.Time != null ? cfn_parse.FromCloudFormation.getString(properties.Time) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetHttpParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetHttpParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetHttpParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('headerParameters', cdk.hashValidator(cdk.validateString))(properties.headerParameters));
    errors.collect(cdk.propertyValidator('pathParameterValues', cdk.listValidator(cdk.validateString))(properties.pathParameterValues));
    errors.collect(cdk.propertyValidator('queryStringParameters', cdk.hashValidator(cdk.validateString))(properties.queryStringParameters));
    return errors.wrap('supplied properties not correct for "PipeTargetHttpParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetHttpParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetHttpParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetHttpParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetHttpParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetHttpParametersPropertyValidator(properties).assertSuccess();
    return {
        HeaderParameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.headerParameters),
        PathParameterValues: cdk.listMapper(cdk.stringToCloudFormation)(properties.pathParameterValues),
        QueryStringParameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.queryStringParameters),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetHttpParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('headerParameters', 'HeaderParameters', properties.HeaderParameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.HeaderParameters) : undefined);
    ret.addPropertyResult('pathParameterValues', 'PathParameterValues', properties.PathParameterValues != null ? cfn_parse.FromCloudFormation.getStringArray(properties.PathParameterValues) : undefined);
    ret.addPropertyResult('queryStringParameters', 'QueryStringParameters', properties.QueryStringParameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.QueryStringParameters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetKinesisStreamParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetKinesisStreamParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetKinesisStreamParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('partitionKey', cdk.requiredValidator)(properties.partitionKey));
    errors.collect(cdk.propertyValidator('partitionKey', cdk.validateString)(properties.partitionKey));
    return errors.wrap('supplied properties not correct for "PipeTargetKinesisStreamParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetKinesisStreamParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetKinesisStreamParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetKinesisStreamParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetKinesisStreamParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetKinesisStreamParametersPropertyValidator(properties).assertSuccess();
    return {
        PartitionKey: cdk.stringToCloudFormation(properties.partitionKey),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetKinesisStreamParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('partitionKey', 'PartitionKey', cfn_parse.FromCloudFormation.getString(properties.PartitionKey));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetLambdaFunctionParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetLambdaFunctionParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetLambdaFunctionParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('invocationType', cdk.validateString)(properties.invocationType));
    return errors.wrap('supplied properties not correct for "PipeTargetLambdaFunctionParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetLambdaFunctionParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetLambdaFunctionParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetLambdaFunctionParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetLambdaFunctionParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetLambdaFunctionParametersPropertyValidator(properties).assertSuccess();
    return {
        InvocationType: cdk.stringToCloudFormation(properties.invocationType),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetLambdaFunctionParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('invocationType', 'InvocationType', properties.InvocationType != null ? cfn_parse.FromCloudFormation.getString(properties.InvocationType) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('batchJobParameters', CfnPipe_PipeTargetBatchJobParametersPropertyValidator)(properties.batchJobParameters));
    errors.collect(cdk.propertyValidator('cloudWatchLogsParameters', CfnPipe_PipeTargetCloudWatchLogsParametersPropertyValidator)(properties.cloudWatchLogsParameters));
    errors.collect(cdk.propertyValidator('ecsTaskParameters', CfnPipe_PipeTargetEcsTaskParametersPropertyValidator)(properties.ecsTaskParameters));
    errors.collect(cdk.propertyValidator('eventBridgeEventBusParameters', CfnPipe_PipeTargetEventBridgeEventBusParametersPropertyValidator)(properties.eventBridgeEventBusParameters));
    errors.collect(cdk.propertyValidator('httpParameters', CfnPipe_PipeTargetHttpParametersPropertyValidator)(properties.httpParameters));
    errors.collect(cdk.propertyValidator('inputTemplate', cdk.validateString)(properties.inputTemplate));
    errors.collect(cdk.propertyValidator('kinesisStreamParameters', CfnPipe_PipeTargetKinesisStreamParametersPropertyValidator)(properties.kinesisStreamParameters));
    errors.collect(cdk.propertyValidator('lambdaFunctionParameters', CfnPipe_PipeTargetLambdaFunctionParametersPropertyValidator)(properties.lambdaFunctionParameters));
    errors.collect(cdk.propertyValidator('redshiftDataParameters', CfnPipe_PipeTargetRedshiftDataParametersPropertyValidator)(properties.redshiftDataParameters));
    errors.collect(cdk.propertyValidator('sageMakerPipelineParameters', CfnPipe_PipeTargetSageMakerPipelineParametersPropertyValidator)(properties.sageMakerPipelineParameters));
    errors.collect(cdk.propertyValidator('sqsQueueParameters', CfnPipe_PipeTargetSqsQueueParametersPropertyValidator)(properties.sqsQueueParameters));
    errors.collect(cdk.propertyValidator('stepFunctionStateMachineParameters', CfnPipe_PipeTargetStateMachineParametersPropertyValidator)(properties.stepFunctionStateMachineParameters));
    return errors.wrap('supplied properties not correct for "PipeTargetParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetParametersPropertyValidator(properties).assertSuccess();
    return {
        BatchJobParameters: cfnPipePipeTargetBatchJobParametersPropertyToCloudFormation(properties.batchJobParameters),
        CloudWatchLogsParameters: cfnPipePipeTargetCloudWatchLogsParametersPropertyToCloudFormation(properties.cloudWatchLogsParameters),
        EcsTaskParameters: cfnPipePipeTargetEcsTaskParametersPropertyToCloudFormation(properties.ecsTaskParameters),
        EventBridgeEventBusParameters: cfnPipePipeTargetEventBridgeEventBusParametersPropertyToCloudFormation(properties.eventBridgeEventBusParameters),
        HttpParameters: cfnPipePipeTargetHttpParametersPropertyToCloudFormation(properties.httpParameters),
        InputTemplate: cdk.stringToCloudFormation(properties.inputTemplate),
        KinesisStreamParameters: cfnPipePipeTargetKinesisStreamParametersPropertyToCloudFormation(properties.kinesisStreamParameters),
        LambdaFunctionParameters: cfnPipePipeTargetLambdaFunctionParametersPropertyToCloudFormation(properties.lambdaFunctionParameters),
        RedshiftDataParameters: cfnPipePipeTargetRedshiftDataParametersPropertyToCloudFormation(properties.redshiftDataParameters),
        SageMakerPipelineParameters: cfnPipePipeTargetSageMakerPipelineParametersPropertyToCloudFormation(properties.sageMakerPipelineParameters),
        SqsQueueParameters: cfnPipePipeTargetSqsQueueParametersPropertyToCloudFormation(properties.sqsQueueParameters),
        StepFunctionStateMachineParameters: cfnPipePipeTargetStateMachineParametersPropertyToCloudFormation(properties.stepFunctionStateMachineParameters),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('batchJobParameters', 'BatchJobParameters', properties.BatchJobParameters != null ? CfnPipePipeTargetBatchJobParametersPropertyFromCloudFormation(properties.BatchJobParameters) : undefined);
    ret.addPropertyResult('cloudWatchLogsParameters', 'CloudWatchLogsParameters', properties.CloudWatchLogsParameters != null ? CfnPipePipeTargetCloudWatchLogsParametersPropertyFromCloudFormation(properties.CloudWatchLogsParameters) : undefined);
    ret.addPropertyResult('ecsTaskParameters', 'EcsTaskParameters', properties.EcsTaskParameters != null ? CfnPipePipeTargetEcsTaskParametersPropertyFromCloudFormation(properties.EcsTaskParameters) : undefined);
    ret.addPropertyResult('eventBridgeEventBusParameters', 'EventBridgeEventBusParameters', properties.EventBridgeEventBusParameters != null ? CfnPipePipeTargetEventBridgeEventBusParametersPropertyFromCloudFormation(properties.EventBridgeEventBusParameters) : undefined);
    ret.addPropertyResult('httpParameters', 'HttpParameters', properties.HttpParameters != null ? CfnPipePipeTargetHttpParametersPropertyFromCloudFormation(properties.HttpParameters) : undefined);
    ret.addPropertyResult('inputTemplate', 'InputTemplate', properties.InputTemplate != null ? cfn_parse.FromCloudFormation.getString(properties.InputTemplate) : undefined);
    ret.addPropertyResult('kinesisStreamParameters', 'KinesisStreamParameters', properties.KinesisStreamParameters != null ? CfnPipePipeTargetKinesisStreamParametersPropertyFromCloudFormation(properties.KinesisStreamParameters) : undefined);
    ret.addPropertyResult('lambdaFunctionParameters', 'LambdaFunctionParameters', properties.LambdaFunctionParameters != null ? CfnPipePipeTargetLambdaFunctionParametersPropertyFromCloudFormation(properties.LambdaFunctionParameters) : undefined);
    ret.addPropertyResult('redshiftDataParameters', 'RedshiftDataParameters', properties.RedshiftDataParameters != null ? CfnPipePipeTargetRedshiftDataParametersPropertyFromCloudFormation(properties.RedshiftDataParameters) : undefined);
    ret.addPropertyResult('sageMakerPipelineParameters', 'SageMakerPipelineParameters', properties.SageMakerPipelineParameters != null ? CfnPipePipeTargetSageMakerPipelineParametersPropertyFromCloudFormation(properties.SageMakerPipelineParameters) : undefined);
    ret.addPropertyResult('sqsQueueParameters', 'SqsQueueParameters', properties.SqsQueueParameters != null ? CfnPipePipeTargetSqsQueueParametersPropertyFromCloudFormation(properties.SqsQueueParameters) : undefined);
    ret.addPropertyResult('stepFunctionStateMachineParameters', 'StepFunctionStateMachineParameters', properties.StepFunctionStateMachineParameters != null ? CfnPipePipeTargetStateMachineParametersPropertyFromCloudFormation(properties.StepFunctionStateMachineParameters) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetRedshiftDataParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetRedshiftDataParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetRedshiftDataParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('database', cdk.requiredValidator)(properties.database));
    errors.collect(cdk.propertyValidator('database', cdk.validateString)(properties.database));
    errors.collect(cdk.propertyValidator('dbUser', cdk.validateString)(properties.dbUser));
    errors.collect(cdk.propertyValidator('secretManagerArn', cdk.validateString)(properties.secretManagerArn));
    errors.collect(cdk.propertyValidator('sqls', cdk.requiredValidator)(properties.sqls));
    errors.collect(cdk.propertyValidator('sqls', cdk.listValidator(cdk.validateString))(properties.sqls));
    errors.collect(cdk.propertyValidator('statementName', cdk.validateString)(properties.statementName));
    errors.collect(cdk.propertyValidator('withEvent', cdk.validateBoolean)(properties.withEvent));
    return errors.wrap('supplied properties not correct for "PipeTargetRedshiftDataParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetRedshiftDataParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetRedshiftDataParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetRedshiftDataParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetRedshiftDataParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetRedshiftDataParametersPropertyValidator(properties).assertSuccess();
    return {
        Database: cdk.stringToCloudFormation(properties.database),
        DbUser: cdk.stringToCloudFormation(properties.dbUser),
        SecretManagerArn: cdk.stringToCloudFormation(properties.secretManagerArn),
        Sqls: cdk.listMapper(cdk.stringToCloudFormation)(properties.sqls),
        StatementName: cdk.stringToCloudFormation(properties.statementName),
        WithEvent: cdk.booleanToCloudFormation(properties.withEvent),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetRedshiftDataParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('database', 'Database', cfn_parse.FromCloudFormation.getString(properties.Database));
    ret.addPropertyResult('dbUser', 'DbUser', properties.DbUser != null ? cfn_parse.FromCloudFormation.getString(properties.DbUser) : undefined);
    ret.addPropertyResult('secretManagerArn', 'SecretManagerArn', properties.SecretManagerArn != null ? cfn_parse.FromCloudFormation.getString(properties.SecretManagerArn) : undefined);
    ret.addPropertyResult('sqls', 'Sqls', cfn_parse.FromCloudFormation.getStringArray(properties.Sqls));
    ret.addPropertyResult('statementName', 'StatementName', properties.StatementName != null ? cfn_parse.FromCloudFormation.getString(properties.StatementName) : undefined);
    ret.addPropertyResult('withEvent', 'WithEvent', properties.WithEvent != null ? cfn_parse.FromCloudFormation.getBoolean(properties.WithEvent) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetSageMakerPipelineParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetSageMakerPipelineParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetSageMakerPipelineParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('pipelineParameterList', cdk.listValidator(CfnPipe_SageMakerPipelineParameterPropertyValidator))(properties.pipelineParameterList));
    return errors.wrap('supplied properties not correct for "PipeTargetSageMakerPipelineParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetSageMakerPipelineParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetSageMakerPipelineParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetSageMakerPipelineParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetSageMakerPipelineParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetSageMakerPipelineParametersPropertyValidator(properties).assertSuccess();
    return {
        PipelineParameterList: cdk.listMapper(cfnPipeSageMakerPipelineParameterPropertyToCloudFormation)(properties.pipelineParameterList),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetSageMakerPipelineParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('pipelineParameterList', 'PipelineParameterList', properties.PipelineParameterList != null ? cfn_parse.FromCloudFormation.getArray(CfnPipeSageMakerPipelineParameterPropertyFromCloudFormation)(properties.PipelineParameterList) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetSqsQueueParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetSqsQueueParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetSqsQueueParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('messageDeduplicationId', cdk.validateString)(properties.messageDeduplicationId));
    errors.collect(cdk.propertyValidator('messageGroupId', cdk.validateString)(properties.messageGroupId));
    return errors.wrap('supplied properties not correct for "PipeTargetSqsQueueParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetSqsQueueParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetSqsQueueParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetSqsQueueParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetSqsQueueParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetSqsQueueParametersPropertyValidator(properties).assertSuccess();
    return {
        MessageDeduplicationId: cdk.stringToCloudFormation(properties.messageDeduplicationId),
        MessageGroupId: cdk.stringToCloudFormation(properties.messageGroupId),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetSqsQueueParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('messageDeduplicationId', 'MessageDeduplicationId', properties.MessageDeduplicationId != null ? cfn_parse.FromCloudFormation.getString(properties.MessageDeduplicationId) : undefined);
    ret.addPropertyResult('messageGroupId', 'MessageGroupId', properties.MessageGroupId != null ? cfn_parse.FromCloudFormation.getString(properties.MessageGroupId) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PipeTargetStateMachineParametersProperty`
 *
 * @param properties - the TypeScript properties of a `PipeTargetStateMachineParametersProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PipeTargetStateMachineParametersPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('invocationType', cdk.validateString)(properties.invocationType));
    return errors.wrap('supplied properties not correct for "PipeTargetStateMachineParametersProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetStateMachineParameters` resource
 *
 * @param properties - the TypeScript properties of a `PipeTargetStateMachineParametersProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PipeTargetStateMachineParameters` resource.
 */
// @ts-ignore TS6133
function cfnPipePipeTargetStateMachineParametersPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PipeTargetStateMachineParametersPropertyValidator(properties).assertSuccess();
    return {
        InvocationType: cdk.stringToCloudFormation(properties.invocationType),
    };
}
// @ts-ignore TS6133
function CfnPipePipeTargetStateMachineParametersPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('invocationType', 'InvocationType', properties.InvocationType != null ? cfn_parse.FromCloudFormation.getString(properties.InvocationType) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PlacementConstraintProperty`
 *
 * @param properties - the TypeScript properties of a `PlacementConstraintProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PlacementConstraintPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('expression', cdk.validateString)(properties.expression));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "PlacementConstraintProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PlacementConstraint` resource
 *
 * @param properties - the TypeScript properties of a `PlacementConstraintProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PlacementConstraint` resource.
 */
// @ts-ignore TS6133
function cfnPipePlacementConstraintPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PlacementConstraintPropertyValidator(properties).assertSuccess();
    return {
        Expression: cdk.stringToCloudFormation(properties.expression),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnPipePlacementConstraintPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('expression', 'Expression', properties.Expression != null ? cfn_parse.FromCloudFormation.getString(properties.Expression) : undefined);
    ret.addPropertyResult('type', 'Type', properties.Type != null ? cfn_parse.FromCloudFormation.getString(properties.Type) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `PlacementStrategyProperty`
 *
 * @param properties - the TypeScript properties of a `PlacementStrategyProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_PlacementStrategyPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('field', cdk.validateString)(properties.field));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "PlacementStrategyProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PlacementStrategy` resource
 *
 * @param properties - the TypeScript properties of a `PlacementStrategyProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.PlacementStrategy` resource.
 */
// @ts-ignore TS6133
function cfnPipePlacementStrategyPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_PlacementStrategyPropertyValidator(properties).assertSuccess();
    return {
        Field: cdk.stringToCloudFormation(properties.field),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnPipePlacementStrategyPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('field', 'Field', properties.Field != null ? cfn_parse.FromCloudFormation.getString(properties.Field) : undefined);
    ret.addPropertyResult('type', 'Type', properties.Type != null ? cfn_parse.FromCloudFormation.getString(properties.Type) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `SageMakerPipelineParameterProperty`
 *
 * @param properties - the TypeScript properties of a `SageMakerPipelineParameterProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_SageMakerPipelineParameterPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "SageMakerPipelineParameterProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SageMakerPipelineParameter` resource
 *
 * @param properties - the TypeScript properties of a `SageMakerPipelineParameterProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SageMakerPipelineParameter` resource.
 */
// @ts-ignore TS6133
function cfnPipeSageMakerPipelineParameterPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_SageMakerPipelineParameterPropertyValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnPipeSageMakerPipelineParameterPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', cfn_parse.FromCloudFormation.getString(properties.Name));
    ret.addPropertyResult('value', 'Value', cfn_parse.FromCloudFormation.getString(properties.Value));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `SelfManagedKafkaAccessConfigurationCredentialsProperty`
 *
 * @param properties - the TypeScript properties of a `SelfManagedKafkaAccessConfigurationCredentialsProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_SelfManagedKafkaAccessConfigurationCredentialsPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('basicAuth', cdk.validateString)(properties.basicAuth));
    errors.collect(cdk.propertyValidator('clientCertificateTlsAuth', cdk.validateString)(properties.clientCertificateTlsAuth));
    errors.collect(cdk.propertyValidator('saslScram256Auth', cdk.validateString)(properties.saslScram256Auth));
    errors.collect(cdk.propertyValidator('saslScram512Auth', cdk.validateString)(properties.saslScram512Auth));
    return errors.wrap('supplied properties not correct for "SelfManagedKafkaAccessConfigurationCredentialsProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SelfManagedKafkaAccessConfigurationCredentials` resource
 *
 * @param properties - the TypeScript properties of a `SelfManagedKafkaAccessConfigurationCredentialsProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SelfManagedKafkaAccessConfigurationCredentials` resource.
 */
// @ts-ignore TS6133
function cfnPipeSelfManagedKafkaAccessConfigurationCredentialsPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_SelfManagedKafkaAccessConfigurationCredentialsPropertyValidator(properties).assertSuccess();
    return {
        BasicAuth: cdk.stringToCloudFormation(properties.basicAuth),
        ClientCertificateTlsAuth: cdk.stringToCloudFormation(properties.clientCertificateTlsAuth),
        SaslScram256Auth: cdk.stringToCloudFormation(properties.saslScram256Auth),
        SaslScram512Auth: cdk.stringToCloudFormation(properties.saslScram512Auth),
    };
}
// @ts-ignore TS6133
function CfnPipeSelfManagedKafkaAccessConfigurationCredentialsPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('basicAuth', 'BasicAuth', properties.BasicAuth != null ? cfn_parse.FromCloudFormation.getString(properties.BasicAuth) : undefined);
    ret.addPropertyResult('clientCertificateTlsAuth', 'ClientCertificateTlsAuth', properties.ClientCertificateTlsAuth != null ? cfn_parse.FromCloudFormation.getString(properties.ClientCertificateTlsAuth) : undefined);
    ret.addPropertyResult('saslScram256Auth', 'SaslScram256Auth', properties.SaslScram256Auth != null ? cfn_parse.FromCloudFormation.getString(properties.SaslScram256Auth) : undefined);
    ret.addPropertyResult('saslScram512Auth', 'SaslScram512Auth', properties.SaslScram512Auth != null ? cfn_parse.FromCloudFormation.getString(properties.SaslScram512Auth) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `SelfManagedKafkaAccessConfigurationVpcProperty`
 *
 * @param properties - the TypeScript properties of a `SelfManagedKafkaAccessConfigurationVpcProperty`
 *
 * @returns the result of the validation.
 */
function CfnPipe_SelfManagedKafkaAccessConfigurationVpcPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('securityGroup', cdk.listValidator(cdk.validateString))(properties.securityGroup));
    errors.collect(cdk.propertyValidator('subnets', cdk.listValidator(cdk.validateString))(properties.subnets));
    return errors.wrap('supplied properties not correct for "SelfManagedKafkaAccessConfigurationVpcProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SelfManagedKafkaAccessConfigurationVpc` resource
 *
 * @param properties - the TypeScript properties of a `SelfManagedKafkaAccessConfigurationVpcProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Pipes::Pipe.SelfManagedKafkaAccessConfigurationVpc` resource.
 */
// @ts-ignore TS6133
function cfnPipeSelfManagedKafkaAccessConfigurationVpcPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnPipe_SelfManagedKafkaAccessConfigurationVpcPropertyValidator(properties).assertSuccess();
    return {
        SecurityGroup: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroup),
        Subnets: cdk.listMapper(cdk.stringToCloudFormation)(properties.subnets),
    };
}
// @ts-ignore TS6133
function CfnPipeSelfManagedKafkaAccessConfigurationVpcPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('securityGroup', 'SecurityGroup', properties.SecurityGroup != null ? cfn_parse.FromCloudFormation.getStringArray(properties.SecurityGroup) : undefined);
    ret.addPropertyResult('subnets', 'Subnets', properties.Subnets != null ? cfn_parse.FromCloudFormation.getStringArray(properties.Subnets) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
