// aws-cdk-lib/aws-servicecatalog/lib/product-stack.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductStack = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const cdk = require("../../core");
const product_stack_synthesizer_1 = require("./private/product-stack-synthesizer");
/**
 * A Service Catalog product stack, which is similar in form to a Cloudformation nested stack.
 * You can add the resources to this stack that you want to define for your service catalog product.
 *
 * This stack will not be treated as an independent deployment
 * artifact (won't be listed in "cdk list" or deployable through "cdk deploy"),
 * but rather only synthesized as a template and uploaded as an asset to S3.
 *
 */
class ProductStack extends cdk.Stack {
    constructor(scope, id, props = {}) {
        super(scope, id, {
            synthesizer: new product_stack_synthesizer_1.ProductStackSynthesizer(props.assetBucket),
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_servicecatalog_ProductStackProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, ProductStack);
            }
            throw error;
        }
        this._parentStack = findParentStack(scope);
        // this is the file name of the synthesized template file within the cloud assembly
        this.templateFile = `${cdk.Names.uniqueId(this)}.product.template.json`;
        this.assetBucket = props.assetBucket;
    }
    /**
     * Set the parent product stack history
     *
     * @internal
     */
    _setParentProductStackHistory(parentProductStackHistory) {
        return this._parentProductStackHistory = parentProductStackHistory;
    }
    /**
     * Fetch the template URL.
     *
     * @internal
     */
    _getTemplateUrl() {
        return cdk.Lazy.uncachedString({ produce: () => this._templateUrl });
    }
    /**
     * Fetch the asset bucket.
     *
     * @internal
     */
    _getAssetBucket() {
        return this.assetBucket;
    }
    /**
     * Fetch the parent Stack.
     *
     * @internal
     */
    _getParentStack() {
        return this._parentStack;
    }
    /**
     * Synthesize the product stack template, overrides the `super` class method.
     *
     * Defines an asset at the parent stack which represents the template of this
     * product stack.
     *
     * @internal
     */
    _synthesizeTemplate(session) {
        const cfn = JSON.stringify(this._toCloudFormation(), undefined, 2);
        const templateHash = crypto.createHash('sha256').update(cfn).digest('hex');
        this._templateUrl = this._parentStack.synthesizer.addFileAsset({
            packaging: cdk.FileAssetPackaging.FILE,
            sourceHash: templateHash,
            fileName: this.templateFile,
        }).httpUrl;
        if (this._parentProductStackHistory) {
            this._parentProductStackHistory._writeTemplateToSnapshot(cfn);
        }
        fs.writeFileSync(path.join(session.assembly.outdir, this.templateFile), cfn);
    }
}
_a = JSII_RTTI_SYMBOL_1;
ProductStack[_a] = { fqn: "aws-cdk-lib.aws_servicecatalog.ProductStack", version: "2.75.1" };
exports.ProductStack = ProductStack;
/**
 * Validates the scope for a product stack, which must be defined within the scope of another `Stack`.
 */
function findParentStack(scope) {
    try {
        const parentStack = cdk.Stack.of(scope);
        return parentStack;
    }
    catch {
        throw new Error('Product stacks must be defined within scope of another non-product stack');
    }
}
