// aws-cdk-lib/aws-opensearchservice/lib/opensearch-access-policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenSearchAccessPolicy = void 0;
const iam = require("../../aws-iam");
const cdk = require("../../core");
const cr = require("../../custom-resources");
/**
 * Creates LogGroup resource policies.
 */
class OpenSearchAccessPolicy extends cr.AwsCustomResource {
    constructor(scope, id, props) {
        super(scope, id, {
            resourceType: 'Custom::OpenSearchAccessPolicy',
            onUpdate: {
                action: 'updateDomainConfig',
                service: 'OpenSearch',
                parameters: {
                    DomainName: props.domainName,
                    AccessPolicies: cdk.Lazy.string({
                        produce: () => JSON.stringify(new iam.PolicyDocument({
                            statements: this.accessPolicyStatements,
                        }).toJSON()),
                    }),
                },
                // this is needed to limit the response body, otherwise it exceeds the CFN 4k limit
                outputPaths: ['DomainConfig.AccessPolicies'],
                physicalResourceId: cr.PhysicalResourceId.of(`${props.domainName}AccessPolicy`),
            },
            policy: cr.AwsCustomResourcePolicy.fromStatements([new iam.PolicyStatement({ actions: ['es:UpdateDomainConfig'], resources: [props.domainArn] })]),
        });
        this.accessPolicyStatements = [];
        this.addAccessPolicies(...props.accessPolicies);
    }
    /**
     * Add policy statements to the domain access policy
     */
    addAccessPolicies(...accessPolicyStatements) {
        this.accessPolicyStatements.push(...accessPolicyStatements);
    }
}
exports.OpenSearchAccessPolicy = OpenSearchAccessPolicy;
