// aws-cdk-lib/aws-cloudwatch-actions/lib/ssm.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmIncidentAction = exports.SsmAction = exports.OpsItemCategory = exports.OpsItemSeverity = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
/**
 * Types of OpsItem severity available
 */
var OpsItemSeverity;
(function (OpsItemSeverity) {
    /**
     * Set the severity to critical
     */
    OpsItemSeverity["CRITICAL"] = "1";
    /**
     * Set the severity to high
     */
    OpsItemSeverity["HIGH"] = "2";
    /**
     * Set the severity to medium
     */
    OpsItemSeverity["MEDIUM"] = "3";
    /**
     * Set the severity to low
     */
    OpsItemSeverity["LOW"] = "4";
})(OpsItemSeverity = exports.OpsItemSeverity || (exports.OpsItemSeverity = {}));
/**
 * Types of OpsItem category available
 */
var OpsItemCategory;
(function (OpsItemCategory) {
    /**
     * Set the category to availability
     */
    OpsItemCategory["AVAILABILITY"] = "Availability";
    /**
     * Set the category to cost
     */
    OpsItemCategory["COST"] = "Cost";
    /**
     * Set the category to performance
     */
    OpsItemCategory["PERFORMANCE"] = "Performance";
    /**
     * Set the category to recovery
     */
    OpsItemCategory["RECOVERY"] = "Recovery";
    /**
     * Set the category to security
     */
    OpsItemCategory["SECURITY"] = "Security";
})(OpsItemCategory = exports.OpsItemCategory || (exports.OpsItemCategory = {}));
/**
 * Use an SSM OpsItem action as an Alarm action
 */
class SsmAction {
    constructor(severity, category) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_actions_OpsItemSeverity(severity);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_actions_OpsItemCategory(category);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, SsmAction);
            }
            throw error;
        }
        this.severity = severity;
        this.category = category;
    }
    /**
     * Returns an alarm action configuration to use an SSM OpsItem action as an alarm action
     */
    bind(_scope, _alarm) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_IAlarm(_alarm);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bind);
            }
            throw error;
        }
        if (this.category === undefined) {
            return { alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:ssm:${core_1.Stack.of(_scope).region}:${core_1.Stack.of(_scope).account}:opsitem:${this.severity}` };
        }
        else {
            return { alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:ssm:${core_1.Stack.of(_scope).region}:${core_1.Stack.of(_scope).account}:opsitem:${this.severity}#CATEGORY=${this.category}` };
        }
    }
}
_a = JSII_RTTI_SYMBOL_1;
SsmAction[_a] = { fqn: "aws-cdk-lib.aws_cloudwatch_actions.SsmAction", version: "2.75.1" };
exports.SsmAction = SsmAction;
/**
 * Use an SSM Incident Response Plan as an Alarm action
 */
class SsmIncidentAction {
    constructor(responsePlanName) {
        this.responsePlanName = responsePlanName;
    }
    /**
     * Returns an alarm action configuration to use an SSM Incident as an alarm action
     * based on an Incident Manager Response Plan
     */
    bind(_scope, _alarm) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_IAlarm(_alarm);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bind);
            }
            throw error;
        }
        return {
            alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:ssm-incidents::${core_1.Stack.of(_scope).account}:response-plan/${this.responsePlanName}`,
        };
    }
}
_b = JSII_RTTI_SYMBOL_1;
SsmIncidentAction[_b] = { fqn: "aws-cdk-lib.aws_cloudwatch_actions.SsmIncidentAction", version: "2.75.1" };
exports.SsmIncidentAction = SsmIncidentAction;
