// aws-cdk-lib/aws-sagemaker/lib/sagemaker-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SageMakerMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class SageMakerMetrics {
    static invocationsSum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'Invocations',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static invocation5XxErrorsSum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'Invocation5XXErrors',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static invocation4XxErrorsSum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'Invocation4XXErrors',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static invocationsPerInstanceSum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'InvocationsPerInstance',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static modelLatencySum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'ModelLatency',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static overheadLatencySum(dimensions) {
        return {
            namespace: 'AWS/SageMaker',
            metricName: 'OverheadLatency',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.SageMakerMetrics = SageMakerMetrics;
