// aws-cdk-lib/aws-cloudfront-origins/lib/http-origin.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpOrigin = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cloudfront = require("../../aws-cloudfront");
const utils_1 = require("./private/utils");
/**
 * An Origin for an HTTP server or S3 bucket configured for website hosting.
 */
class HttpOrigin extends cloudfront.OriginBase {
    constructor(domainName, props = {}) {
        super(domainName, props);
        this.props = props;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudfront_origins_HttpOriginProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, HttpOrigin);
            }
            throw error;
        }
        (0, utils_1.validateSecondsInRangeOrUndefined)('readTimeout', 1, 180, props.readTimeout);
        (0, utils_1.validateSecondsInRangeOrUndefined)('keepaliveTimeout', 1, 180, props.keepaliveTimeout);
    }
    renderCustomOriginConfig() {
        return {
            originSslProtocols: this.props.originSslProtocols ?? [cloudfront.OriginSslPolicy.TLS_V1_2],
            originProtocolPolicy: this.props.protocolPolicy ?? cloudfront.OriginProtocolPolicy.HTTPS_ONLY,
            httpPort: this.props.httpPort,
            httpsPort: this.props.httpsPort,
            originReadTimeout: this.props.readTimeout?.toSeconds(),
            originKeepaliveTimeout: this.props.keepaliveTimeout?.toSeconds(),
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
HttpOrigin[_a] = { fqn: "aws-cdk-lib.aws_cloudfront_origins.HttpOrigin", version: "2.75.1" };
exports.HttpOrigin = HttpOrigin;
