// aws-cdk-lib/aws-logs-destinations/lib/lambda.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaDestination = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("../../aws-iam");
/**
 * Use a Lambda Function as the destination for a log subscription
 */
class LambdaDestination {
    /**  LambdaDestinationOptions */
    constructor(fn, options = {}) {
        this.fn = fn;
        this.options = options;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_lambda_IFunction(fn);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_logs_destinations_LambdaDestinationOptions(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, LambdaDestination);
            }
            throw error;
        }
    }
    bind(scope, logGroup) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_logs_ILogGroup(logGroup);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bind);
            }
            throw error;
        }
        const arn = logGroup.logGroupArn;
        if (this.options.addPermissions !== false) {
            const permissionId = 'CanInvokeLambda';
            this.fn.addPermission(permissionId, {
                principal: new iam.ServicePrincipal('logs.amazonaws.com'),
                sourceArn: arn,
                // Using SubScription Filter as scope is okay, since every Subscription Filter has only
                // one destination.
                scope,
            });
            // Need to add a dependency, otherwise the SubscriptionFilter can be created before the
            // Permission that allows the interaction.
            const cfnPermission = scope.node.tryFindChild(permissionId);
            if (cfnPermission) {
                scope.node.addDependency(cfnPermission);
            }
        }
        return { arn: this.fn.functionArn };
    }
}
_a = JSII_RTTI_SYMBOL_1;
LambdaDestination[_a] = { fqn: "aws-cdk-lib.aws_logs_destinations.LambdaDestination", version: "2.75.1" };
exports.LambdaDestination = LambdaDestination;
