// aws-cdk-lib/aws-omics/lib/omics.generated.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnWorkflow = exports.CfnVariantStore = exports.CfnSequenceStore = exports.CfnRunGroup = exports.CfnReferenceStore = exports.CfnAnnotationStore = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("../../core");
const cfn_parse = require("../../core/lib/helpers-internal");
/**
 * Determine whether the given properties match those of a `CfnAnnotationStoreProps`
 *
 * @param properties - the TypeScript properties of a `CfnAnnotationStoreProps`
 *
 * @returns the result of the validation.
 */
function CfnAnnotationStorePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('reference', CfnAnnotationStore_ReferenceItemPropertyValidator)(properties.reference));
    errors.collect(cdk.propertyValidator('sseConfig', CfnAnnotationStore_SseConfigPropertyValidator)(properties.sseConfig));
    errors.collect(cdk.propertyValidator('storeFormat', cdk.requiredValidator)(properties.storeFormat));
    errors.collect(cdk.propertyValidator('storeFormat', cdk.validateString)(properties.storeFormat));
    errors.collect(cdk.propertyValidator('storeOptions', CfnAnnotationStore_StoreOptionsPropertyValidator)(properties.storeOptions));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnAnnotationStoreProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore` resource
 *
 * @param properties - the TypeScript properties of a `CfnAnnotationStoreProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore` resource.
 */
// @ts-ignore TS6133
function cfnAnnotationStorePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnAnnotationStorePropsValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        StoreFormat: cdk.stringToCloudFormation(properties.storeFormat),
        Description: cdk.stringToCloudFormation(properties.description),
        Reference: cfnAnnotationStoreReferenceItemPropertyToCloudFormation(properties.reference),
        SseConfig: cfnAnnotationStoreSseConfigPropertyToCloudFormation(properties.sseConfig),
        StoreOptions: cfnAnnotationStoreStoreOptionsPropertyToCloudFormation(properties.storeOptions),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnAnnotationStorePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', cfn_parse.FromCloudFormation.getString(properties.Name));
    ret.addPropertyResult('storeFormat', 'StoreFormat', cfn_parse.FromCloudFormation.getString(properties.StoreFormat));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('reference', 'Reference', properties.Reference != null ? CfnAnnotationStoreReferenceItemPropertyFromCloudFormation(properties.Reference) : undefined);
    ret.addPropertyResult('sseConfig', 'SseConfig', properties.SseConfig != null ? CfnAnnotationStoreSseConfigPropertyFromCloudFormation(properties.SseConfig) : undefined);
    ret.addPropertyResult('storeOptions', 'StoreOptions', properties.StoreOptions != null ? CfnAnnotationStoreStoreOptionsPropertyFromCloudFormation(properties.StoreOptions) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::AnnotationStore`
 *
 * Creates an annotation store.
 *
 * @cloudformationResource AWS::Omics::AnnotationStore
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-annotationstore.html
 */
class CfnAnnotationStore extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnAnnotationStorePropsFromCloudFormation(resourceProperties);
        const ret = new CfnAnnotationStore(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::AnnotationStore`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnAnnotationStore.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnAnnotationStoreProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnAnnotationStore);
            }
            throw error;
        }
        cdk.requireProperty(props, 'name', this);
        cdk.requireProperty(props, 'storeFormat', this);
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrId = cdk.Token.asString(this.getAtt('Id', cdk.ResolutionTypeHint.STRING));
        this.attrStatus = cdk.Token.asString(this.getAtt('Status', cdk.ResolutionTypeHint.STRING));
        this.attrStatusMessage = cdk.Token.asString(this.getAtt('StatusMessage', cdk.ResolutionTypeHint.STRING));
        this.attrStoreArn = cdk.Token.asString(this.getAtt('StoreArn', cdk.ResolutionTypeHint.STRING));
        this.attrStoreSizeBytes = this.getAtt('StoreSizeBytes', cdk.ResolutionTypeHint.STRING);
        this.attrUpdateTime = cdk.Token.asString(this.getAtt('UpdateTime', cdk.ResolutionTypeHint.STRING));
        this.name = props.name;
        this.storeFormat = props.storeFormat;
        this.description = props.description;
        this.reference = props.reference;
        this.sseConfig = props.sseConfig;
        this.storeOptions = props.storeOptions;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::AnnotationStore", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnAnnotationStore.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            name: this.name,
            storeFormat: this.storeFormat,
            description: this.description,
            reference: this.reference,
            sseConfig: this.sseConfig,
            storeOptions: this.storeOptions,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnAnnotationStorePropsToCloudFormation(props);
    }
}
_a = JSII_RTTI_SYMBOL_1;
CfnAnnotationStore[_a] = { fqn: "aws-cdk-lib.aws_omics.CfnAnnotationStore", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnAnnotationStore.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::AnnotationStore";
exports.CfnAnnotationStore = CfnAnnotationStore;
/**
 * Determine whether the given properties match those of a `ReferenceItemProperty`
 *
 * @param properties - the TypeScript properties of a `ReferenceItemProperty`
 *
 * @returns the result of the validation.
 */
function CfnAnnotationStore_ReferenceItemPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('referenceArn', cdk.requiredValidator)(properties.referenceArn));
    errors.collect(cdk.propertyValidator('referenceArn', cdk.validateString)(properties.referenceArn));
    return errors.wrap('supplied properties not correct for "ReferenceItemProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.ReferenceItem` resource
 *
 * @param properties - the TypeScript properties of a `ReferenceItemProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.ReferenceItem` resource.
 */
// @ts-ignore TS6133
function cfnAnnotationStoreReferenceItemPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnAnnotationStore_ReferenceItemPropertyValidator(properties).assertSuccess();
    return {
        ReferenceArn: cdk.stringToCloudFormation(properties.referenceArn),
    };
}
// @ts-ignore TS6133
function CfnAnnotationStoreReferenceItemPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('referenceArn', 'ReferenceArn', cfn_parse.FromCloudFormation.getString(properties.ReferenceArn));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `SseConfigProperty`
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the result of the validation.
 */
function CfnAnnotationStore_SseConfigPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('keyArn', cdk.validateString)(properties.keyArn));
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "SseConfigProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.SseConfig` resource
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.SseConfig` resource.
 */
// @ts-ignore TS6133
function cfnAnnotationStoreSseConfigPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnAnnotationStore_SseConfigPropertyValidator(properties).assertSuccess();
    return {
        KeyArn: cdk.stringToCloudFormation(properties.keyArn),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnAnnotationStoreSseConfigPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('keyArn', 'KeyArn', properties.KeyArn != null ? cfn_parse.FromCloudFormation.getString(properties.KeyArn) : undefined);
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `StoreOptionsProperty`
 *
 * @param properties - the TypeScript properties of a `StoreOptionsProperty`
 *
 * @returns the result of the validation.
 */
function CfnAnnotationStore_StoreOptionsPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('tsvStoreOptions', cdk.requiredValidator)(properties.tsvStoreOptions));
    errors.collect(cdk.propertyValidator('tsvStoreOptions', CfnAnnotationStore_TsvStoreOptionsPropertyValidator)(properties.tsvStoreOptions));
    return errors.wrap('supplied properties not correct for "StoreOptionsProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.StoreOptions` resource
 *
 * @param properties - the TypeScript properties of a `StoreOptionsProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.StoreOptions` resource.
 */
// @ts-ignore TS6133
function cfnAnnotationStoreStoreOptionsPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnAnnotationStore_StoreOptionsPropertyValidator(properties).assertSuccess();
    return {
        TsvStoreOptions: cfnAnnotationStoreTsvStoreOptionsPropertyToCloudFormation(properties.tsvStoreOptions),
    };
}
// @ts-ignore TS6133
function CfnAnnotationStoreStoreOptionsPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('tsvStoreOptions', 'TsvStoreOptions', CfnAnnotationStoreTsvStoreOptionsPropertyFromCloudFormation(properties.TsvStoreOptions));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `TsvStoreOptionsProperty`
 *
 * @param properties - the TypeScript properties of a `TsvStoreOptionsProperty`
 *
 * @returns the result of the validation.
 */
function CfnAnnotationStore_TsvStoreOptionsPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('annotationType', cdk.validateString)(properties.annotationType));
    errors.collect(cdk.propertyValidator('formatToHeader', cdk.hashValidator(cdk.validateString))(properties.formatToHeader));
    errors.collect(cdk.propertyValidator('schema', cdk.validateObject)(properties.schema));
    return errors.wrap('supplied properties not correct for "TsvStoreOptionsProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.TsvStoreOptions` resource
 *
 * @param properties - the TypeScript properties of a `TsvStoreOptionsProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::AnnotationStore.TsvStoreOptions` resource.
 */
// @ts-ignore TS6133
function cfnAnnotationStoreTsvStoreOptionsPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnAnnotationStore_TsvStoreOptionsPropertyValidator(properties).assertSuccess();
    return {
        AnnotationType: cdk.stringToCloudFormation(properties.annotationType),
        FormatToHeader: cdk.hashMapper(cdk.stringToCloudFormation)(properties.formatToHeader),
        Schema: cdk.objectToCloudFormation(properties.schema),
    };
}
// @ts-ignore TS6133
function CfnAnnotationStoreTsvStoreOptionsPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('annotationType', 'AnnotationType', properties.AnnotationType != null ? cfn_parse.FromCloudFormation.getString(properties.AnnotationType) : undefined);
    ret.addPropertyResult('formatToHeader', 'FormatToHeader', properties.FormatToHeader != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.FormatToHeader) : undefined);
    ret.addPropertyResult('schema', 'Schema', properties.Schema != null ? cfn_parse.FromCloudFormation.getAny(properties.Schema) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `CfnReferenceStoreProps`
 *
 * @param properties - the TypeScript properties of a `CfnReferenceStoreProps`
 *
 * @returns the result of the validation.
 */
function CfnReferenceStorePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('sseConfig', CfnReferenceStore_SseConfigPropertyValidator)(properties.sseConfig));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnReferenceStoreProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::ReferenceStore` resource
 *
 * @param properties - the TypeScript properties of a `CfnReferenceStoreProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::ReferenceStore` resource.
 */
// @ts-ignore TS6133
function cfnReferenceStorePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnReferenceStorePropsValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Description: cdk.stringToCloudFormation(properties.description),
        SseConfig: cfnReferenceStoreSseConfigPropertyToCloudFormation(properties.sseConfig),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnReferenceStorePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', cfn_parse.FromCloudFormation.getString(properties.Name));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('sseConfig', 'SseConfig', properties.SseConfig != null ? CfnReferenceStoreSseConfigPropertyFromCloudFormation(properties.SseConfig) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::ReferenceStore`
 *
 * Creates a reference store.
 *
 * @cloudformationResource AWS::Omics::ReferenceStore
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-referencestore.html
 */
class CfnReferenceStore extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnReferenceStorePropsFromCloudFormation(resourceProperties);
        const ret = new CfnReferenceStore(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::ReferenceStore`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnReferenceStore.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnReferenceStoreProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnReferenceStore);
            }
            throw error;
        }
        cdk.requireProperty(props, 'name', this);
        this.attrArn = cdk.Token.asString(this.getAtt('Arn', cdk.ResolutionTypeHint.STRING));
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrReferenceStoreId = cdk.Token.asString(this.getAtt('ReferenceStoreId', cdk.ResolutionTypeHint.STRING));
        this.name = props.name;
        this.description = props.description;
        this.sseConfig = props.sseConfig;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::ReferenceStore", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnReferenceStore.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            name: this.name,
            description: this.description,
            sseConfig: this.sseConfig,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnReferenceStorePropsToCloudFormation(props);
    }
}
_b = JSII_RTTI_SYMBOL_1;
CfnReferenceStore[_b] = { fqn: "aws-cdk-lib.aws_omics.CfnReferenceStore", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnReferenceStore.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::ReferenceStore";
exports.CfnReferenceStore = CfnReferenceStore;
/**
 * Determine whether the given properties match those of a `SseConfigProperty`
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the result of the validation.
 */
function CfnReferenceStore_SseConfigPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('keyArn', cdk.validateString)(properties.keyArn));
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "SseConfigProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::ReferenceStore.SseConfig` resource
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::ReferenceStore.SseConfig` resource.
 */
// @ts-ignore TS6133
function cfnReferenceStoreSseConfigPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnReferenceStore_SseConfigPropertyValidator(properties).assertSuccess();
    return {
        KeyArn: cdk.stringToCloudFormation(properties.keyArn),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnReferenceStoreSseConfigPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('keyArn', 'KeyArn', properties.KeyArn != null ? cfn_parse.FromCloudFormation.getString(properties.KeyArn) : undefined);
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `CfnRunGroupProps`
 *
 * @param properties - the TypeScript properties of a `CfnRunGroupProps`
 *
 * @returns the result of the validation.
 */
function CfnRunGroupPropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('maxCpus', cdk.validateNumber)(properties.maxCpus));
    errors.collect(cdk.propertyValidator('maxDuration', cdk.validateNumber)(properties.maxDuration));
    errors.collect(cdk.propertyValidator('maxRuns', cdk.validateNumber)(properties.maxRuns));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnRunGroupProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::RunGroup` resource
 *
 * @param properties - the TypeScript properties of a `CfnRunGroupProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::RunGroup` resource.
 */
// @ts-ignore TS6133
function cfnRunGroupPropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnRunGroupPropsValidator(properties).assertSuccess();
    return {
        MaxCpus: cdk.numberToCloudFormation(properties.maxCpus),
        MaxDuration: cdk.numberToCloudFormation(properties.maxDuration),
        MaxRuns: cdk.numberToCloudFormation(properties.maxRuns),
        Name: cdk.stringToCloudFormation(properties.name),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnRunGroupPropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('maxCpus', 'MaxCpus', properties.MaxCpus != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaxCpus) : undefined);
    ret.addPropertyResult('maxDuration', 'MaxDuration', properties.MaxDuration != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaxDuration) : undefined);
    ret.addPropertyResult('maxRuns', 'MaxRuns', properties.MaxRuns != null ? cfn_parse.FromCloudFormation.getNumber(properties.MaxRuns) : undefined);
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::RunGroup`
 *
 * Creates a run group.
 *
 * @cloudformationResource AWS::Omics::RunGroup
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-rungroup.html
 */
class CfnRunGroup extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnRunGroupPropsFromCloudFormation(resourceProperties);
        const ret = new CfnRunGroup(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::RunGroup`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props = {}) {
        super(scope, id, { type: CfnRunGroup.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnRunGroupProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnRunGroup);
            }
            throw error;
        }
        this.attrArn = cdk.Token.asString(this.getAtt('Arn', cdk.ResolutionTypeHint.STRING));
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrId = cdk.Token.asString(this.getAtt('Id', cdk.ResolutionTypeHint.STRING));
        this.maxCpus = props.maxCpus;
        this.maxDuration = props.maxDuration;
        this.maxRuns = props.maxRuns;
        this.name = props.name;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::RunGroup", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnRunGroup.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            maxCpus: this.maxCpus,
            maxDuration: this.maxDuration,
            maxRuns: this.maxRuns,
            name: this.name,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnRunGroupPropsToCloudFormation(props);
    }
}
_c = JSII_RTTI_SYMBOL_1;
CfnRunGroup[_c] = { fqn: "aws-cdk-lib.aws_omics.CfnRunGroup", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnRunGroup.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::RunGroup";
exports.CfnRunGroup = CfnRunGroup;
/**
 * Determine whether the given properties match those of a `CfnSequenceStoreProps`
 *
 * @param properties - the TypeScript properties of a `CfnSequenceStoreProps`
 *
 * @returns the result of the validation.
 */
function CfnSequenceStorePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('sseConfig', CfnSequenceStore_SseConfigPropertyValidator)(properties.sseConfig));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnSequenceStoreProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::SequenceStore` resource
 *
 * @param properties - the TypeScript properties of a `CfnSequenceStoreProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::SequenceStore` resource.
 */
// @ts-ignore TS6133
function cfnSequenceStorePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnSequenceStorePropsValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Description: cdk.stringToCloudFormation(properties.description),
        SseConfig: cfnSequenceStoreSseConfigPropertyToCloudFormation(properties.sseConfig),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnSequenceStorePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', cfn_parse.FromCloudFormation.getString(properties.Name));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('sseConfig', 'SseConfig', properties.SseConfig != null ? CfnSequenceStoreSseConfigPropertyFromCloudFormation(properties.SseConfig) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::SequenceStore`
 *
 * Creates a sequence store.
 *
 * @cloudformationResource AWS::Omics::SequenceStore
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-sequencestore.html
 */
class CfnSequenceStore extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnSequenceStorePropsFromCloudFormation(resourceProperties);
        const ret = new CfnSequenceStore(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::SequenceStore`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnSequenceStore.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnSequenceStoreProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnSequenceStore);
            }
            throw error;
        }
        cdk.requireProperty(props, 'name', this);
        this.attrArn = cdk.Token.asString(this.getAtt('Arn', cdk.ResolutionTypeHint.STRING));
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrSequenceStoreId = cdk.Token.asString(this.getAtt('SequenceStoreId', cdk.ResolutionTypeHint.STRING));
        this.name = props.name;
        this.description = props.description;
        this.sseConfig = props.sseConfig;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::SequenceStore", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnSequenceStore.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            name: this.name,
            description: this.description,
            sseConfig: this.sseConfig,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnSequenceStorePropsToCloudFormation(props);
    }
}
_d = JSII_RTTI_SYMBOL_1;
CfnSequenceStore[_d] = { fqn: "aws-cdk-lib.aws_omics.CfnSequenceStore", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnSequenceStore.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::SequenceStore";
exports.CfnSequenceStore = CfnSequenceStore;
/**
 * Determine whether the given properties match those of a `SseConfigProperty`
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the result of the validation.
 */
function CfnSequenceStore_SseConfigPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('keyArn', cdk.validateString)(properties.keyArn));
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "SseConfigProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::SequenceStore.SseConfig` resource
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::SequenceStore.SseConfig` resource.
 */
// @ts-ignore TS6133
function cfnSequenceStoreSseConfigPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnSequenceStore_SseConfigPropertyValidator(properties).assertSuccess();
    return {
        KeyArn: cdk.stringToCloudFormation(properties.keyArn),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnSequenceStoreSseConfigPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('keyArn', 'KeyArn', properties.KeyArn != null ? cfn_parse.FromCloudFormation.getString(properties.KeyArn) : undefined);
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `CfnVariantStoreProps`
 *
 * @param properties - the TypeScript properties of a `CfnVariantStoreProps`
 *
 * @returns the result of the validation.
 */
function CfnVariantStorePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('reference', cdk.requiredValidator)(properties.reference));
    errors.collect(cdk.propertyValidator('reference', CfnVariantStore_ReferenceItemPropertyValidator)(properties.reference));
    errors.collect(cdk.propertyValidator('sseConfig', CfnVariantStore_SseConfigPropertyValidator)(properties.sseConfig));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnVariantStoreProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::VariantStore` resource
 *
 * @param properties - the TypeScript properties of a `CfnVariantStoreProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::VariantStore` resource.
 */
// @ts-ignore TS6133
function cfnVariantStorePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnVariantStorePropsValidator(properties).assertSuccess();
    return {
        Name: cdk.stringToCloudFormation(properties.name),
        Reference: cfnVariantStoreReferenceItemPropertyToCloudFormation(properties.reference),
        Description: cdk.stringToCloudFormation(properties.description),
        SseConfig: cfnVariantStoreSseConfigPropertyToCloudFormation(properties.sseConfig),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnVariantStorePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('name', 'Name', cfn_parse.FromCloudFormation.getString(properties.Name));
    ret.addPropertyResult('reference', 'Reference', CfnVariantStoreReferenceItemPropertyFromCloudFormation(properties.Reference));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('sseConfig', 'SseConfig', properties.SseConfig != null ? CfnVariantStoreSseConfigPropertyFromCloudFormation(properties.SseConfig) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::VariantStore`
 *
 * Create a store for variant data.
 *
 * @cloudformationResource AWS::Omics::VariantStore
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-variantstore.html
 */
class CfnVariantStore extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnVariantStorePropsFromCloudFormation(resourceProperties);
        const ret = new CfnVariantStore(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::VariantStore`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnVariantStore.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnVariantStoreProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnVariantStore);
            }
            throw error;
        }
        cdk.requireProperty(props, 'name', this);
        cdk.requireProperty(props, 'reference', this);
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrId = cdk.Token.asString(this.getAtt('Id', cdk.ResolutionTypeHint.STRING));
        this.attrStatus = cdk.Token.asString(this.getAtt('Status', cdk.ResolutionTypeHint.STRING));
        this.attrStatusMessage = cdk.Token.asString(this.getAtt('StatusMessage', cdk.ResolutionTypeHint.STRING));
        this.attrStoreArn = cdk.Token.asString(this.getAtt('StoreArn', cdk.ResolutionTypeHint.STRING));
        this.attrStoreSizeBytes = this.getAtt('StoreSizeBytes', cdk.ResolutionTypeHint.STRING);
        this.attrUpdateTime = cdk.Token.asString(this.getAtt('UpdateTime', cdk.ResolutionTypeHint.STRING));
        this.name = props.name;
        this.reference = props.reference;
        this.description = props.description;
        this.sseConfig = props.sseConfig;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::VariantStore", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnVariantStore.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            name: this.name,
            reference: this.reference,
            description: this.description,
            sseConfig: this.sseConfig,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnVariantStorePropsToCloudFormation(props);
    }
}
_e = JSII_RTTI_SYMBOL_1;
CfnVariantStore[_e] = { fqn: "aws-cdk-lib.aws_omics.CfnVariantStore", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnVariantStore.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::VariantStore";
exports.CfnVariantStore = CfnVariantStore;
/**
 * Determine whether the given properties match those of a `ReferenceItemProperty`
 *
 * @param properties - the TypeScript properties of a `ReferenceItemProperty`
 *
 * @returns the result of the validation.
 */
function CfnVariantStore_ReferenceItemPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('referenceArn', cdk.requiredValidator)(properties.referenceArn));
    errors.collect(cdk.propertyValidator('referenceArn', cdk.validateString)(properties.referenceArn));
    return errors.wrap('supplied properties not correct for "ReferenceItemProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::VariantStore.ReferenceItem` resource
 *
 * @param properties - the TypeScript properties of a `ReferenceItemProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::VariantStore.ReferenceItem` resource.
 */
// @ts-ignore TS6133
function cfnVariantStoreReferenceItemPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnVariantStore_ReferenceItemPropertyValidator(properties).assertSuccess();
    return {
        ReferenceArn: cdk.stringToCloudFormation(properties.referenceArn),
    };
}
// @ts-ignore TS6133
function CfnVariantStoreReferenceItemPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('referenceArn', 'ReferenceArn', cfn_parse.FromCloudFormation.getString(properties.ReferenceArn));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `SseConfigProperty`
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the result of the validation.
 */
function CfnVariantStore_SseConfigPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('keyArn', cdk.validateString)(properties.keyArn));
    errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
    errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
    return errors.wrap('supplied properties not correct for "SseConfigProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::VariantStore.SseConfig` resource
 *
 * @param properties - the TypeScript properties of a `SseConfigProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::VariantStore.SseConfig` resource.
 */
// @ts-ignore TS6133
function cfnVariantStoreSseConfigPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnVariantStore_SseConfigPropertyValidator(properties).assertSuccess();
    return {
        KeyArn: cdk.stringToCloudFormation(properties.keyArn),
        Type: cdk.stringToCloudFormation(properties.type),
    };
}
// @ts-ignore TS6133
function CfnVariantStoreSseConfigPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('keyArn', 'KeyArn', properties.KeyArn != null ? cfn_parse.FromCloudFormation.getString(properties.KeyArn) : undefined);
    ret.addPropertyResult('type', 'Type', cfn_parse.FromCloudFormation.getString(properties.Type));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `CfnWorkflowProps`
 *
 * @param properties - the TypeScript properties of a `CfnWorkflowProps`
 *
 * @returns the result of the validation.
 */
function CfnWorkflowPropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('definitionUri', cdk.validateString)(properties.definitionUri));
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('engine', cdk.validateString)(properties.engine));
    errors.collect(cdk.propertyValidator('main', cdk.validateString)(properties.main));
    errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
    errors.collect(cdk.propertyValidator('parameterTemplate', cdk.hashValidator(CfnWorkflow_WorkflowParameterPropertyValidator))(properties.parameterTemplate));
    errors.collect(cdk.propertyValidator('storageCapacity', cdk.validateNumber)(properties.storageCapacity));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    return errors.wrap('supplied properties not correct for "CfnWorkflowProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::Workflow` resource
 *
 * @param properties - the TypeScript properties of a `CfnWorkflowProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::Workflow` resource.
 */
// @ts-ignore TS6133
function cfnWorkflowPropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnWorkflowPropsValidator(properties).assertSuccess();
    return {
        DefinitionUri: cdk.stringToCloudFormation(properties.definitionUri),
        Description: cdk.stringToCloudFormation(properties.description),
        Engine: cdk.stringToCloudFormation(properties.engine),
        Main: cdk.stringToCloudFormation(properties.main),
        Name: cdk.stringToCloudFormation(properties.name),
        ParameterTemplate: cdk.hashMapper(cfnWorkflowWorkflowParameterPropertyToCloudFormation)(properties.parameterTemplate),
        StorageCapacity: cdk.numberToCloudFormation(properties.storageCapacity),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
    };
}
// @ts-ignore TS6133
function CfnWorkflowPropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('definitionUri', 'DefinitionUri', properties.DefinitionUri != null ? cfn_parse.FromCloudFormation.getString(properties.DefinitionUri) : undefined);
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('engine', 'Engine', properties.Engine != null ? cfn_parse.FromCloudFormation.getString(properties.Engine) : undefined);
    ret.addPropertyResult('main', 'Main', properties.Main != null ? cfn_parse.FromCloudFormation.getString(properties.Main) : undefined);
    ret.addPropertyResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addPropertyResult('parameterTemplate', 'ParameterTemplate', properties.ParameterTemplate != null ? cfn_parse.FromCloudFormation.getMap(CfnWorkflowWorkflowParameterPropertyFromCloudFormation)(properties.ParameterTemplate) : undefined);
    ret.addPropertyResult('storageCapacity', 'StorageCapacity', properties.StorageCapacity != null ? cfn_parse.FromCloudFormation.getNumber(properties.StorageCapacity) : undefined);
    ret.addPropertyResult('tags', 'Tags', properties.Tags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::Omics::Workflow`
 *
 * Creates a workflow.
 *
 * @cloudformationResource AWS::Omics::Workflow
 * @stability external
 *
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-omics-workflow.html
 */
class CfnWorkflow extends cdk.CfnResource {
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnWorkflowPropsFromCloudFormation(resourceProperties);
        const ret = new CfnWorkflow(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Create a new `AWS::Omics::Workflow`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope, id, props = {}) {
        super(scope, id, { type: CfnWorkflow.CFN_RESOURCE_TYPE_NAME, properties: props });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_omics_CfnWorkflowProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnWorkflow);
            }
            throw error;
        }
        this.attrArn = cdk.Token.asString(this.getAtt('Arn', cdk.ResolutionTypeHint.STRING));
        this.attrCreationTime = cdk.Token.asString(this.getAtt('CreationTime', cdk.ResolutionTypeHint.STRING));
        this.attrId = cdk.Token.asString(this.getAtt('Id', cdk.ResolutionTypeHint.STRING));
        this.attrStatus = cdk.Token.asString(this.getAtt('Status', cdk.ResolutionTypeHint.STRING));
        this.attrType = cdk.Token.asString(this.getAtt('Type', cdk.ResolutionTypeHint.STRING));
        this.definitionUri = props.definitionUri;
        this.description = props.description;
        this.engine = props.engine;
        this.main = props.main;
        this.name = props.name;
        this.parameterTemplate = props.parameterTemplate;
        this.storageCapacity = props.storageCapacity;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::Omics::Workflow", props.tags, { tagPropertyName: 'tags' });
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes
     *
     */
    inspect(inspector) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.inspect);
            }
            throw error;
        }
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnWorkflow.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    get cfnProperties() {
        return {
            definitionUri: this.definitionUri,
            description: this.description,
            engine: this.engine,
            main: this.main,
            name: this.name,
            parameterTemplate: this.parameterTemplate,
            storageCapacity: this.storageCapacity,
            tags: this.tags.renderTags(),
        };
    }
    renderProperties(props) {
        return cfnWorkflowPropsToCloudFormation(props);
    }
}
_f = JSII_RTTI_SYMBOL_1;
CfnWorkflow[_f] = { fqn: "aws-cdk-lib.aws_omics.CfnWorkflow", version: "2.75.1" };
/**
 * The CloudFormation resource type name for this resource class.
 */
CfnWorkflow.CFN_RESOURCE_TYPE_NAME = "AWS::Omics::Workflow";
exports.CfnWorkflow = CfnWorkflow;
/**
 * Determine whether the given properties match those of a `WorkflowParameterProperty`
 *
 * @param properties - the TypeScript properties of a `WorkflowParameterProperty`
 *
 * @returns the result of the validation.
 */
function CfnWorkflow_WorkflowParameterPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('optional', cdk.validateBoolean)(properties.optional));
    return errors.wrap('supplied properties not correct for "WorkflowParameterProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::Omics::Workflow.WorkflowParameter` resource
 *
 * @param properties - the TypeScript properties of a `WorkflowParameterProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::Omics::Workflow.WorkflowParameter` resource.
 */
// @ts-ignore TS6133
function cfnWorkflowWorkflowParameterPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnWorkflow_WorkflowParameterPropertyValidator(properties).assertSuccess();
    return {
        Description: cdk.stringToCloudFormation(properties.description),
        Optional: cdk.booleanToCloudFormation(properties.optional),
    };
}
// @ts-ignore TS6133
function CfnWorkflowWorkflowParameterPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('optional', 'Optional', properties.Optional != null ? cfn_parse.FromCloudFormation.getBoolean(properties.Optional) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
