// aws-cdk-lib/aws-route53-targets/lib/cloudfront-target.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudFrontTarget = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
/**
 * Use a CloudFront Distribution as an alias record target
 */
class CloudFrontTarget {
    /**
     * Get the hosted zone id for the current scope.
     *
     * @param scope - scope in which this resource is defined
     */
    static getHostedZoneId(scope) {
        const mappingName = 'AWSCloudFrontPartitionHostedZoneIdMap';
        const scopeStack = core_1.Stack.of(scope);
        let mapping = scopeStack.node.tryFindChild(mappingName) ??
            new core_1.CfnMapping(scopeStack, mappingName, {
                mapping: {
                    ['aws']: {
                        zoneId: 'Z2FDTNDATAQYW2', // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html
                    },
                    ['aws-cn']: {
                        zoneId: 'Z3RFFRIM2A3IF5', // https://docs.amazonaws.cn/en_us/aws/latest/userguide/route53.html
                    },
                },
            });
        return mapping.findInMap(core_1.Aws.PARTITION, 'zoneId');
    }
    constructor(distribution) {
        this.distribution = distribution;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudfront_IDistribution(distribution);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CloudFrontTarget);
            }
            throw error;
        }
    }
    bind(_record, _zone) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_route53_IRecordSet(_record);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_route53_IHostedZone(_zone);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.bind);
            }
            throw error;
        }
        return {
            hostedZoneId: CloudFrontTarget.getHostedZoneId(this.distribution),
            dnsName: this.distribution.distributionDomainName,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
CloudFrontTarget[_a] = { fqn: "aws-cdk-lib.aws_route53_targets.CloudFrontTarget", version: "2.75.1" };
/**
 * The hosted zone Id if using an alias record in Route53.
 * This value never changes.
 */
CloudFrontTarget.CLOUDFRONT_ZONE_ID = 'Z2FDTNDATAQYW2';
exports.CloudFrontTarget = CloudFrontTarget;
