// aws-cdk-lib/aws-iot/lib/iot-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoTMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class IoTMetrics {
    static topicMatchSum(dimensions) {
        return {
            namespace: 'AWS/IoT',
            metricName: 'TopicMatch',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static parseErrorSum(dimensions) {
        return {
            namespace: 'AWS/IoT',
            metricName: 'ParseError',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static ruleMessageThrottledSum(dimensions) {
        return {
            namespace: 'AWS/IoT',
            metricName: 'RuleMessageThrottled',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static ruleNotFoundSum(dimensions) {
        return {
            namespace: 'AWS/IoT',
            metricName: 'RuleNotFound',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.IoTMetrics = IoTMetrics;
