// aws-cdk-lib/aws-backup/lib/backupable-resources-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupableResourcesCollector = void 0;
const dynamodb = require("../../aws-dynamodb");
const ec2 = require("../../aws-ec2");
const efs = require("../../aws-efs");
const rds = require("../../aws-rds");
const core_1 = require("../../core");
class BackupableResourcesCollector {
    constructor() {
        this.resources = [];
    }
    visit(node) {
        if (node instanceof efs.CfnFileSystem) {
            this.resources.push(core_1.Stack.of(node).formatArn({
                service: 'elasticfilesystem',
                resource: 'file-system',
                resourceName: node.ref,
            }));
        }
        if (node instanceof dynamodb.CfnTable) {
            this.resources.push(core_1.Stack.of(node).formatArn({
                service: 'dynamodb',
                resource: 'table',
                resourceName: node.ref,
            }));
        }
        if (node instanceof ec2.CfnInstance) {
            this.resources.push(core_1.Stack.of(node).formatArn({
                service: 'ec2',
                resource: 'instance',
                resourceName: node.ref,
            }));
        }
        if (node instanceof ec2.CfnVolume) {
            this.resources.push(core_1.Stack.of(node).formatArn({
                service: 'ec2',
                resource: 'volume',
                resourceName: node.ref,
            }));
        }
        if (node instanceof rds.CfnDBInstance) {
            const dbInstance = node;
            if (!dbInstance.dbClusterIdentifier) {
                this.resources.push(core_1.Stack.of(node).formatArn({
                    service: 'rds',
                    resource: 'db',
                    arnFormat: core_1.ArnFormat.COLON_RESOURCE_NAME,
                    resourceName: node.ref,
                }));
            }
        }
        if (node instanceof rds.CfnDBCluster) {
            this.resources.push(core_1.Stack.of(node).formatArn({
                service: 'rds',
                resource: 'cluster',
                arnFormat: core_1.ArnFormat.COLON_RESOURCE_NAME,
                resourceName: node.ref,
            }));
        }
    }
}
exports.BackupableResourcesCollector = BackupableResourcesCollector;
