// aws-cdk-lib/pipelines/lib/private/javascript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDefined = exports.partition = exports.maybeSuffix = exports.noUndefined = exports.noEmptyObject = exports.mkdict = exports.mapValues = exports.filterEmpty = exports.flatten = exports.expectProp = exports.enumerate = exports.flatMap = exports.extract = exports.addAll = void 0;
function addAll(into, from) {
    for (const x of from) {
        into.add(x);
    }
}
exports.addAll = addAll;
function extract(from, key) {
    const ret = from.get(key);
    from.delete(key);
    return ret;
}
exports.extract = extract;
function* flatMap(xs, fn) {
    for (const x of xs) {
        for (const y of fn(x)) {
            yield y;
        }
    }
}
exports.flatMap = flatMap;
function* enumerate(xs) {
    let i = 0;
    for (const x of xs) {
        yield [i++, x];
    }
}
exports.enumerate = enumerate;
function expectProp(obj, key) {
    if (!obj[key]) {
        throw new Error(`Expecting '${String(key)}' to be set!`);
    }
    return obj[key];
}
exports.expectProp = expectProp;
function* flatten(xs) {
    for (const x of xs) {
        for (const y of x) {
            yield y;
        }
    }
}
exports.flatten = flatten;
function filterEmpty(xs) {
    return xs.filter(x => x);
}
exports.filterEmpty = filterEmpty;
function mapValues(xs, fn) {
    const ret = {};
    for (const [k, v] of Object.entries(xs)) {
        ret[k] = fn(v);
    }
    return ret;
}
exports.mapValues = mapValues;
function mkdict(xs) {
    const ret = {};
    for (const [k, v] of xs) {
        ret[k] = v;
    }
    return ret;
}
exports.mkdict = mkdict;
function noEmptyObject(xs) {
    if (Object.keys(xs).length === 0) {
        return undefined;
    }
    return xs;
}
exports.noEmptyObject = noEmptyObject;
function noUndefined(xs) {
    return mkdict(Object.entries(xs).filter(([_, v]) => isDefined(v)));
}
exports.noUndefined = noUndefined;
function maybeSuffix(x, suffix) {
    if (x === undefined) {
        return undefined;
    }
    return `${x}${suffix}`;
}
exports.maybeSuffix = maybeSuffix;
/**
 * Partition a collection by dividing it into two collections, one that matches the predicate and one that don't
 */
function partition(xs, pred) {
    const yes = [];
    const no = [];
    for (const x of xs) {
        (pred(x) ? yes : no).push(x);
    }
    return [yes, no];
}
exports.partition = partition;
function isDefined(x) {
    return x !== undefined;
}
exports.isDefined = isDefined;
