// aws-cdk-lib/core/lib/validation/private/construct-tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructTree = void 0;
const app_1 = require("../../app");
const cfn_resource_1 = require("../../cfn-resource");
const stack_1 = require("../../stack");
/**
 * Utility class to help accessing information on constructs in the
 * construct tree. This can be created once and shared between
 * all the validation plugin executions.
 */
class ConstructTree {
    constructor(root) {
        this.root = root;
        /**
         * A cache of the ConstructTrace by node.path. Each construct
         */
        this._traceCache = new Map();
        this._constructByPath = new Map();
        this._constructByTemplatePathAndLogicalId = new Map();
        if (app_1.App.isApp(this.root)) {
            this.treeMetadata = this.root.node.tryFindChild('Tree');
        }
        else {
            this.treeMetadata = app_1.App.of(this.root)?.node.tryFindChild('Tree');
        }
        this._constructByPath.set(this.root.node.path, root);
        // do this once at the start so we don't have to traverse
        // the entire tree everytime we want to find a nested node
        this.root.node.findAll().forEach(child => {
            this._constructByPath.set(child.node.path, child);
            const defaultChild = child.node.defaultChild;
            if (defaultChild && cfn_resource_1.CfnResource.isCfnResource(defaultChild)) {
                const stack = stack_1.Stack.of(defaultChild);
                const logicalId = stack.resolve(defaultChild.logicalId);
                this.setLogicalId(stack, logicalId, child);
            }
        });
        // Another pass to include all the L1s that haven't been added yet
        this.root.node.findAll().forEach(child => {
            if (cfn_resource_1.CfnResource.isCfnResource(child)) {
                const stack = stack_1.Stack.of(child);
                const logicalId = stack_1.Stack.of(child).resolve(child.logicalId);
                this.setLogicalId(stack, logicalId, child);
            }
        });
    }
    setLogicalId(stack, logicalId, child) {
        if (!this._constructByTemplatePathAndLogicalId.has(stack.templateFile)) {
            this._constructByTemplatePathAndLogicalId.set(stack.templateFile, new Map([[logicalId, child]]));
        }
        else {
            this._constructByTemplatePathAndLogicalId.get(stack.templateFile)?.set(logicalId, child);
        }
    }
    /**
     * Get the stack trace from the construct node metadata.
     * The stack trace only gets recorded if the node is a `CfnResource`,
     * but the stack trace will have entries for all types of parent construct
     * scopes
     */
    getTraceMetadata(size, node) {
        if (node) {
            const construct = this.getConstructByPath(node.path);
            if (construct) {
                let trace;
                if (cfn_resource_1.CfnResource.isCfnResource(construct)) {
                    trace = construct.node.metadata.find(meta => !!meta.trace)?.trace ?? [];
                }
                else {
                    trace = construct.node.defaultChild?.node.metadata.find(meta => !!meta.trace)?.trace ?? [];
                }
                // take just the items we need and reverse it since we are
                // displaying the trace bottom up
                return Object.create(trace.slice(0, size));
            }
        }
        return [];
    }
    /**
     * Only the `CfnResource` constructs contain the trace information
     * So we need to go down the tree and find that resource (its always the last one)
     *
     * @param node Node the entire tree where the bottom is the violating resource
     * @return Node the bottom of the tree which will be the violating resource
     */
    getNodeWithTrace(node) {
        if (node.children) {
            return this.getNodeWithTrace(this.getChild(node.children));
        }
        return node;
    }
    /**
     * Get a ConstructTrace from the cache for a given construct
     *
     * Construct the stack trace of constructs. This will start with the
     * root of the tree and go down to the construct that has the violation
     */
    getTrace(node, locations) {
        const trace = this._traceCache.get(node.path);
        if (trace) {
            return trace;
        }
        const size = this.nodeSize(node);
        const nodeWithTrace = this.getNodeWithTrace(node);
        const metadata = (locations ?? this.getTraceMetadata(size, nodeWithTrace));
        const thisLocation = metadata.pop();
        const constructTrace = {
            id: node.id,
            path: node.path,
            // the "child" trace will be the "parent" node
            // since we are going bottom up
            child: node.children
                ? this.getTrace(this.getChild(node.children), metadata)
                : undefined,
            construct: node.constructInfo?.fqn,
            libraryVersion: node.constructInfo?.version,
            location: thisLocation ?? "Run with '--debug' to include location info",
        };
        this._traceCache.set(constructTrace.path, constructTrace);
        return constructTrace;
    }
    /**
     * Each node will only have a single child so just
     * return that
     */
    getChild(children) {
        return Object.values(children)[0];
    }
    /**
     * Get the size of a Node, i.e. how many levels is it
     */
    nodeSize(node) {
        let size = 1;
        if (!node.children) {
            return size;
        }
        let children = this.getChild(node.children);
        do {
            size++;
            children = children.children
                ? this.getChild(children.children)
                : undefined;
        } while (children);
        return size;
    }
    /**
     * Get a specific node in the tree by construct path
     *
     * @param path the construct path of the node to return
     * @returns the TreeMetadata Node
     */
    getTreeNode(path) {
        return this.treeMetadata._getNodeBranch(path);
    }
    /**
     * Get a specific Construct by the node.addr
     *
     * @param path the node.addr of the construct
     * @returns the Construct
     */
    getConstructByPath(path) {
        return this._constructByPath.get(path);
    }
    /**
     * Get a specific Construct by the CfnResource logical ID. This will
     * be the construct.node.defaultChild with the given ID
     *
     * @param logicalId the ID of the CfnResource
     * @returns the Construct
     */
    getConstructByLogicalId(templateFile, logicalId) {
        return this._constructByTemplatePathAndLogicalId.get(templateFile)?.get(logicalId);
    }
}
exports.ConstructTree = ConstructTree;
