// aws-cdk-lib/aws-ec2/lib/placement-group.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlacementGroup = exports.PlacementGroupStrategy = exports.PlacementGroupSpreadLevel = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const ec2_generated_1 = require("./ec2.generated");
/**
 * Determines how this placement group spreads instances
 */
var PlacementGroupSpreadLevel;
(function (PlacementGroupSpreadLevel) {
    /**
     * Host spread level placement groups are only available with AWS Outposts.
     * For host spread level placement groups, there are no restrictions for running instances per Outposts.
     *
     * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/placement-groups-outpost.html
     */
    PlacementGroupSpreadLevel["HOST"] = "host";
    /**
     * Each instance is launched on a separate rack.
     * Each has its own network and power source.
     * A rack spread placement group can span multiple Availability Zones in the same Region.
     * For rack spread level placement groups, you can have a maximum of seven running instances per Availability Zone per group.
     */
    PlacementGroupSpreadLevel["RACK"] = "rack";
})(PlacementGroupSpreadLevel = exports.PlacementGroupSpreadLevel || (exports.PlacementGroupSpreadLevel = {}));
/**
 * Which strategy to use when launching instances
 */
var PlacementGroupStrategy;
(function (PlacementGroupStrategy) {
    /**
     * Packs instances close together inside an Availability Zone.
     * This strategy enables workloads to achieve the low-latency network
     * performance necessary for tightly-coupled node-to-node communication that
     * is typical of high-performance computing (HPC) applications.
     */
    PlacementGroupStrategy["CLUSTER"] = "cluster";
    /**
     * Spreads your instances across logical partitions such that groups of instances
     * in one partition do not share the underlying hardware with groups of instances
     * in different partitions.
     *
     * This strategy is typically used by large distributed and replicated workloads,
     * such as Hadoop, Cassandra, and Kafka.
     */
    PlacementGroupStrategy["PARTITION"] = "partition";
    /**
     * Strictly places a small group of instances across distinct underlying hardware
     * to reduce correlated failures.
     */
    PlacementGroupStrategy["SPREAD"] = "spread";
})(PlacementGroupStrategy = exports.PlacementGroupStrategy || (exports.PlacementGroupStrategy = {}));
/**
 * Defines a placement group. Placement groups give you fine-grained control over
 * where your instances are provisioned.
 */
class PlacementGroup extends core_1.Resource {
    /**
     * Import a PlacementGroup by its arn
     */
    static fromPlacementGroupName(scope, id, placementGroupName) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.placementGroupName = placementGroupName;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props) {
        super(scope, id, {
            physicalName: undefined,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_PlacementGroupProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, PlacementGroup);
            }
            throw error;
        }
        this.partitions = props?.partitions;
        this.spreadLevel = props?.spreadLevel;
        this.strategy = props?.strategy;
        if (this.partitions && this.strategy) {
            if (this.strategy !== PlacementGroupStrategy.PARTITION) {
                throw new Error(`PlacementGroup '${id}' can only specify 'partitions' with the 'PARTITION' strategy`);
            }
        }
        else if (this.partitions && !this.strategy) {
            this.strategy = PlacementGroupStrategy.PARTITION;
        }
        if (this.spreadLevel) {
            if (!this.strategy) {
                this.strategy = PlacementGroupStrategy.SPREAD;
            }
            if (this.strategy !== PlacementGroupStrategy.SPREAD) {
                throw new Error(`PlacementGroup '${id}' can only specify 'spreadLevel' with the 'SPREAD' strategy`);
            }
        }
        const resource = new ec2_generated_1.CfnPlacementGroup(this, 'Resource', {
            partitionCount: this.partitions,
            spreadLevel: this.spreadLevel,
            strategy: this.strategy,
        });
        this.placementGroupName = this.getResourceArnAttribute(resource.attrGroupName, {
            service: 'batch',
            resource: 'compute-environment',
            resourceName: this.physicalName,
        });
    }
}
_a = JSII_RTTI_SYMBOL_1;
PlacementGroup[_a] = { fqn: "aws-cdk-lib.aws_ec2.PlacementGroup", version: "2.75.1" };
exports.PlacementGroup = PlacementGroup;
