// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/alb.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationLoadBalancerEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _util_1 = require("./_util");
/**
 * Use an Application Load Balancer as a Global Accelerator Endpoint
 */
class ApplicationLoadBalancerEndpoint {
    constructor(loadBalancer, options = {}) {
        this.loadBalancer = loadBalancer;
        this.options = options;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_IApplicationLoadBalancer(loadBalancer);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_ApplicationLoadBalancerEndpointOptions(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, ApplicationLoadBalancerEndpoint);
            }
            throw error;
        }
        (0, _util_1.validateWeight)(options.weight);
        this.region = loadBalancer.env.region;
    }
    renderEndpointConfiguration() {
        return {
            endpointId: this.loadBalancer.loadBalancerArn,
            weight: this.options.weight,
            clientIpPreservationEnabled: this.options.preserveClientIp,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
ApplicationLoadBalancerEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.ApplicationLoadBalancerEndpoint", version: "2.75.1" };
exports.ApplicationLoadBalancerEndpoint = ApplicationLoadBalancerEndpoint;
