// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/eip.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnEipEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const _util_1 = require("./_util");
/**
 * Use an EC2 Instance as a Global Accelerator Endpoint
 */
class CfnEipEndpoint {
    constructor(eip, options = {}) {
        this.eip = eip;
        this.options = options;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_CfnEIP(eip);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_CfnEipEndpointProps(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, CfnEipEndpoint);
            }
            throw error;
        }
        (0, _util_1.validateWeight)(options.weight);
        this.region = core_1.Stack.of(eip).region;
    }
    renderEndpointConfiguration() {
        return {
            endpointId: this.eip.attrAllocationId,
            weight: this.options.weight,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
CfnEipEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.CfnEipEndpoint", version: "2.75.1" };
exports.CfnEipEndpoint = CfnEipEndpoint;
