// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/instance.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _util_1 = require("./_util");
/**
 * Use an EC2 Instance as a Global Accelerator Endpoint
 */
class InstanceEndpoint {
    constructor(instance, options = {}) {
        this.instance = instance;
        this.options = options;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_IInstance(instance);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_InstanceEndpointProps(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, InstanceEndpoint);
            }
            throw error;
        }
        (0, _util_1.validateWeight)(options.weight);
        this.region = instance.env.region;
    }
    renderEndpointConfiguration() {
        return {
            endpointId: this.instance.instanceId,
            weight: this.options.weight,
            clientIpPreservationEnabled: this.options.preserveClientIp,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
InstanceEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.InstanceEndpoint", version: "2.75.1" };
exports.InstanceEndpoint = InstanceEndpoint;
