// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/nlb.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkLoadBalancerEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _util_1 = require("./_util");
/**
 * Use a Network Load Balancer as a Global Accelerator Endpoint
 */
class NetworkLoadBalancerEndpoint {
    constructor(loadBalancer, options = {}) {
        this.loadBalancer = loadBalancer;
        this.options = options;
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_INetworkLoadBalancer(loadBalancer);
            jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_NetworkLoadBalancerEndpointProps(options);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, NetworkLoadBalancerEndpoint);
            }
            throw error;
        }
        (0, _util_1.validateWeight)(options.weight);
        this.region = loadBalancer.env.region;
    }
    renderEndpointConfiguration() {
        return {
            endpointId: this.loadBalancer.loadBalancerArn,
            weight: this.options.weight,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
NetworkLoadBalancerEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.NetworkLoadBalancerEndpoint", version: "2.75.1" };
exports.NetworkLoadBalancerEndpoint = NetworkLoadBalancerEndpoint;
