// aws-cdk-lib/aws-appmesh/lib/http-route-method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRouteMethod = void 0;
/**
 * Supported values for matching routes based on the HTTP request method
 */
var HttpRouteMethod;
(function (HttpRouteMethod) {
    /**
     * GET request
     */
    HttpRouteMethod["GET"] = "GET";
    /**
     * HEAD request
     */
    HttpRouteMethod["HEAD"] = "HEAD";
    /**
     * POST request
     */
    HttpRouteMethod["POST"] = "POST";
    /**
     * PUT request
     */
    HttpRouteMethod["PUT"] = "PUT";
    /**
     * DELETE request
     */
    HttpRouteMethod["DELETE"] = "DELETE";
    /**
     * CONNECT request
     */
    HttpRouteMethod["CONNECT"] = "CONNECT";
    /**
     * OPTIONS request
     */
    HttpRouteMethod["OPTIONS"] = "OPTIONS";
    /**
     * TRACE request
     */
    HttpRouteMethod["TRACE"] = "TRACE";
    /**
     * PATCH request
     */
    HttpRouteMethod["PATCH"] = "PATCH";
})(HttpRouteMethod = exports.HttpRouteMethod || (exports.HttpRouteMethod = {}));
