// aws-cdk-lib/aws-appmesh/lib/tls-certificate.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MutualTlsCertificate = exports.TlsCertificate = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Represents a TLS certificate
 */
class TlsCertificate {
    /**
     * Returns an File TLS Certificate
     */
    static file(certificateChainPath, privateKeyPath) {
        return new FileTlsCertificate(certificateChainPath, privateKeyPath);
    }
    /**
     * Returns an ACM TLS Certificate
     */
    static acm(certificate) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_certificatemanager_ICertificate(certificate);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.acm);
            }
            throw error;
        }
        return new AcmTlsCertificate(certificate);
    }
    /**
     * Returns an SDS TLS Certificate
     */
    static sds(secretName) {
        return new SdsTlsCertificate(secretName);
    }
}
_a = JSII_RTTI_SYMBOL_1;
TlsCertificate[_a] = { fqn: "aws-cdk-lib.aws_appmesh.TlsCertificate", version: "2.75.1" };
exports.TlsCertificate = TlsCertificate;
/**
 * Represents a TLS certificate that is supported for mutual TLS authentication.
 */
class MutualTlsCertificate extends TlsCertificate {
    constructor() {
        super(...arguments);
        // TypeScript uses structural typing, so we need a property different from TlsCertificate
        this.differentiator = false;
    }
}
_b = JSII_RTTI_SYMBOL_1;
MutualTlsCertificate[_b] = { fqn: "aws-cdk-lib.aws_appmesh.MutualTlsCertificate", version: "2.75.1" };
exports.MutualTlsCertificate = MutualTlsCertificate;
/**
 * Represents a ACM provided TLS certificate
 */
class AcmTlsCertificate extends TlsCertificate {
    constructor(certificate) {
        super();
        this.acmCertificate = certificate;
    }
    bind(_scope) {
        return {
            tlsCertificate: {
                acm: {
                    certificateArn: this.acmCertificate.certificateArn,
                },
            },
        };
    }
}
/**
 * Represents a file provided TLS certificate
 */
class FileTlsCertificate extends MutualTlsCertificate {
    constructor(certificateChainPath, privateKeyPath) {
        super();
        this.certificateChain = certificateChainPath;
        this.privateKey = privateKeyPath;
    }
    bind(_scope) {
        return {
            tlsCertificate: {
                file: {
                    certificateChain: this.certificateChain,
                    privateKey: this.privateKey,
                },
            },
        };
    }
}
/**
 * Represents a SDS provided TLS certificate
 */
class SdsTlsCertificate extends MutualTlsCertificate {
    constructor(secretName) {
        super();
        this.secretName = secretName;
    }
    bind(_scope) {
        return {
            tlsCertificate: {
                sds: {
                    secretName: this.secretName,
                },
            },
        };
    }
}
