// aws-cdk-lib/aws-events/lib/event-pattern.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Match = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
/**
 * An event pattern matcher
 */
class Match {
    /**
     * Matches a null value in the JSON of the event
     */
    static nullValue() {
        return this.fromObjects([null]);
    }
    /**
     * Matches when the field is present in the JSON of the event
     */
    static exists() {
        return this.fromObjects([{ exists: true }]);
    }
    /**
     * Matches when the field is absent from the JSON of the event
     */
    static doesNotExist() {
        return this.fromObjects([{ exists: false }]);
    }
    /**
     * Matches a string, exactly, in the JSON of the event
     */
    static exactString(value) {
        return this.fromObjects([value]);
    }
    /**
     * Matches a string, regardless of case, in the JSON of the event
     */
    static equalsIgnoreCase(value) {
        return this.fromObjects([{ 'equals-ignore-case': value }]);
    }
    /**
     * Matches strings with the given prefix in the JSON of the event
     */
    static prefix(value) {
        return this.fromObjects([{ prefix: value }]);
    }
    /**
     * Matches strings with the given suffix in the JSON of the event
     */
    static suffix(value) {
        return this.fromObjects([{ suffix: value }]);
    }
    /**
     * Matches IPv4 and IPv6 network addresses using the Classless Inter-Domain Routing (CIDR) format
     */
    static cidr(range) {
        const ipv4Regex = /^([0-9]{1,3}\.){3}[0-9]{1,3}(\/([0-9]|[1-2][0-9]|3[0-2]))?$/igm;
        const ipv6Regex = /^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(\/([0-9]|[1-9][0-9]|1[0-1][0-9]|12[0-8]))?$/igm;
        if (!ipv4Regex.test(range) && !ipv6Regex.test(range)) {
            throw new Error(`Invalid IP address range: ${range}`);
        }
        return this.fromObjects([{ cidr: range }]);
    }
    /**
     * Matches IPv4 and IPv6 network addresses using the Classless Inter-Domain Routing (CIDR) format.
     * Alias of `cidr()`.
     */
    static ipAddressRange(range) {
        return Match.cidr(range);
    }
    /**
     * Matches anything except what's provided in the rule. The list of provided values must contain
     * only strings or only numbers.
     */
    static anythingBut(...values) {
        if (values.length === 0) {
            throw new Error('anythingBut matchers must be non-empty lists');
        }
        const allNumbers = values.every(v => typeof (v) === 'number');
        const allStrings = values.every(v => typeof (v) === 'string');
        if (!(allNumbers || allStrings)) {
            throw new Error('anythingBut matchers must be lists that contain only strings or only numbers.');
        }
        return this.fromObjects([{ 'anything-but': values }]);
    }
    /**
     * Matches any string that doesn't start with the given prefix.
     */
    static anythingButPrefix(prefix) {
        return this.fromObjects([{ 'anything-but': { prefix: prefix } }]);
    }
    /**
     * Matches numbers greater than the provided value
     */
    static greaterThan(value) {
        return this.numeric('>', value);
    }
    /**
     * Matches numbers greater than, or equal to, the provided value
     */
    static greaterThanOrEqual(value) {
        return this.numeric('>=', value);
    }
    /**
     * Matches numbers less than the provided value
     */
    static lessThan(value) {
        return this.numeric('<', value);
    }
    /**
     * Matches numbers less than, or equal to, the provided value
     */
    static lessThanOrEqual(value) {
        return this.numeric('<=', value);
    }
    /**
     * Matches numbers equal to the provided value
     */
    static equal(value) {
        return this.numeric('=', value);
    }
    /**
     * Matches numbers inside a closed numeric interval. Equivalent to:
     *
     *    Match.allOf(Match.greaterThanOrEqual(lower), Match.lessThanOrEqual(upper))
     *
     * @param lower Lower bound (inclusive)
     * @param upper Upper bound (inclusive)
     */
    static interval(lower, upper) {
        if (lower > upper) {
            throw new Error(`Invalid interval: [${lower}, ${upper}]`);
        }
        return Match.allOf(Match.greaterThanOrEqual(lower), Match.lessThanOrEqual(upper));
    }
    /**
     * Matches an event if any of the provided matchers do. Only numeric matchers are accepted.
     */
    static allOf(...matchers) {
        if (matchers.length === 0) {
            throw new Error('A list of matchers must contain at least one element.');
        }
        return this.fromMergedObjects(matchers);
    }
    /**
     * Matches an event if any of the provided matchers does.
     */
    static anyOf(...matchers) {
        if (matchers.length === 0) {
            throw new Error('A list of matchers must contain at least one element.');
        }
        return this.fromObjects(matchers);
    }
    static numeric(operator, value) {
        return this.fromObjects([{ numeric: [operator, value] }]);
    }
    static fromObjects(values) {
        return new Match(values, { mergeMatchers: false }).asList();
    }
    static fromMergedObjects(values) {
        return new Match(values, { mergeMatchers: true }).asList();
    }
    constructor(matchers, options) {
        this.matchers = matchers;
        this.options = options;
        this.creationStack = (0, core_1.captureStackTrace)();
    }
    resolve(context) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_IResolveContext(context);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.resolve);
            }
            throw error;
        }
        const matchers = this.matchers.flatMap(matcher => context.resolve(matcher));
        return this.options.mergeMatchers ? this.merge(matchers) : matchers;
    }
    merge(matchers) {
        // This is the only supported case for merging at the moment.
        // We can generalize this logic if EventBridge starts supporting more cases in the future.
        if (!matchers.every(matcher => matcher?.numeric)) {
            throw new Error('Only numeric matchers can be merged into a single matcher.');
        }
        return [{ numeric: matchers.flatMap(matcher => matcher.numeric) }];
    }
    toString() {
        return core_1.Token.asString(this);
    }
    /**
     * A representation of this matcher as a list of strings
     */
    asList() {
        return core_1.Token.asList(this);
    }
}
_a = JSII_RTTI_SYMBOL_1;
Match[_a] = { fqn: "aws-cdk-lib.aws_events.Match", version: "2.75.1" };
exports.Match = Match;
