// aws-cdk-lib/aws-ses/lib/dedicated-ip-pool.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DedicatedIpPool = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const ses_generated_1 = require("./ses.generated");
/**
 * A dedicated IP pool
 */
class DedicatedIpPool extends core_1.Resource {
    /**
     * Use an existing dedicated IP pool
     */
    static fromDedicatedIpPoolName(scope, id, dedicatedIpPoolName) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.dedicatedIpPoolName = dedicatedIpPoolName;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props = {}) {
        super(scope, id, {
            physicalName: props.dedicatedIpPoolName,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ses_DedicatedIpPoolProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, DedicatedIpPool);
            }
            throw error;
        }
        const pool = new ses_generated_1.CfnDedicatedIpPool(this, 'Resource', {
            poolName: this.physicalName,
        });
        this.dedicatedIpPoolName = pool.ref;
    }
}
_a = JSII_RTTI_SYMBOL_1;
DedicatedIpPool[_a] = { fqn: "aws-cdk-lib.aws_ses.DedicatedIpPool", version: "2.75.1" };
exports.DedicatedIpPool = DedicatedIpPool;
