// aws-cdk-lib/aws-ses/lib/vdm-attributes.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VdmAttributes = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const ses_generated_1 = require("./ses.generated");
/**
 * Virtual Deliverablity Manager (VDM) attributes
 */
class VdmAttributes extends core_1.Resource {
    /**
     * Use an existing Virtual Deliverablity Manager attributes resource
     */
    static fromVdmAttributesName(scope, id, vdmAttributesName) {
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.vdmAttributesName = vdmAttributesName;
            }
        }
        return new Import(scope, id);
    }
    constructor(scope, id, props = {}) {
        super(scope, id);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ses_VdmAttributesProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, VdmAttributes);
            }
            throw error;
        }
        const resource = new ses_generated_1.CfnVdmAttributes(this, 'Resource', {
            dashboardAttributes: {
                engagementMetrics: booleanToEnabledDisabled(props.engagementMetrics ?? true),
            },
            guardianAttributes: {
                optimizedSharedDelivery: booleanToEnabledDisabled(props.optimizedSharedDelivery ?? true),
            },
        });
        this.vdmAttributesName = resource.ref;
        this.vdmAttributesResourceId = resource.attrVdmAttributesResourceId;
    }
}
_a = JSII_RTTI_SYMBOL_1;
VdmAttributes[_a] = { fqn: "aws-cdk-lib.aws_ses.VdmAttributes", version: "2.75.1" };
exports.VdmAttributes = VdmAttributes;
function booleanToEnabledDisabled(value) {
    return value === true
        ? 'ENABLED'
        : 'DISABLED';
}
