// aws-cdk-lib/aws-servicecatalog/lib/private/product-stack-synthesizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductStackSynthesizer = void 0;
const path = require("path");
const aws_s3_deployment_1 = require("../../../aws-s3-deployment");
const cdk = require("../../../core");
/**
 * Deployment environment for an AWS Service Catalog product stack.
 *
 * Interoperates with the StackSynthesizer of the parent stack.
 */
class ProductStackSynthesizer extends cdk.StackSynthesizer {
    constructor(assetBucket) {
        super();
        this.assetBucket = assetBucket;
    }
    addFileAsset(asset) {
        if (!this.assetBucket) {
            throw new Error('An Asset Bucket must be provided to use Assets');
        }
        const outdir = cdk.App.of(this.boundStack)?.outdir ?? 'cdk.out';
        const assetPath = `${outdir}/${asset.fileName}`;
        if (!this.bucketDeployment) {
            const parentStack = this.boundStack._getParentStack();
            if (!cdk.Resource.isOwnedResource(this.assetBucket)) {
                cdk.Annotations.of(parentStack).addWarning('[WARNING] Bucket Policy Permissions cannot be added to' +
                    ' referenced Bucket. Please make sure your bucket has the correct permissions');
            }
            this.bucketDeployment = new aws_s3_deployment_1.BucketDeployment(parentStack, 'AssetsBucketDeployment', {
                sources: [aws_s3_deployment_1.Source.asset(assetPath)],
                destinationBucket: this.assetBucket,
                extract: false,
                prune: false,
            });
        }
        else {
            this.bucketDeployment.addSource(aws_s3_deployment_1.Source.asset(assetPath));
        }
        const physicalName = this.physicalNameOfBucket(this.assetBucket);
        const bucketName = physicalName;
        if (!asset.fileName) {
            throw new Error('Asset file name is undefined');
        }
        const assetFileBaseName = path.basename(asset.fileName);
        const s3Filename = assetFileBaseName.split('.')[1] + '.zip';
        const objectKey = `${s3Filename}`;
        const s3ObjectUrl = `s3://${bucketName}/${objectKey}`;
        const httpUrl = `https://s3.${bucketName}/${objectKey}`;
        return { bucketName, objectKey, httpUrl, s3ObjectUrl, s3Url: httpUrl };
    }
    physicalNameOfBucket(bucket) {
        let resolvedName;
        if (cdk.Resource.isOwnedResource(bucket)) {
            resolvedName = cdk.Stack.of(bucket).resolve(bucket.node.defaultChild.bucketName);
        }
        else {
            resolvedName = bucket.bucketName;
        }
        if (resolvedName === undefined) {
            throw new Error('A bucketName must be provided to use Assets');
        }
        return resolvedName;
    }
    addDockerImageAsset(_asset) {
        throw new Error('Service Catalog Product Stacks cannot use Assets');
    }
    synthesize(session) {
        // Synthesize the template, but don't emit as a cloud assembly artifact.
        // It will be registered as an S3 asset of its parent instead.
        this.synthesizeTemplate(session);
    }
}
exports.ProductStackSynthesizer = ProductStackSynthesizer;
