// aws-cdk-lib/aws-servicecatalog/lib/private/validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputValidator = void 0;
const cdk = require("../../../core");
/**
 * Class to validate that inputs match requirements.
 */
class InputValidator {
    /**
     * Validates length is between allowed min and max lengths.
     */
    static validateLength(resourceName, inputName, minLength, maxLength, inputString) {
        if (!cdk.Token.isUnresolved(inputString) && inputString !== undefined && (inputString.length < minLength || inputString.length > maxLength)) {
            throw new Error(`Invalid ${inputName} for resource ${resourceName}, must have length between ${minLength} and ${maxLength}, got: '${this.truncateString(inputString, 100)}'`);
        }
    }
    /**
     * Validates string matches the allowed regex pattern.
     */
    static validateRegex(resourceName, inputName, regexp, inputString) {
        if (!cdk.Token.isUnresolved(inputString) && inputString !== undefined && !regexp.test(inputString)) {
            throw new Error(`Invalid ${inputName} for resource ${resourceName}, must match regex pattern ${regexp}, got: '${this.truncateString(inputString, 100)}'`);
        }
    }
    /**
     * Validates string matches the valid URL regex pattern.
     */
    static validateUrl(resourceName, inputName, inputString) {
        this.validateRegex(resourceName, inputName, /^https?:\/\/.*/, inputString);
    }
    /**
    * Validates string matches the valid email regex pattern.
    */
    static validateEmail(resourceName, inputName, inputString) {
        this.validateRegex(resourceName, inputName, /^[\w\d.%+\-]+@[a-z\d.\-]+\.[a-z]{2,4}$/i, inputString);
    }
    /**
    * Validates that a role being used as a local launch role has the role name set
    */
    static validateRoleNameSetForLocalLaunchRole(role) {
        if (role.node.defaultChild) {
            if (cdk.Token.isUnresolved(role.node.defaultChild.roleName)) {
                throw new Error(`Role ${role.node.id} used for Local Launch Role must have roleName explicitly set`);
            }
        }
    }
    static truncateString(string, maxLength) {
        if (string.length > maxLength) {
            return string.substring(0, maxLength) + '[truncated]';
        }
        return string;
    }
}
exports.InputValidator = InputValidator;
