// aws-cdk-lib/aws-route53resolver/lib/route53resolver-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53ResolverMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class Route53ResolverMetrics {
    static inboundQueryVolumeSum(dimensions) {
        return {
            namespace: 'AWS/Route53Resolver',
            metricName: 'InboundQueryVolume',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static outboundQueryVolumeSum(dimensions) {
        return {
            namespace: 'AWS/Route53Resolver',
            metricName: 'OutboundQueryVolume',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static outboundQueryAggregateVolumeSum(dimensions) {
        return {
            namespace: 'AWS/Route53Resolver',
            metricName: 'OutboundQueryAggregateVolume',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.Route53ResolverMetrics = Route53ResolverMetrics;
