// aws-cdk-lib/aws-appmesh/lib/private/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateGrpcGatewayRouteMatch = exports.validateGrpcRouteMatch = exports.validateGrpcMatchArrayLength = exports.validateHttpMatchArrayLength = exports.renderMeshOwner = exports.renderListenerTlsOptions = exports.renderTlsClientPolicy = void 0;
const core_1 = require("../../../core");
/**
 * This is the helper method to render TLS property of client policy.
 */
function renderTlsClientPolicy(scope, tlsClientPolicy) {
    const certificate = tlsClientPolicy?.mutualTlsCertificate?.bind(scope).tlsCertificate;
    const sans = tlsClientPolicy?.validation.subjectAlternativeNames;
    return tlsClientPolicy
        ? {
            certificate: certificate,
            ports: tlsClientPolicy.ports,
            enforce: tlsClientPolicy.enforce,
            validation: {
                subjectAlternativeNames: sans
                    ? {
                        match: sans.bind(scope).subjectAlternativeNamesMatch,
                    }
                    : undefined,
                trust: tlsClientPolicy.validation.trust.bind(scope).tlsValidationTrust,
            },
        }
        : undefined;
}
exports.renderTlsClientPolicy = renderTlsClientPolicy;
/**
 * This is the helper method to render the TLS config for a listener.
 */
function renderListenerTlsOptions(scope, listenerTls) {
    const tlsValidation = listenerTls?.mutualTlsValidation;
    return listenerTls
        ? {
            certificate: listenerTls.certificate.bind(scope).tlsCertificate,
            mode: listenerTls.mode,
            validation: tlsValidation
                ? {
                    subjectAlternativeNames: tlsValidation.subjectAlternativeNames
                        ? {
                            match: tlsValidation.subjectAlternativeNames.bind(scope).subjectAlternativeNamesMatch,
                        }
                        : undefined,
                    trust: tlsValidation.trust.bind(scope).tlsValidationTrust,
                }
                : undefined,
        }
        : undefined;
}
exports.renderListenerTlsOptions = renderListenerTlsOptions;
/**
 * This is the helper method to populate mesh owner when it is a shared mesh scenario
 */
function renderMeshOwner(resourceAccount, meshAccount) {
    const comparison = core_1.Token.compareStrings(resourceAccount, meshAccount);
    return comparison === core_1.TokenComparison.DIFFERENT || comparison === core_1.TokenComparison.ONE_UNRESOLVED
        ? meshAccount
        : undefined;
}
exports.renderMeshOwner = renderMeshOwner;
/**
 * This is the helper method to validate the length of HTTP match array when it is specified.
 */
function validateHttpMatchArrayLength(headers, queryParameters) {
    const MIN_LENGTH = 1;
    const MAX_LENGTH = 10;
    if (headers && (headers.length < MIN_LENGTH || headers.length > MAX_LENGTH)) {
        throw new Error(`Number of headers provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${headers.length}`);
    }
    if (queryParameters && (queryParameters.length < MIN_LENGTH || queryParameters.length > MAX_LENGTH)) {
        throw new Error(`Number of query parameters provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${queryParameters.length}`);
    }
}
exports.validateHttpMatchArrayLength = validateHttpMatchArrayLength;
/**
 * This is the helper method to validate the length of gRPC match array when it is specified.
 */
function validateGrpcMatchArrayLength(metadata) {
    const MIN_LENGTH = 1;
    const MAX_LENGTH = 10;
    if (metadata && (metadata.length < MIN_LENGTH || metadata.length > MAX_LENGTH)) {
        throw new Error(`Number of metadata provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${metadata.length}`);
    }
}
exports.validateGrpcMatchArrayLength = validateGrpcMatchArrayLength;
/**
 * This is the helper method to validate at least one of gRPC route match type is defined.
 */
function validateGrpcRouteMatch(match) {
    if (match.serviceName === undefined && match.metadata === undefined && match.methodName === undefined) {
        throw new Error('At least one gRPC route match property must be provided');
    }
}
exports.validateGrpcRouteMatch = validateGrpcRouteMatch;
/**
 * This is the helper method to validate at least one of gRPC gateway route match type is defined.
 */
function validateGrpcGatewayRouteMatch(match) {
    if (match.serviceName === undefined && match.metadata === undefined && match.hostname === undefined) {
        throw new Error('At least one gRPC gateway route match property beside rewriteRequestHostname must be provided');
    }
}
exports.validateGrpcGatewayRouteMatch = validateGrpcGatewayRouteMatch;
