// aws-cdk-lib/aws-appsync/lib/appsync-canned-metrics.generated.js
"use strict";
// Copyright 2012-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppSyncMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class AppSyncMetrics {
    static _4XxErrorSum(dimensions) {
        return {
            namespace: 'AWS/AppSync',
            metricName: '4XXError',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static _5XxErrorSum(dimensions) {
        return {
            namespace: 'AWS/AppSync',
            metricName: '5XXError',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static latencyAverage(dimensions) {
        return {
            namespace: 'AWS/AppSync',
            metricName: 'Latency',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
}
exports.AppSyncMetrics = AppSyncMetrics;
