// aws-cdk-lib/aws-appsync/lib/runtime.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionRuntime = exports.FunctionRuntimeFamily = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Appsync supported runtimes. Only JavaScript as of now
 */
var FunctionRuntimeFamily;
(function (FunctionRuntimeFamily) {
    /**
     * AppSync JavaScript runtime
     */
    FunctionRuntimeFamily["JS"] = "APPSYNC_JS";
})(FunctionRuntimeFamily = exports.FunctionRuntimeFamily || (exports.FunctionRuntimeFamily = {}));
/**
 * Utility class for specifying specific appsync runtime versions
 */
class FunctionRuntime {
    constructor(family, version) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_appsync_FunctionRuntimeFamily(family);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, FunctionRuntime);
            }
            throw error;
        }
        this.name = family;
        this.version = version;
    }
    /**
     * Convert to Cfn runtime configuration property format
     */
    toProperties() {
        return {
            name: this.name,
            runtimeVersion: this.version,
        };
    }
}
_a = JSII_RTTI_SYMBOL_1;
FunctionRuntime[_a] = { fqn: "aws-cdk-lib.aws_appsync.FunctionRuntime", version: "2.75.1" };
/**
 * APPSYNC_JS v1.0.0 runtime
 */
FunctionRuntime.JS_1_0_0 = new FunctionRuntime(FunctionRuntimeFamily.JS, '1.0.0');
exports.FunctionRuntime = FunctionRuntime;
