// aws-cdk-lib/triggers/lib/trigger-function.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerFunction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _1 = require(".");
const lambda = require("../../aws-lambda");
/**
 * Invokes an AWS Lambda function during deployment.
 */
class TriggerFunction extends lambda.Function {
    constructor(scope, id, props) {
        super(scope, id, props);
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_triggers_TriggerFunctionProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, TriggerFunction);
            }
            throw error;
        }
        this.trigger = new _1.Trigger(this, 'Trigger', {
            ...props,
            handler: this,
        });
    }
    executeAfter(...scopes) {
        this.trigger.executeAfter(...scopes);
    }
    executeBefore(...scopes) {
        this.trigger.executeBefore(...scopes);
    }
}
_a = JSII_RTTI_SYMBOL_1;
TriggerFunction[_a] = { fqn: "aws-cdk-lib.triggers.TriggerFunction", version: "2.75.1" };
exports.TriggerFunction = TriggerFunction;
